---
title: "STREAM: A comprehensive guide to inferring enhancer-driven gene regulatory networks from scRNA-seq and scATAC-seq data"
output: html_document
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---

```{r setup, include = FALSE, warning=FALSE, message=FALSE}

library(knitr)
knitr::opts_chunk$set(echo = TRUE)
work.dir <- "/fs/ess/PCON0022/liyang/STREAM-BIB/tutorial-simul/"
opts_knit$set(root.dir = work.dir)

```


In this tutorial, we will guide you through the process of inferring enhancer regulons, known as eRegulons, and enhancer-driven 
gene regulatory networks or eGRNs, utilizing the STREAM tool. Our demonstration leverages a select subset of data from the 
publicly accessible 10x Genomics Multiome dataset, specifically focusing on human Peripheral Blood Mononuclear Cells (PBMCs).
For those interested in the comprehensive dataset, the raw data captured by the 10X Genomics Multiome ATAC+GEX can be downloaded directly from 
[10x Genomics](https://www.10xgenomics.com/resources/datasets/pbmc-from-a-healthy-donor-granulocytes-removed-through-cell-sorting-10-k-1-standard-1-0-0). 
As for our tutorial, the subsetted data we employ, stored in a Seurat object format, is available for download on 
[Figshare](https://figshare.com/ndownloader/files/42039747).

Initially, we ran STREAM on the PBMC data to infer eRegulons, each of which includes its regulating TF, its included genes and 
enhancers (and their relations), as well as the eRegulon-active cells. We evaluated these eRegulons from three perspectives, i.e., 
pathway enrichment of eRegulon genes, overlap of eRegulon enhancers against chromatin regions curated in databases (e.g., histone mark ChIP-seq peaks), 
and intersection of enhancer-gene relations against relations curated in databases (e.g., EnhancerAtlas 2.0 and scEnhancer). 
Then, we performed hyper-geometric test for eRegulon-active cells against each cell type. 
If the eRegulon-active cells are significantly enriched in a cell type (e.g., CD4 TCM), we denoted the corresponding eRegulon as 
cell-type-active eRegulon (CD4 TCM-active eRegulon). For each cell type, we merged all cell-type-active eRegulons, leading to 
an eGRN. Finally, by splitting the eGRN into sub-networks according to the regulating TFs, we obtained cell-type-specific eRegulons. 

```{r, eval = TRUE, echo = TRUE, warning=FALSE, message=FALSE}

dyn.load(x = "/users/PAS1475/liyang/libs/hdf5_1.10.6/lib/libhdf5_hl.so.100")
suppressWarnings(invisible(library(stream2) ) )

```


```{r, eval=FALSE, echo=FALSE, warning=FALSE, message=FALSE}

source.dir <- "/fs/ess/PCON0022/liyang/STREAM/Codes/stream_v2/R/"
code.list <- list.files(source.dir, pattern = ".R")
lapply(code.list, function(x) {
  message (x)
  source(paste0(source.dir, x))
})

```


## Data loading

Load a `Seurat` object containing jointly profiled scRNA-seq and scATAC-seq assays from human PBMCs.

```{r, warning=FALSE, message=FALSE, warning=FALSE, message=FALSE}

pbmc <- qs::qread(paste0(work.dir, "tutorial_pbmc.qsave"))
pbmc

```

## eRegulon inference

Run STREAM on the Seurat object with default parameters. 

In the function `run_stream`, variable annotations are presented as follows:

* `obj`: `Seurat` object composed of both RNA and ATAC assays

* `qubic.path`: File path of the `QUBIC2` biclustering program

* `peak.assay`: Name of the ATAC assay

* `var.genes`: Number of highly variable genes for LTMG modeling

* `top.peaks`: Number of top-ranked peaks to build heterogeneous graph

* `min.cells`: Minimum number of cells for quality control

* `org`: Organism assembly version

* `c.cutoff`: Cutoff of expression consistency among genes in the core part of hybrid biclusters

* `distance`: Size of window to connect co-accessible peaks

* `BlockOverlap`: Threshold of overlap between prior identified hybrid biclusters and the current seed

* `Extension`: Cutoff of expression consistency among genes in the dual part of hybrid biclusters

When it comes to jointly profiled RNA+ATAC data via different techniques, we suggest different strategies. For datasets with low scRNA-seq depth 
(e.g., [SHARE-seq](https://www.sciencedirect.com/science/article/pii/S0092867420312538)), 
we recommend imputation using mainstream approaches like [MAGIC](https://www.cell.com/cell/pdf/S0092-8674(18)30724-4.pdf), 
[SAVER](https://www.nature.com/articles/s41592-018-0033-z), and [scImpute](https://www.nature.com/articles/s41467-018-03405-7). 
For datasets with low coverage in both modalities (e.g., [SNARE-seq](https://www.nature.com/articles/s41587-019-0290-0) 
and [PAIRED-seq](https://www.nature.com/articles/s41594-019-0323-x)), 
we suggest using [SEACells](https://www.nature.com/articles/s41587-023-01716-9) to build metacells by pooling similar cells. 
For PAIRED-seq, which has even lower coverage than SNARE-seq, more stringent SEACells parameters should be used. 

```{r, warning=FALSE, message=FALSE}

# This process can be resource-intensive. We advise against executing it on a personal computer.
time1 <- system.time( en.regs <- run_stream(obj = pbmc,
                      qubic.path = "/users/PAS1475/liyang/software/QUBIC2/qubic",
                      peak.assay = "peaks",
                      var.genes = 3000,
                      top.peaks = 3000,
                      min.cells = 10,
                      org = "hg38",
                      c.cutoff = 1.0,
                      distance = 5e+05,
                      BlockOverlap = 0.50,
                      Extension = 1.0
                      ) )
qs::qsave(en.regs, paste0(work.dir, "eRegulons.qsave"))
names(en.regs[[1]])
head(en.regs[[1]]$genes)
head(en.regs[[1]]$peaks)
head(en.regs[[1]]$links)

```

```{r}
message ("The running time of eRegulon prediction is: ", time1["elapsed"])
```


## eGRN construction

Construct eGRN in each cell type.

In the function `get_cts_en_GRNs`, variabled are defined below:

* `obj`: `Seurat` object composed of both RNA and ATAC assays

* `celltype`: Name of the metadata column indicating cell types

* `en.regs`: List of predicted eRegulons

* `peak.assay`: Name of the ATAC assay

* `rna.dims`: Number of RNA dimensions for multimodal clustering using `Seurat`

* `atac.dims`: Number of ATAC dimensions for multimodal clustering using `Seurat`

* `padj.cutoff`: Threshold of adjusted P-values when mapping eRegulons onto cell types

* `out.dir`: Directory to save the final result and intermediate results

```{r, warning=FALSE, message=FALSE}

time2 <- system.time( en.grns <- get_cts_en_GRNs(obj = pbmc, celltype = "predicted.id",
                           en.regs = en.regs, peak.assay = "peaks",
                           rna.dims = 50, atac.dims = 50,
                           padj.cutoff = 0.05,
                           out.dir = "./") )
qs::qsave(en.grns, paste0(work.dir, "eGRNs.qsave"))
names(en.grns[[1]])
head(en.grns[[1]]$links)

```

```{r}
message ("The running time of eGRN construction is: ", time2["elapsed"])

```


## Cell-type-specific eRegulon prediction

Construct cell-type-specific eRegulons.

In the function `get_cts_en_regs`, variables are introduced as follows:

* `obj`: `Seurat` object composed of both RNA and ATAC assays

* `peak.assay`: Name of the ATAC assay

* `de.genes`: List of differentially expressed genes (DEGs)

* `cts.en.grns`: Cell-type-specific eGRNs saved in GRanges object, the name of each of which is cell type, and the GRanges object contains metadata columns "gene" and "TF".

* `out.dir`: Directory to save the final result and intermediate results

* `celltype`: Name of the metadata column indicating cell types

* `min.pct`: Cutoff of overlap percentage in calculating DEGs

* `logfc.threshold`: Threshold of log-fold change in predicting DEGs

* `padj.cutoff`: Cutoff of adjusted P-values in identifying DEGs

```{r, warning=FALSE, message=FALSE}

time3 <- system.time( cts.en.regs <- get_cts_en_regs(obj = pbmc, peak.assay = "peaks", de.genes = NULL,
                               cts.en.grns = en.grns, out.dir = "./", celltype = "predicted.id",
                               min.pct = 0.25, logfc.threshold = 0.25, padj.cutoff = 0.05) )
qs::qsave(cts.en.regs, paste0(work.dir, "cell-type-specific-eRegulons.qsave"))
names(cts.en.regs[[1]])
head(cts.en.regs[[1]]$genes)
head(cts.en.regs[[1]]$enhancers)
head(cts.en.regs[[1]]$links)

```

```{r}
message ("The running time of cell-type-specific eRegulon discovery is: ", time3["elapsed"])

```


## Evaluation of cell-type-specific eRegulons

Firstly, we performed pathway enrichment analysis for gene lists in cell-type-specific eRegulons against Gene Ontology (GO) or KEGG.

In the function `enrich_genes`, variables are defined as follows:

* `regs`: The list of enhancer regulons (eRegulons) or cell-type-specific eRegulons.

* `dbs`: The list of databases to run enrichment analysis, c("GO", "KEGG") by default.

```{r, warning=FALSE, message=FALSE}

library(pbapply)
time4 <- system.time( pathways <- enrich_genes(regs = cts.en.regs, dbs = "GO") )
sig.pathways <- pbapply::pblapply(pathways, function(x) {
  x[x$Adjusted.P.value < 0.05,, drop = FALSE]
})
qs::qsave(sig.pathways, paste0(work.dir, "enriched_pathways.qsave"))
head(sig.pathways[[1]])

```

```{r}
message ("The running time of pathway enrichment analysis for eRegulons is: ", time4["elapsed"])

```


Then, we intersected the enhancer list of a cell-type-specific eRegulon against H3K27ac ChIP-seq peaks in CD4+ T cells with ENCODE accession code: 
[ENCFF498QFU](https://www.encodeproject.org/files/ENCFF498QFU/).

In the function `intersect_peaks`, variables are described as follows:

* `x`: The first GRanges object or data.frame.

* `y`: The first GRanges object or data.frame.

```{r, warning=FALSE, message=FALSE}

library(dplyr)
library(Signac)
chipseq.df <- read.table(paste0(work.dir, "ENCFF498QFU.bed"), sep = "\t") %>% dplyr::select(V1, V2, V3)
chipseq.peaks <- StringToGRanges(paste0(chipseq.df[, 1], "-", chipseq.df[, 2], "-", chipseq.df[, 3]))
time5 <- system.time( permTest.res <- intersect_peaks(x = Signac::StringToGRanges(cts.en.regs[[1]]$enhancers), 
                                y = chipseq.peaks) )
permTest.res$numOverlaps

```

```{r}
message ("The running time of evaluating TF-enhancer relations in eRegulons is: ", time5['elapsed'])

```


Finally, we overlapped the enhancer-gene relations in a cell-type-specific eRegulon against enhancer-target pairs in CD4 TCM cells curated 
by [scEnhancer](http://enhanceratlas.net/scenhancer/data/download/interaction/hs/Memory_CD4_T_interaction.txt).

In the function `intersect_enhancer_gene_relations`, variables are introduced as follows:

* `x`: The first GRanges object saving enhancer-gene relations with gene symbols saved in "gene" meta column

* `y`: The second GRanges object saving enhancer-gene relations with gene symbols saved in "gene" meta column

```{r, warning=FALSE, message=FALSE}

# Preprocess the peak-gene interaction pairs in scEnhancer database
enhGene.df <- read.table(paste0(work.dir, "Memory_CD4_T_interaction.txt"), sep = "\t") %>% dplyr::select(V1, V2)
enhGene.pairs <- Signac::StringToGRanges(strsplit(enhGene.df$V1, split = "\\|") %>% sapply(., "[[", 1), sep = c(":", "-"))
mcols(enhGene.pairs)$gene <- strsplit(enhGene.df$V2, split = "[\\||\\:]") %>% sapply(., "[[", 3)
head(enhGene.pairs)

```


```{r, warning=FALSE, message=FALSE}

# Calculate the overlaps between enhancer-gene relations in a cell-type-specific eRegulon
time6 <- system.time(overlaps <- intersect_enhancer_gene_relations(x = resize(cts.en.regs[[1]]$links, width = 50000, fix = "center"), 
                                              y = resize(enhGene.pairs, width = 50000, fix = "center") ) )
overlaps

```

```{r}
message ("The running time of assessing enhancer-gene relations in eRegulons is: ", time6["elapsed"], "\n\n", 
         "The total time for the implementation and evaluation of eRegulon/eGRN prediction is: ", 
         time1["elapsed"] + time2["elapsed"] + time3["elapsed"] + time4["elapsed"] + time5["elapsed"] + time6["elapsed"])
```


```{r, eval=FALSE, echo=FALSE, warning=FALSE, message=FALSE}

# Check session information
sessionInfo()

```