<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Enhancer-driven gene regulatory network inference from single-cell RNA-seq and ATAC-seq  • stream</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Enhancer-driven gene regulatory network inference from single-cell RNA-seq and ATAC-seq ">
<meta property="og:description" content="We present a new algorithm, STREAM, for enhancer-driven gene regulatory network (eGRN) inference 
    from transcriptome and chromatin accessibility profiled from the same single cell populations. The algorithm 
    substantially improves the prediction accuracy of relations among transcription factors (TFs), enhancers, and 
    genes, by achieving global optimization based on two key new ideas: (i) we developed the Steiner forest problem
    (SFP) model based on a heterogeneous graph to identify the set of highly confident enhancer-gene relations which
    underlie a context-specific functional gene module (FGM); and (ii) we designed a hybrid biclustering pipeline
    integrated with submodular optimization for inferring eGRNs by identifying the optimal subset from a set of 
    hybrid biclusters (HBCs), each of which represents co-regulated genes by the same TF and, co-accessible 
    enhancers bound by the same TF, occurring over a subset of cells. These two key ideas are embedded in an iterative
    framework for eGRN inference through by finding patterns in a pair of transcriptome and chromatin accessibility
    matrices. Benchmarking analysis shows that the performance, assessed by f-scores, precision, or recall, was
    significantly improved by our program compared to four other state-of-the-art tools on ten single-cell sequencing
    datasets from seven sequencing techniques. The applicative powerbility of STREAM was demonstrated through two
    biological case studies. By implementing STREAM on an Alzheimer’s disease dataset over a time coursediffuse small
    lymphocytic lymphoma dataset, we showcased its capability to identify TF-enhancer-gene relations associated with
    pseudotime and investigate the changing of enhancer-gene alongside cell lineagesexcavated the key TF-enhancer-gene
    relations and cooperation among TFs underlying diseased cell types. Additionally, by implementing STREAM on a
    diffuse small lymphocytic lymphoma dataset, we excavated the key TF-enhancer-gene relations and cooperation among
    TFs underlying diseased cell types.STREAM showcased its capability to identify TF-enhancer-gene relations which
    were associated with pseudotime and investigate the changing of enhancer-gene alongside cell lineages from an
    Alzheimer’s disease dataset over a time course. ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">stream</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="articles/install.html">Install</a>
</li>
<li>
  <a href="news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/data-simulation.html">Simulated data analyses</a>
    </li>
    <li>
      <a href="articles/eRegulon-inference.html">Biological data analyses</a>
    </li>
  </ul>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/YangLi-Bio/stream2" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="stream">STREAM<a class="anchor" aria-label="anchor" href="#stream"></a>
</h1></div>
<p>STREAM (<strong>S</strong>ingle-cell enhancer regula<strong>T</strong>ory netwo<strong>R</strong>k inference from gene <strong>E</strong>xpression <strong>A</strong>nd chro<strong>M</strong>atin accessibility) is a computational method specifically designed to infer enhancer-driven gene regulatory networks (eGRNs) by analyzing transcriptome and chromatin accessibility jointly profiled from the same single cells. The method aims to improve the mapping accuracy of relationships between transcription factors (TFs), enhancers, and target genes. The algorithm employs two main strategies. First, it uses a <a href="https://www.sciencedirect.com/science/article/pii/S1570866709000628" class="external-link">Steiner Forest Problem (SFP) model</a> based on a heterogeneous graph to identify enhancer-gene relations that are highly relevant in a specific context. Second, it incorporates a hybrid biclustering approach combined with <a href="https://link.springer.com/article/10.1007/s10626-019-00308-7" class="external-link">submodular optimization</a> to determine eGRNs by selecting an optimal subset from sets of co-regulated genes and co-accessible enhancers. Based on these strategies, STREAM operates through an iterative framework to find enhancer-driven regulons (eRegulons) based on both transcriptome and chromatin accessibility data. eRegulons specific to a cell type collectively compose an eGRN. It is based on several popular packages for single-cell transcriptome and chromatin accessibility analysis, particularly, <a href="https://satijalab.org/seurat/" class="external-link">Seurat</a>, <a href="https://stuartlab.org/signac/" class="external-link">Signac</a>, <a href="https://cole-trapnell-lab.github.io/cicero-release/docs_m3/" class="external-link">cicero</a>, and <a href="https://bioconductor.org/packages/release/bioc/html/IRISFGM.html" class="external-link">IRIS-FGM</a>. Picture below showed the methodological details of STREAM for eGRN inference.</p>
<p><img src="reference/figures/details.jpg"></p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://github.com/YangLi-Bio/stream2" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/YangLi-Bio/stream2/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing stream</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Yang Li <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-7677-9028" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Anjun Ma <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0001-6269-398X" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li>Yizhong Wang <br><small class="roles"> Author </small>  </li>
<li>Qi Guo <br><small class="roles"> Author </small>  </li>
<li>Cankun Wang <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-0225-9855" target="orcid.widget" aria-label="ORCID" class="external-link"><span class="fab fa-orcid orcid" aria-hidden="true"></span></a> </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>



  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Yang Li, Anjun Ma, Yizhong Wang, Qi Guo, Cankun Wang.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
