<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>From Simulation to Insight: Inferring eRegulons and eGRNs using STREAM • stream</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="From Simulation to Insight: Inferring eRegulons and eGRNs using STREAM"><meta property="og:description" content="stream"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stream</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/install.html">Install</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/data-simulation.html">Simulated data analyses</a>
    </li>
    <li>
      <a href="../articles/eRegulon-inference.html">Biological data analyses</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/YangLi-Bio/stream2" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header>





<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>From Simulation to Insight: Inferring eRegulons
and eGRNs using STREAM</h1>
            
            <h4 data-toc-skip class="date">Compiled: August 26,
2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/YangLi-Bio/stream2/blob/HEAD/vignettes/data-simulation.Rmd" class="external-link"><code>vignettes/data-simulation.Rmd</code></a></small>
      <div class="hidden name"><code>data-simulation.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we will generate a simulated dataset that jointly
profiles scRNA-seq and scATAC-seq, drawing upon a <a
href="https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis#wnn-analysis-of-10x-multiome-rna-atac">Seurat
object</a> composed of both scRNA-seq and scATAC-seq assays, <a
href="https://saezlab.github.io/dorothea/">DoRothEA</a> — a gene
regulatory network (GRN) enriched with signed transcription factor (TF),
and <a href="https://jaspar.genereg.net/">JASPAR</a> — a collection of
curated, non-redundant TF binding profiles to target gene interactions.
Using this simulated dataset, we will subsequently infer enhancer
regulons (eRegulons) and derive enhancer-driven GRNs (eGRNs).</p>
<p>Initially, we selected a subset of TFs at random from the JASPAR 2022
database, along with their curated binding sites. Utilizing these TF
binding sites, we identified peaks associated with these TFs. From the
regulon genes in DoRothEA, we randomly chose certain genes as target
genes for these TFs. For each TF, we associated every target gene with a
set of TF binding peaks that are located within a predetermined distance
from the gene’s transcription start site (TSS). We consider these
eRegulons peaks as enhancers.</p>
<p>Upon determining the TFs, genes, and enhancers for the eRegulon, we
randomly sampled cell subsets where the eRegulons are active, denoted as
eRegulon-active cells. For every eRegulon gene, its expression value in
eRegulon-active cells was set to its highest expression value across all
cells, multiplied by a factor, e.g., one hundred. In a similar fashion,
we increased the chromatin accessibility value for each enhancer within
eRegulon-active cells.</p>
<p>Subsequently, we selected a distinct set of genes, peaks, and cells –
ones that were not considered as eRegulon genes, peaks, or
eRegulon-active cells – to serve as background genes, peaks, and cells.
In the end, we crafted scRNA-seq and scATAC-seq matrices by tailoring
the original matrices. These new matrices incorporate both eRegulon and
background genes/peaks as rows, and a combination of eRegulon-active and
background cells as columns.</p>
<p>Execute the commands below to simulate a dataset. The scRNA-seq
matrix has dimensions 1000 x 1000, while the scATAC-seq matrix is sized
at 3000 x 1000. The dataset encompasses five eRegulons. Each eRegulon
contains around 100 target genes, and each gene is modulated by a
transcription factor via roughly two enhancers.</p>
<p>Run the following commands to load the STREAM library:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">dyn.load</span>(<span class="at">x =</span> <span class="st">&quot;/users/PAS1475/liyang/libs/hdf5_1.10.6/lib/libhdf5_hl.so.100&quot;</span>)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="co"># library(easypackages)</span></span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="co"># libs &lt;- c(</span></span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="co">#   &quot;qs&quot;, </span></span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="co">#   &quot;hdf5r&quot;,</span></span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="co">#   &quot;Seurat&quot;,</span></span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="co">#   &quot;Signac&quot;, </span></span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="co">#   &quot;pbmcapply&quot;, </span></span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="co">#   &quot;pbapply&quot;, </span></span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="co">#   &quot;parallel&quot;,</span></span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="co">#   &quot;EnsDb.Mmusculus.v79&quot;, </span></span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="co">#   &quot;EnsDb.Hsapiens.v86&quot;,</span></span>
<span id="cb1-13"><a href="#cb1-13" tabindex="-1"></a><span class="co">#   &quot;dplyr&quot;, </span></span>
<span id="cb1-14"><a href="#cb1-14" tabindex="-1"></a><span class="co">#   &quot;ggplot2&quot;, </span></span>
<span id="cb1-15"><a href="#cb1-15" tabindex="-1"></a><span class="co">#   &quot;ggpubr&quot;, </span></span>
<span id="cb1-16"><a href="#cb1-16" tabindex="-1"></a><span class="co">#   &quot;igraph&quot;, </span></span>
<span id="cb1-17"><a href="#cb1-17" tabindex="-1"></a><span class="co">#   &quot;Matrix&quot;</span></span>
<span id="cb1-18"><a href="#cb1-18" tabindex="-1"></a><span class="co"># )</span></span>
<span id="cb1-19"><a href="#cb1-19" tabindex="-1"></a><span class="co"># libraries(libs)</span></span>
<span id="cb1-20"><a href="#cb1-20" tabindex="-1"></a></span>
<span id="cb1-21"><a href="#cb1-21" tabindex="-1"></a><span class="fu">library</span>(stream2)</span></code></pre></div>
<pre><code>## Loading required package: Biobase</code></pre>
<pre><code>## Loading required package: BiocGenerics</code></pre>
<pre><code>## 
## Attaching package: &#39;BiocGenerics&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:stats&#39;:
## 
##     IQR, mad, sd, var, xtabs</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     anyDuplicated, aperm, append, as.data.frame, basename, cbind,
##     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
##     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
##     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
##     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,
##     table, tapply, union, unique, unsplit, which.max, which.min</code></pre>
<pre><code>## Welcome to Bioconductor
## 
##     Vignettes contain introductory material; view with
##     &#39;browseVignettes()&#39;. To cite Bioconductor, see
##     &#39;citation(&quot;Biobase&quot;)&#39;, and for packages &#39;citation(&quot;pkgname&quot;)&#39;.</code></pre>
<pre><code>## The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,
## which was just loaded, will retire in October 2023.
## Please refer to R-spatial evolution reports for details, especially
## https://r-spatial.org/r/2023/05/15/evolution4.html.
## It may be desirable to make the sf package available;
## package maintainers should consider adding sf to Suggests:.
## The sp package is now running under evolution status 2
##      (status 2 uses the sf package in place of rgdal)</code></pre>
<pre><code>## Warning: replacing previous import &#39;Seurat::Assays&#39; by
## &#39;SummarizedExperiment::Assays&#39; when loading &#39;stream2&#39;</code></pre>
<pre><code>## Warning: replacing previous import &#39;SummarizedExperiment::shift&#39; by
## &#39;data.table::shift&#39; when loading &#39;stream2&#39;</code></pre>
<pre><code>## Warning: replacing previous import &#39;data.table::last&#39; by &#39;dplyr::last&#39; when
## loading &#39;stream2&#39;</code></pre>
<pre><code>## Warning: replacing previous import &#39;data.table::first&#39; by &#39;dplyr::first&#39; when
## loading &#39;stream2&#39;</code></pre>
<pre><code>## Warning: replacing previous import &#39;data.table::between&#39; by &#39;dplyr::between&#39;
## when loading &#39;stream2&#39;</code></pre>
<pre><code>## Warning: replacing previous import &#39;dplyr::union&#39; by &#39;igraph::union&#39; when
## loading &#39;stream2&#39;</code></pre>
<pre><code>## Warning: replacing previous import &#39;dplyr::as_data_frame&#39; by
## &#39;igraph::as_data_frame&#39; when loading &#39;stream2&#39;</code></pre>
<pre><code>## Warning: replacing previous import &#39;dplyr::groups&#39; by &#39;igraph::groups&#39; when
## loading &#39;stream2&#39;</code></pre>
<div id="data-simulation" class="section level2">
<h2>Data simulation</h2>
<p>Load a Seurat object that comprises jointly profiled scRNA-seq and
scATAC-seq assays derived from human PBMCs and produce simulated data.
Based on this object, construct a simulated Seurat object encompassing
1,000 genes, 3,000 peaks, and 1,000 cells. This object features five
eRegulons, with each eRegulon linked to roughly five target genes and
approximately two enhancers. These eRegulons are active in 100
cells.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a>pbmc <span class="ot">&lt;-</span> qs<span class="sc">::</span><span class="fu">qread</span>(<span class="fu">paste0</span>(work.dir, <span class="st">&quot;tutorial_pbmc.qsave&quot;</span>))</span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a>pbmc</span></code></pre></div>
<pre><code>## An object of class Seurat 
## 6193 features across 1000 samples within 3 assays 
## Active assay: peaks (3991 features, 3991 variable features)
##  2 other assays present: RNA, SCT
##  3 dimensional reductions calculated: pca, lsi, umap</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a>simul <span class="ot">&lt;-</span> <span class="fu">create_rna_atac</span>(<span class="at">obj =</span> pbmc, <span class="at">ntfs =</span> <span class="dv">5</span>, <span class="at">ngenes =</span> <span class="dv">100</span>,</span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a>                         <span class="at">ncells =</span> <span class="dv">100</span>, <span class="at">all.genes =</span> <span class="dv">1000</span>, <span class="at">all.enhs =</span> <span class="dv">3000</span>, <span class="at">all.cells =</span> <span class="dv">1000</span>,</span>
<span id="cb19-3"><a href="#cb19-3" tabindex="-1"></a>                         <span class="at">org =</span> <span class="st">&quot;hg38&quot;</span>, <span class="at">atac.assay =</span> <span class="st">&quot;peaks&quot;</span>, <span class="at">gene.links =</span> <span class="dv">2</span></span>
<span id="cb19-4"><a href="#cb19-4" tabindex="-1"></a>                         )</span></code></pre></div>
<pre><code>## Loading TF binding sites from JASPAR 2022 ...</code></pre>
<pre><code>## There are 203 TFs from the JASPAR 2022 database.
## 5 eRegulons regulated by different TFs will be generated.
## The Seurat object contains 1101 genes, 3991 enhancers, and 1000 cells.</code></pre>
<pre><code>## There are 197 candidate TFs curated by JASPAR 2022</code></pre>
<pre><code>## Loaded 1207 TFs collected in DoRothEA database.</code></pre>
<pre><code>## There are 500 common selected TFs between JASPAR and DoRothEA.</code></pre>
<pre><code>## Generated regulons composed of 1-142 genes.</code></pre>
<pre><code>## Loading required package: EnsDb.Hsapiens.v75</code></pre>
<pre><code>## Loading required package: ensembldb</code></pre>
<pre><code>## Loading required package: GenomicRanges</code></pre>
<pre><code>## Loading required package: stats4</code></pre>
<pre><code>## Loading required package: S4Vectors</code></pre>
<pre><code>## 
## Attaching package: &#39;S4Vectors&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:base&#39;:
## 
##     expand.grid, I, unname</code></pre>
<pre><code>## Loading required package: IRanges</code></pre>
<pre><code>## Loading required package: GenomeInfoDb</code></pre>
<pre><code>## Loading required package: GenomicFeatures</code></pre>
<pre><code>## Loading required package: AnnotationDbi</code></pre>
<pre><code>## Loading required package: AnnotationFilter</code></pre>
<pre><code>## 
## Attaching package: &#39;ensembldb&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:stats&#39;:
## 
##     filter</code></pre>
<pre><code>## Finding nearby genes for each peak within 5e+05 bp ...</code></pre>
<pre><code>## Identified 2849 enhancers within 5e+05 from the selected genes.</code></pre>
<pre><code>## Identified the binding enhancers of 65 TFs, ranging from 47 to 322803 bound enhancers</code></pre>
<pre><code>## Generated 65 eRegulons composed of 0-28 enhancer-gene relations.</code></pre>
<pre><code>## Retained 5 eRegulons composed of 12-28 enhancer-gene relations.</code></pre>
<pre><code>## Began building hybrid biclusters (HBCs) based on simulated eRegulons ...</code></pre>
<pre><code>## 
  |                                                        
  |                                                  |   0%
  |                                                        
  |==========                                        |  20%
  |                                                        
  |====================                              |  40%
  |                                                        
  |==============================                    |  60%
  |                                                        
  |========================================          |  80%
  |                                                        
  |==================================================| 100%</code></pre>
<pre><code>## Simulated hybrid biclusters (HBCs) cover 62 genes, 78 enhancers, and 431 cells.</code></pre>
<pre><code>## We have 938 genes, 2922 enhancers, and 569 cells as background.</code></pre>
<pre><code>## Generated simulated scRNA-seq and scATAC-seq matrices of sizes: 1000 x 1000 and 3000 x 1000, respectively.</code></pre>
<pre><code>## Loaded scRNA-seq and scATAC-seq matrices of sizes: 1000 x 1000 and 3000 x 1000.</code></pre>
<pre><code>## The organism version is hg38.</code></pre>
<pre><code>## Loading the database EnsDb.Hsapiens.v86 for assembly hg38</code></pre>
<pre><code>## Loading required package: EnsDb.Hsapiens.v86</code></pre>
<pre><code>## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)

## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use
##   suppressWarnings() to suppress this warning.)</code></pre>
<pre><code>## The simulated scRNA-seq and scATAC-seq data contains:
## 1. Hybrid biclusters (HBCs) in nested list,
## 2. Seurat object containing the scRNA-seq and scATAC-seq data.</code></pre>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" tabindex="-1"></a>simul<span class="sc">$</span>Seurat</span></code></pre></div>
<pre><code>## An object of class Seurat 
## 2836 features across 1000 samples within 2 assays 
## Active assay: RNA (1000 features, 0 variable features)
##  1 other assay present: ATAC</code></pre>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" tabindex="-1"></a><span class="fu">head</span>(simul<span class="sc">$</span>HBCs[[<span class="dv">1</span>]]<span class="sc">$</span>genes)</span></code></pre></div>
<pre><code>## [1] &quot;STX11&quot; &quot;NRP1&quot;  &quot;DOCK9&quot; &quot;TYMP&quot;  &quot;IL1R1&quot; &quot;GCH1&quot;</code></pre>
<div class="sourceCode" id="cb60"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" tabindex="-1"></a><span class="fu">head</span>(simul<span class="sc">$</span>HBCs[[<span class="dv">1</span>]]<span class="sc">$</span>enhancers)</span></code></pre></div>
<pre><code>## [1] &quot;chr6-144149667-144151675&quot; &quot;chr10-33134015-33134576&quot; 
## [3] &quot;chr10-33264223-33264655&quot;  &quot;chr13-99322063-99323136&quot; 
## [5] &quot;chr22-50538228-50538489&quot;  &quot;chr2-102226906-102228140&quot;</code></pre>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1" tabindex="-1"></a><span class="fu">head</span>(simul<span class="sc">$</span>HBCs[[<span class="dv">1</span>]]<span class="sc">$</span>links)</span></code></pre></div>
<pre><code>## GRanges object with 6 ranges and 1 metadata column:
##       seqnames              ranges strand |        gene
##          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   [1]     chr6 144149667-144151675      * |       STX11
##   [2]    chr10   33134015-33134576      * |        NRP1
##   [3]    chr10   33264223-33264655      * |        NRP1
##   [4]    chr13   99322063-99323136      * |       DOCK9
##   [5]    chr22   50538228-50538489      * |        TYMP
##   [6]     chr2 102226906-102228140      * |       IL1R1
##   -------
##   seqinfo: 23 sequences from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" tabindex="-1"></a><span class="fu">sapply</span>(simul<span class="sc">$</span>HBCs, <span class="st">&quot;[[&quot;</span>, <span class="st">&quot;TF&quot;</span>)</span></code></pre></div>
<pre><code>## [1] &quot;BATF3&quot;  &quot;KLF15&quot;  &quot;TCF12&quot;  &quot;ZSCAN4&quot; &quot;MEF2A&quot;</code></pre>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb66-1"><a href="#cb66-1" tabindex="-1"></a><span class="co"># Check session information</span></span>
<span id="cb66-2"><a href="#cb66-2" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.2.1 (2022-06-23)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Red Hat Enterprise Linux
## 
## Matrix products: default
## BLAS/LAPACK: /opt/intel/2021.3/mkl/2021.3.0/lib/intel64/libmkl_gf_lp64.so.1
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
##  [1] EnsDb.Hsapiens.v86_2.99.0 EnsDb.Hsapiens.v75_2.99.0
##  [3] ensembldb_2.22.0          AnnotationFilter_1.22.0  
##  [5] GenomicFeatures_1.50.4    AnnotationDbi_1.60.2     
##  [7] GenomicRanges_1.50.2      GenomeInfoDb_1.34.9      
##  [9] IRanges_2.32.0            S4Vectors_0.36.2         
## [11] stream2_2.0.0             Biobase_2.58.0           
## [13] BiocGenerics_0.44.0       knitr_1.43               
## 
## loaded via a namespace (and not attached):
##   [1] rappdirs_0.3.3              rtracklayer_1.58.0         
##   [3] scattermore_1.2             SeuratObject_4.1.3         
##   [5] ragg_1.2.5                  tidyr_1.3.0                
##   [7] ggplot2_3.4.2               bit64_4.0.5                
##   [9] irlba_2.3.5.1               DelayedArray_0.24.0        
##  [11] data.table_1.14.8           rpart_4.1.16               
##  [13] KEGGREST_1.38.0             RCurl_1.98-1.12            
##  [15] generics_0.1.3              terra_1.7-39               
##  [17] cowplot_1.1.1               RSQLite_2.3.1              
##  [19] RApiSerialize_0.1.2         RANN_2.6.1                 
##  [21] future_1.33.0               bit_4.0.5                  
##  [23] spatstat.data_3.0-1         xml2_1.3.5                 
##  [25] httpuv_1.6.11               SummarizedExperiment_1.28.0
##  [27] xfun_0.40                   hms_1.1.3                  
##  [29] jquerylib_0.1.4             evaluate_0.21              
##  [31] promises_1.2.1              fansi_1.0.4                
##  [33] restfulr_0.0.15             progress_1.2.2             
##  [35] dbplyr_2.3.3                igraph_1.5.0               
##  [37] DBI_1.1.3                   htmlwidgets_1.6.2          
##  [39] spatstat.geom_3.2-2         purrr_1.0.2                
##  [41] ellipsis_0.3.2              dplyr_1.1.2                
##  [43] backports_1.4.1             RcppParallel_5.1.7         
##  [45] biomaRt_2.54.1              deldir_1.0-9               
##  [47] MatrixGenerics_1.10.0       vctrs_0.6.3                
##  [49] SingleCellExperiment_1.20.1 ROCR_1.0-11                
##  [51] abind_1.4-5                 cachem_1.0.8               
##  [53] withr_2.5.0                 BSgenome_1.66.3            
##  [55] progressr_0.14.0            checkmate_2.2.0            
##  [57] sctransform_0.3.5           GenomicAlignments_1.34.1   
##  [59] prettyunits_1.1.1           goftest_1.2-3              
##  [61] cluster_2.1.3               lazyeval_0.2.2             
##  [63] crayon_1.5.2                spatstat.explore_3.2-1     
##  [65] pkgconfig_2.0.3             nlme_3.1-157               
##  [67] ProtGenerics_1.30.0         nnet_7.3-17                
##  [69] rlang_1.1.1                 globals_0.16.2             
##  [71] lifecycle_1.0.3             miniUI_0.1.1.1             
##  [73] filelock_1.0.2              BiocFileCache_2.6.1        
##  [75] dichromat_2.0-0.1           rprojroot_2.0.3            
##  [77] polyclip_1.10-4             matrixStats_1.0.0          
##  [79] lmtest_0.9-40               Matrix_1.6-0               
##  [81] boot_1.3-28                 zoo_1.8-12                 
##  [83] base64enc_0.1-3             ggridges_0.5.4             
##  [85] png_0.1-8                   viridisLite_0.4.2          
##  [87] rjson_0.2.21                stringfish_0.15.8          
##  [89] bitops_1.0-7                KernSmooth_2.23-20         
##  [91] Biostrings_2.66.0           blob_1.2.4                 
##  [93] stringr_1.5.0               rTRM_1.36.0                
##  [95] parallelly_1.36.0           spatstat.random_3.1-5      
##  [97] scales_1.2.1                memoise_2.0.1              
##  [99] magrittr_2.0.3              plyr_1.8.8                 
## [101] ica_1.0-3                   zlibbioc_1.44.0            
## [103] compiler_4.2.1              BiocIO_1.8.0               
## [105] RColorBrewer_1.1-3          lme4_1.1-34                
## [107] fitdistrplus_1.1-11         Rsamtools_2.14.0           
## [109] cli_3.6.1                   XVector_0.38.0             
## [111] listenv_0.9.0               patchwork_1.1.2            
## [113] pbapply_1.7-2               htmlTable_2.4.1            
## [115] Formula_1.2-5               MASS_7.3-57                
## [117] tidyselect_1.2.0            stringi_1.7.12             
## [119] textshaping_0.3.6           yaml_2.3.7                 
## [121] ggrepel_0.9.3               pbmcapply_1.5.1            
## [123] grid_4.2.1                  sass_0.4.7                 
## [125] VariantAnnotation_1.44.1    bcellViper_1.34.0          
## [127] fastmatch_1.1-4             tools_4.2.1                
## [129] future.apply_1.11.0         parallel_4.2.1             
## [131] rstudioapi_0.15.0           qs_0.25.5                  
## [133] foreign_0.8-82              gridExtra_2.3              
## [135] Rtsne_0.16                  digest_0.6.33              
## [137] shiny_1.7.4.1               monocle3_1.3.1             
## [139] Rcpp_1.0.11                 later_1.3.1                
## [141] RcppAnnoy_0.0.21            httr_1.4.6                 
## [143] biovizBase_1.46.0           colorspace_2.1-0           
## [145] XML_3.99-0.14               fs_1.6.3                   
## [147] tensor_1.5                  reticulate_1.31            
## [149] splines_4.2.1               uwot_0.1.16                
## [151] RcppRoll_0.3.0              spatstat.utils_3.0-3       
## [153] pkgdown_2.0.7               sp_2.0-0                   
## [155] plotly_4.10.2               systemfonts_1.0.4          
## [157] dorothea_1.10.0             xtable_1.8-4               
## [159] jsonlite_1.8.7              nloptr_2.0.3               
## [161] R6_2.5.1                    Hmisc_5.1-0                
## [163] pillar_1.9.0                htmltools_0.5.6            
## [165] mime_0.12                   glue_1.6.2                 
## [167] fastmap_1.1.1               minqa_1.2.5                
## [169] BiocParallel_1.32.6         codetools_0.2-18           
## [171] Signac_1.10.0               utf8_1.2.3                 
## [173] lattice_0.20-45             bslib_0.5.1                
## [175] spatstat.sparse_3.0-2       tibble_3.2.1               
## [177] curl_5.0.1                  leiden_0.4.3               
## [179] survival_3.3-1              rmarkdown_2.23             
## [181] desc_1.4.2                  munsell_0.5.0              
## [183] GenomeInfoDbData_1.2.9      reshape2_1.4.4             
## [185] gtable_0.3.4                Seurat_4.3.0.1</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
      </div>

</div>



      <footer><div class="copyright">
  <p></p><p>Developed by Yang Li, Anjun Ma, Yizhong Wang, Qi Guo, Cankun Wang, Bingqiang Liu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>
