<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>From Simulation to Insight: Inferring eRegulons and eGRNs using STREAM • stream</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="From Simulation to Insight: Inferring eRegulons and eGRNs using STREAM">
<meta property="og:description" content="stream">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stream</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/install.html">Install</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data-simulation.html">Simulated data analyses</a>
    </li>
    <li>
      <a href="../articles/eRegulon-inference.html">Biological data analyses</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/YangLi-Bio/stream2" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>From Simulation to Insight: Inferring eRegulons
and eGRNs using STREAM</h1>
            
            <h4 data-toc-skip class="date">Compiled: June 16, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/YangLi-Bio/stream2/blob/HEAD/vignettes/data-simulation.Rmd" class="external-link"><code>vignettes/data-simulation.Rmd</code></a></small>
      <div class="hidden name"><code>data-simulation.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we will generate a simulated dataset that jointly
profiles scRNA-seq and scATAC-seq, drawing upon a <a href="https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis#wnn-analysis-of-10x-multiome-rna-atac" class="external-link">Seurat
object</a> composed of both scRNA-seq and scATAC-seq assays, <a href="https://saezlab.github.io/dorothea/" class="external-link">DoRothEA</a> — a gene
regulatory network (GRN) enriched with signed transcription factor (TF),
and <a href="https://jaspar.genereg.net/" class="external-link">JASPAR</a> — a collection of
curated, non-redundant TF binding profiles to target gene interactions.
Using this simulated dataset, we will subsequently infer enhancer
regulons (eRegulons) and derive enhancer-driven GRNs (eGRNs).</p>
<p>Initially, we selected a subset of TFs at random from the JASPAR 2022
database, along with their curated binding sites. Utilizing these TF
binding sites, we identified peaks associated with these TFs. From the
regulon genes in <code>DoRothEA</code>, we randomly chose certain genes
as target genes for these TFs. For each TF, we associated every target
gene with a set of TF binding peaks that are located within a
predetermined distance from the gene’s transcription start site (TSS).
We consider these eRegulons peaks as enhancers.</p>
<p>Upon determining the TFs, genes, and enhancers for the eRegulon, we
randomly sampled cell subsets where the eRegulons are active, denoted as
eRegulon-active cells. For every eRegulon gene, its expression value in
eRegulon-active cells was set to its highest expression value across all
cells, multiplied by a factor, e.g., one hundred. In a similar fashion,
we increased the chromatin accessibility value for each enhancer within
eRegulon-active cells.</p>
<p>Subsequently, we selected a distinct set of genes, peaks, and cells –
ones that were not considered as eRegulon genes, peaks, or
eRegulon-active cells – to serve as background genes, peaks, and cells.
In the end, we crafted scRNA-seq and scATAC-seq matrices by tailoring
the original matrices. These new matrices incorporate both eRegulon and
background genes/peaks as rows, and a combination of eRegulon-active and
background cells as columns.</p>
<p>Execute the commands below to simulate a dataset. The scRNA-seq
matrix has dimensions 1000 x 1000, while the scATAC-seq matrix is sized
at 3000 x 1000. The dataset encompasses five eRegulons. Each eRegulon
contains around 100 target genes, and each gene is modulated by a
transcription factor via roughly two enhancers.</p>
<p>Run the following commands to load the <code>STREAM</code>
library:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dynload.html" class="external-link">dyn.load</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"/users/PAS1475/liyang/libs/hdf5_1.10.6/lib/libhdf5_hl.so.100"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/YangLi-Bio/stream2" class="external-link">stream2</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="data-simulation">Data simulation<a class="anchor" aria-label="anchor" href="#data-simulation"></a>
</h2>
<p>Load a Seurat object that contains jointly profiled scRNA-seq and
scATAC-seq assays derived from human PBMCs, and generate simulated data
based on this object. Construct a simulated Seurat object comprising
1,000 genes, 3,000 peaks, and 1,000 cells. This object should feature
five eRegulons regulated by five different TFs, with each eRegulon
linked to approximately fifty target genes. Each target gene should be
associated with roughly two enhancers. These eRegulons should be active
in 100 cells.</p>
<p>In the function <code>create_rna_atac</code>, variable annotations
are given as follows:</p>
<ul>
<li><p><code>obj</code>: <code>Seurat</code> object composed of both RNA
and ATAC assays</p></li>
<li><p><code>ntfs</code>: Number of eRegulons/TFs to be
implanted</p></li>
<li><p><code>ngenes</code>: Average number of genes in eRegulons to be
implanted</p></li>
<li><p><code>ncells</code>: Average number of cells in eRegulons to be
implanted</p></li>
<li><p><code>all.genes</code>: Number of genes in dataset to be
simulated</p></li>
<li><p><code>all.enhs</code>: Number of peaks in dataset to be
simulated</p></li>
<li><p><code>all.cells</code>: Number of cells in dataset to be
simulated</p></li>
<li><p><code>org</code>: Organism assembly version</p></li>
<li><p><code>atac.assay</code>: Name of the ATAC assay</p></li>
<li><p><code>gene.links</code>: Average number of enhancers linked to
each gene</p></li>
</ul>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html" class="external-link">qread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">work.dir</span>, <span class="st">"tutorial_pbmc.qsave"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pbmc</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 6193 features across 1000 samples within 3 assays </span></span>
<span><span class="co">## Active assay: peaks (3991 features, 3991 variable features)</span></span>
<span><span class="co">##  2 other assays present: RNA, SCT</span></span>
<span><span class="co">##  3 dimensional reductions calculated: pca, lsi, umap</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simul</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_rna_atac.html">create_rna_atac</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">pbmc</span>, ntfs <span class="op">=</span> <span class="fl">5</span>, ngenes <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                         ncells <span class="op">=</span> <span class="fl">100</span>, all.genes <span class="op">=</span> <span class="fl">1000</span>, all.enhs <span class="op">=</span> <span class="fl">3000</span>, all.cells <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                         org <span class="op">=</span> <span class="st">"hg38"</span>, atac.assay <span class="op">=</span> <span class="st">"peaks"</span>, gene.links <span class="op">=</span> <span class="fl">2</span></span>
<span>                         <span class="op">)</span></span></code></pre></div>
<pre><code>## 
  |                                                        
  |                                                  |   0%
  |                                                        
  |==========                                        |  20%
  |                                                        
  |====================                              |  40%
  |                                                        
  |==============================                    |  60%
  |                                                        
  |========================================          |  80%
  |                                                        
  |==================================================| 100%</code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simul</span><span class="op">$</span><span class="va">Seurat</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 2811 features across 1000 samples within 2 assays </span></span>
<span><span class="co">## Active assay: RNA (1000 features, 0 variable features)</span></span>
<span><span class="co">##  1 other assay present: ATAC</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">simul</span><span class="op">$</span><span class="va">HBCs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ITGB1"   "ANK3"    "PLEKHG1" "DUSP6"   "ZDHHC14" "ARMC4"</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">simul</span><span class="op">$</span><span class="va">HBCs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">enhancers</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "chr10-33134015-33134576"  "chr10-33264223-33264655" </span></span>
<span><span class="co">## [3] "chr10-62050806-62051384"  "chr10-62120709-62121465" </span></span>
<span><span class="co">## [5] "chr6-150622501-150623717" "chr12-89768730-89769666"</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">simul</span><span class="op">$</span><span class="va">HBCs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">links</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 1 metadata column:</span></span>
<span><span class="co">##       seqnames              ranges strand |        gene</span></span>
<span><span class="co">##          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##   [1]    chr10   33134015-33134576      * |       ITGB1</span></span>
<span><span class="co">##   [2]    chr10   33264223-33264655      * |       ITGB1</span></span>
<span><span class="co">##   [3]    chr10   62050806-62051384      * |        ANK3</span></span>
<span><span class="co">##   [4]    chr10   62120709-62121465      * |        ANK3</span></span>
<span><span class="co">##   [5]     chr6 150622501-150623717      * |     PLEKHG1</span></span>
<span><span class="co">##   [6]    chr12   89768730-89769666      * |       DUSP6</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">simul</span><span class="op">$</span><span class="va">HBCs</span>, <span class="st">"[["</span>, <span class="st">"TF"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "BATF3" "KLF15" "BATF"  "CEBPA" "TCF12"</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yang Li, Anjun Ma, Yizhong Wang, Qi Guo, Cankun Wang.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
