<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>From Simulation to Insight: Inferring eRegulons and eGRNs using STREAM • stream2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="From Simulation to Insight: Inferring eRegulons and eGRNs using STREAM">
<meta property="og:description" content="stream2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stream2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/install.html">Install</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/simul-analyses.html">Simulated data analyses</a>
    </li>
    <li>
      <a href="../articles/bio-analyses.html">Biological data analyses</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/YangLi-Bio/stream2" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>From Simulation to Insight: Inferring eRegulons
and eGRNs using STREAM</h1>
            
            <h4 data-toc-skip class="date">Compiled: August 20,
2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/YangLi-Bio/stream2/blob/HEAD/vignettes/data-simulation.Rmd" class="external-link"><code>vignettes/data-simulation.Rmd</code></a></small>
      <div class="hidden name"><code>data-simulation.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we will generate a simulated dataset that jointly
profiles scRNA-seq and scATAC-seq, drawing upon a [Seurat object] (<a href="https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis#wnn-analysis-of-10x-multiome-rna-atac" class="external-link uri">https://satijalab.org/seurat/articles/weighted_nearest_neighbor_analysis#wnn-analysis-of-10x-multiome-rna-atac</a>)
composed of both scRNA-seq and scATAC-seq assays, [DoRothEA] (<a href="https://saezlab.github.io/dorothea/" class="external-link uri">https://saezlab.github.io/dorothea/</a>) — a gene regulatory
network (GRN) enriched with signed transcription factor (TF), and
[JASPAR] (<a href="https://jaspar.genereg.net/" class="external-link uri">https://jaspar.genereg.net/</a>) — a collection of curated,
non-redundant TF binding profiles to target gene interactions. Using
this simulated dataset, we will subsequently infer enhancer regulons
(eRegulons) and derive enhancer-driven GRNs (eGRNs).</p>
<p>Initially, we selected a subset of TFs at random from the JASPAR 2022
database, along with their curated binding sites. Utilizing these TF
binding sites, we identified peaks associated with these TFs. From the
regulon genes in DoRothEA, we randomly chose certain genes as target
genes for these TFs. For each TF, we associated every target gene with a
set of TF binding peaks that are located within a predetermined distance
from the gene’s transcription start site (TSS). We consider these
eRegulons peaks as enhancers.</p>
<p>Upon determining the TFs, genes, and enhancers for the eRegulon, we
randomly sampled cell subsets where the eRegulons are active, denoted as
eRegulon-active cells. For every eRegulon gene, its expression value in
eRegulon-active cells was set to its highest expression value across all
cells, multiplied by a factor, e.g., one hundred. In a similar fashion,
we increased the chromatin accessibility value for each enhancer within
eRegulon-active cells.</p>
<p>Subsequently, we selected a distinct set of genes, peaks, and cells –
ones that were not considered as eRegulon genes, peaks, or
eRegulon-active cells – to serve as background genes, peaks, and cells.
In the end, we crafted scRNA-seq and scATAC-seq matrices by tailoring
the original matrices. These new matrices incorporate both eRegulon and
background genes/peaks as rows, and a combination of eRegulon-active and
background cells as columns.</p>
<p>Execute the commands below to simulate a dataset. The scRNA-seq
matrix has dimensions 1000 x 1000, while the scATAC-seq matrix is sized
at 3000 x 1000. The dataset encompasses five eRegulons. Each eRegulon
contains around 100 target genes, and each gene is modulated by a
transcription factor via roughly two enhancers.</p>
<p>Run the following commands to load the STREAM library:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dynload.html" class="external-link">dyn.load</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"/users/PAS1475/liyang/libs/hdf5_1.10.6/lib/libhdf5_hl.so.100"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">easypackages</span><span class="op">)</span></span>
<span><span class="va">libs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"qs"</span>, </span>
<span>  <span class="st">"hdf5r"</span>,</span>
<span>  <span class="st">"Seurat"</span>,</span>
<span>  <span class="st">"Signac"</span>, </span>
<span>  <span class="st">"pbmcapply"</span>, </span>
<span>  <span class="st">"pbapply"</span>, </span>
<span>  <span class="st">"parallel"</span>,</span>
<span>  <span class="st">"EnsDb.Mmusculus.v79"</span>, </span>
<span>  <span class="st">"EnsDb.Hsapiens.v86"</span>,</span>
<span>  <span class="st">"dplyr"</span>, </span>
<span>  <span class="st">"ggplot2"</span>, </span>
<span>  <span class="st">"ggpubr"</span>, </span>
<span>  <span class="st">"igraph"</span>, </span>
<span>  <span class="st">"Matrix"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/easypackages/man/libraries.html" class="external-link">libraries</a></span><span class="op">(</span><span class="va">libs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: qs</span></span></code></pre>
<pre><code><span><span class="co">## qs 0.25.5</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: hdf5r</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: Seurat</span></span></code></pre>
<pre><code><span><span class="co">## The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,</span></span>
<span><span class="co">## which was just loaded, will retire in October 2023.</span></span>
<span><span class="co">## Please refer to R-spatial evolution reports for details, especially</span></span>
<span><span class="co">## https://r-spatial.org/r/2023/05/15/evolution4.html.</span></span>
<span><span class="co">## It may be desirable to make the sf package available;</span></span>
<span><span class="co">## package maintainers should consider adding sf to Suggests:.</span></span>
<span><span class="co">## The sp package is now running under evolution status 2</span></span>
<span><span class="co">##      (status 2 uses the sf package in place of rgdal)</span></span></code></pre>
<pre><code><span><span class="co">## Attaching SeuratObject</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: Signac</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: pbmcapply</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: parallel</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: pbapply</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: EnsDb.Mmusculus.v79</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: ensembldb</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: BiocGenerics</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'BiocGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     IQR, mad, sd, var, xtabs</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">##     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">##     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">##     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">##     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">##     table, tapply, union, unique, unsplit, which.max, which.min</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomicRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: stats4</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: S4Vectors</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'S4Vectors'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:hdf5r':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     values</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand.grid, I, unname</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: IRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomeInfoDb</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomicFeatures</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: AnnotationDbi</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: Biobase</span></span></code></pre>
<pre><code><span><span class="co">## Welcome to Bioconductor</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Vignettes contain introductory material; view with</span></span>
<span><span class="co">##     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">##     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: AnnotationFilter</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'ensembldb'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     filter</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: EnsDb.Hsapiens.v86</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: dplyr</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'dplyr'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:ensembldb':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     filter, select</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:AnnotationDbi':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     select</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     combine</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:GenomicRanges':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect, setdiff, union</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:GenomeInfoDb':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:IRanges':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     collapse, desc, intersect, setdiff, slice, union</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:S4Vectors':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     first, intersect, rename, setdiff, setequal, union</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:BiocGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     combine, intersect, setdiff, union</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     filter, lag</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect, setdiff, setequal, union</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: ggplot2</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: ggpubr</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: igraph</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'igraph'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:dplyr':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     as_data_frame, groups, union</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:GenomicRanges':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     union</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:IRanges':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     union</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:S4Vectors':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     union</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:BiocGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     normalize, path, union</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     decompose, spectrum</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     union</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: Matrix</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Matrix'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:S4Vectors':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand</span></span></code></pre>
<pre><code><span><span class="co">## All packages loaded successfully</span></span></code></pre>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/YangLi-Bio/stream2" class="external-link">stream2</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: replacing previous import 'SummarizedExperiment::shift' by</span></span>
<span><span class="co">## 'data.table::shift' when loading 'stream2'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'data.table::last' by 'dplyr::last' when</span></span>
<span><span class="co">## loading 'stream2'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'data.table::first' by 'dplyr::first' when</span></span>
<span><span class="co">## loading 'stream2'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'data.table::between' by 'dplyr::between'</span></span>
<span><span class="co">## when loading 'stream2'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'dplyr::union' by 'igraph::union' when</span></span>
<span><span class="co">## loading 'stream2'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'dplyr::as_data_frame' by</span></span>
<span><span class="co">## 'igraph::as_data_frame' when loading 'stream2'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'dplyr::groups' by 'igraph::groups' when</span></span>
<span><span class="co">## loading 'stream2'</span></span></code></pre>
<pre><code><span><span class="co">## cistrome_tools.R</span></span></code></pre>
<pre><code><span><span class="co">## epigenome_tools.R</span></span></code></pre>
<pre><code><span><span class="co">## multiome_tools.R</span></span></code></pre>
<pre><code><span><span class="co">## protein.R</span></span></code></pre>
<pre><code><span><span class="co">## reguome_tools.R</span></span></code></pre>
<pre><code><span><span class="co">## run_stream.R</span></span></code></pre>
<pre><code><span><span class="co">## stream2.R</span></span></code></pre>
<pre><code><span><span class="co">## transcriptome_tools.R</span></span></code></pre>
<pre><code><span><span class="co">## utilities.R</span></span></code></pre>
<pre><code><span><span class="co">## visual.R</span></span></code></pre>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## [[1]]$value</span></span>
<span><span class="co">## function (object, peak.assay = "ATAC", expression.assay = "RNA", </span></span>
<span><span class="co">##     peak.slot = "counts", expression.slot = "data", method = "pearson", </span></span>
<span><span class="co">##     gene.coords = NULL, distance = 5e+05, min.distance = NULL, </span></span>
<span><span class="co">##     min.cells = 10, genes.use = NULL, n_sample = 100, pvalue_cutoff = 0.05, </span></span>
<span><span class="co">##     score_cutoff = 0.05, gene.id = FALSE, verbose = TRUE) </span></span>
<span><span class="co">## {</span></span>
<span><span class="co">##     if (!requireNamespace(package = "qlcMatrix", quietly = TRUE)) {</span></span>
<span><span class="co">##         stop("Please install qlcMatrix: install.packages('qlcMatrix')")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if (!inherits(x = object[[peak.assay]], what = "ChromatinAssay")) {</span></span>
<span><span class="co">##         stop("The requested assay is not a ChromatinAssay")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if (!is.null(x = min.distance)) {</span></span>
<span><span class="co">##         if (!is.numeric(x = min.distance)) {</span></span>
<span><span class="co">##             stop("min.distance should be a numeric value")</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         if (min.distance &lt; 0) {</span></span>
<span><span class="co">##             warning("Requested a negative min.distance value, setting min.distance to zero")</span></span>
<span><span class="co">##             min.distance &lt;- NULL</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         else if (min.distance == 0) {</span></span>
<span><span class="co">##             min.distance &lt;- NULL</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     features.match &lt;- c("GC.percent", "count", "sequence.length")</span></span>
<span><span class="co">##     if (method == "pearson") {</span></span>
<span><span class="co">##         cor_method &lt;- qlcMatrix::corSparse</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     else if (method == "spearman") {</span></span>
<span><span class="co">##         cor_method &lt;- SparseSpearmanCor</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     else {</span></span>
<span><span class="co">##         stop("method can be one of 'pearson' or 'spearman'.")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if (is.null(x = gene.coords)) {</span></span>
<span><span class="co">##         annot &lt;- Signac::Annotation(object = object[[peak.assay]])</span></span>
<span><span class="co">##         if (is.null(x = annot)) {</span></span>
<span><span class="co">##             stop("Gene annotations not found")</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         gene.coords &lt;- CollapseToLongestTranscript(ranges = annot)</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     meta.features &lt;- Seurat::GetAssayData(object = object, assay = peak.assay, </span></span>
<span><span class="co">##         slot = "meta.features")</span></span>
<span><span class="co">##     if (!(all(c("GC.percent", "sequence.length") %in% colnames(x = meta.features)))) {</span></span>
<span><span class="co">##         stop("DNA sequence information for each peak has not been computed.\n", </span></span>
<span><span class="co">##             "Run RegionsStats before calling this function.")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if (!("count" %in% colnames(x = meta.features))) {</span></span>
<span><span class="co">##         data.use &lt;- Seurat::GetAssayData(object = object[[peak.assay]], </span></span>
<span><span class="co">##             slot = "counts")</span></span>
<span><span class="co">##         hvf.info &lt;- Signac::FindTopFeatures(object = data.use, </span></span>
<span><span class="co">##             verbose = FALSE)</span></span>
<span><span class="co">##         hvf.info &lt;- hvf.info[rownames(meta.features), , drop = FALSE]</span></span>
<span><span class="co">##         meta.features &lt;- cbind(meta.features, hvf.info)</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     peak.data &lt;- Seurat::GetAssayData(object = object, assay = peak.assay, </span></span>
<span><span class="co">##         slot = peak.slot)</span></span>
<span><span class="co">##     expression.data &lt;- Seurat::GetAssayData(object = object, </span></span>
<span><span class="co">##         assay = expression.assay, slot = expression.slot)</span></span>
<span><span class="co">##     peakcounts &lt;- rowSums(x = peak.data &gt; 0)</span></span>
<span><span class="co">##     genecounts &lt;- rowSums(x = expression.data &gt; 0)</span></span>
<span><span class="co">##     peaks.keep &lt;- peakcounts &gt; min.cells</span></span>
<span><span class="co">##     genes.keep &lt;- genecounts &gt; min.cells</span></span>
<span><span class="co">##     peak.data &lt;- peak.data[peaks.keep, , drop = FALSE]</span></span>
<span><span class="co">##     if (!is.null(x = genes.use)) {</span></span>
<span><span class="co">##         genes.keep &lt;- intersect(x = names(x = genes.keep[genes.keep]), </span></span>
<span><span class="co">##             y = genes.use)</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     expression.data &lt;- expression.data[genes.keep, , drop = FALSE]</span></span>
<span><span class="co">##     if (verbose) {</span></span>
<span><span class="co">##         message("Testing ", nrow(x = expression.data), " genes and ", </span></span>
<span><span class="co">##             sum(peaks.keep), " peaks")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     genes &lt;- rownames(x = expression.data)</span></span>
<span><span class="co">##     if (gene.id) {</span></span>
<span><span class="co">##         gene.coords.use &lt;- gene.coords[gene.coords$gene_id %in% </span></span>
<span><span class="co">##             genes, ]</span></span>
<span><span class="co">##         gene.coords.use$gene_name &lt;- gene.coords.use$gene_id</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     else {</span></span>
<span><span class="co">##         gene.coords.use &lt;- gene.coords[gene.coords$gene_name %in% </span></span>
<span><span class="co">##             genes, ]</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if (length(x = gene.coords.use) == 0) {</span></span>
<span><span class="co">##         stop("Could not find gene coordinates for requested genes")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if (length(x = gene.coords.use) &lt; nrow(x = expression.data)) {</span></span>
<span><span class="co">##         message("Found gene coordinates for ", length(x = gene.coords.use), </span></span>
<span><span class="co">##             " genes")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     peaks &lt;- Signac::StringToGRanges(rownames(object[[peak.assay]]))</span></span>
<span><span class="co">##     peaks &lt;- peaks[peaks.keep]</span></span>
<span><span class="co">##     peak_distance_matrix &lt;- DistanceToTSS(peaks = peaks, genes = gene.coords.use, </span></span>
<span><span class="co">##         distance = distance)</span></span>
<span><span class="co">##     if (!is.null(x = min.distance)) {</span></span>
<span><span class="co">##         peak_distance_matrix_min &lt;- DistanceToTSS(peaks = peaks, </span></span>
<span><span class="co">##             genes = gene.coords.use, distance = min.distance)</span></span>
<span><span class="co">##         peak_distance_matrix &lt;- peak_distance_matrix - peak_distance_matrix_min</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if (sum(peak_distance_matrix) == 0) {</span></span>
<span><span class="co">##         stop("No peaks fall within distance threshold\n", "Have you set the proper genome and seqlevelsStyle for ", </span></span>
<span><span class="co">##             peak.assay, " assay?")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     genes.use &lt;- colnames(x = peak_distance_matrix)</span></span>
<span><span class="co">##     all.peaks &lt;- rownames(x = peak.data)</span></span>
<span><span class="co">##     peak.data &lt;- quiet(SeuratDisk::Transpose(peak.data))</span></span>
<span><span class="co">##     coef.vec &lt;- c()</span></span>
<span><span class="co">##     gene.vec &lt;- c()</span></span>
<span><span class="co">##     zscore.vec &lt;- c()</span></span>
<span><span class="co">##     if (TRUE) {</span></span>
<span><span class="co">##         mylapply &lt;- pbmcapply::pbmclapply</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     else {</span></span>
<span><span class="co">##         mylapply &lt;- ifelse(test = verbose, yes = pblapply, no = lapply)</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     res &lt;- mylapply(X = seq_along(along.with = genes.use), mc.cores = max(parallel::detectCores()/2, </span></span>
<span><span class="co">##         1), FUN = function(i) {</span></span>
<span><span class="co">##         peak.use &lt;- as.logical(x = peak_distance_matrix[, genes.use[[i]]])</span></span>
<span><span class="co">##         gene.expression &lt;- t(x = expression.data[genes.use[[i]], </span></span>
<span><span class="co">##             , drop = FALSE])</span></span>
<span><span class="co">##         gene.chrom &lt;- as.character(x = seqnames(x = gene.coords.use[i]))</span></span>
<span><span class="co">##         if (sum(peak.use) &lt; 2) {</span></span>
<span><span class="co">##             return(list(gene = NULL, coef = NULL, zscore = NULL))</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         else {</span></span>
<span><span class="co">##             peak.access &lt;- peak.data[, peak.use, drop = FALSE]</span></span>
<span><span class="co">##             coef.result &lt;- cor_method(X = peak.access, Y = gene.expression)</span></span>
<span><span class="co">##             rownames(x = coef.result) &lt;- colnames(x = peak.access)</span></span>
<span><span class="co">##             coef.result &lt;- coef.result[abs(x = coef.result) &gt; </span></span>
<span><span class="co">##                 score_cutoff, , drop = FALSE]</span></span>
<span><span class="co">##             if (nrow(x = coef.result) == 0) {</span></span>
<span><span class="co">##                 return(list(gene = NULL, coef = NULL, zscore = NULL))</span></span>
<span><span class="co">##             }</span></span>
<span><span class="co">##             else {</span></span>
<span><span class="co">##                 peaks.test &lt;- rownames(x = coef.result)</span></span>
<span><span class="co">##                 trans.peaks &lt;- all.peaks[!grepl(pattern = paste0("^", </span></span>
<span><span class="co">##                   gene.chrom), x = all.peaks)]</span></span>
<span><span class="co">##                 meta.use &lt;- meta.features[trans.peaks, ]</span></span>
<span><span class="co">##                 pk.use &lt;- meta.features[peaks.test, ]</span></span>
<span><span class="co">##                 bg.peaks &lt;- lapply(X = seq_len(length.out = nrow(x = pk.use)), </span></span>
<span><span class="co">##                   FUN = function(x) {</span></span>
<span><span class="co">##                     Signac::MatchRegionStats(meta.feature = meta.use, </span></span>
<span><span class="co">##                       query.feature = pk.use[x, , drop = FALSE], </span></span>
<span><span class="co">##                       features.match = features.match, n = n_sample, </span></span>
<span><span class="co">##                       verbose = FALSE)</span></span>
<span><span class="co">##                   })</span></span>
<span><span class="co">##                 bg.access &lt;- peak.data[, unlist(x = bg.peaks), </span></span>
<span><span class="co">##                   drop = FALSE]</span></span>
<span><span class="co">##                 bg.coef &lt;- cor_method(X = bg.access, Y = gene.expression)</span></span>
<span><span class="co">##                 rownames(bg.coef) &lt;- colnames(bg.access)</span></span>
<span><span class="co">##                 zscores &lt;- vector(mode = "numeric", length = length(x = peaks.test))</span></span>
<span><span class="co">##                 for (j in seq_along(along.with = peaks.test)) {</span></span>
<span><span class="co">##                   coef.use &lt;- bg.coef[(((j - 1) * n_sample) + </span></span>
<span><span class="co">##                     1):(j * n_sample), ]</span></span>
<span><span class="co">##                   z &lt;- (coef.result[j] - mean(x = coef.use))/sd(x = coef.use)</span></span>
<span><span class="co">##                   zscores[[j]] &lt;- z</span></span>
<span><span class="co">##                 }</span></span>
<span><span class="co">##                 names(x = coef.result) &lt;- peaks.test</span></span>
<span><span class="co">##                 names(x = zscores) &lt;- peaks.test</span></span>
<span><span class="co">##                 zscore.vec &lt;- c(zscore.vec, zscores)</span></span>
<span><span class="co">##                 gene.vec &lt;- c(gene.vec, rep(i, length(x = coef.result)))</span></span>
<span><span class="co">##                 coef.vec &lt;- c(coef.vec, coef.result)</span></span>
<span><span class="co">##             }</span></span>
<span><span class="co">##             gc(verbose = FALSE)</span></span>
<span><span class="co">##             pval.vec &lt;- pnorm(q = -abs(x = zscore.vec))</span></span>
<span><span class="co">##             links.keep &lt;- pval.vec &lt; pvalue_cutoff</span></span>
<span><span class="co">##             if (sum(x = links.keep) == 0) {</span></span>
<span><span class="co">##                 return(list(gene = NULL, coef = NULL, zscore = NULL))</span></span>
<span><span class="co">##             }</span></span>
<span><span class="co">##             else {</span></span>
<span><span class="co">##                 gene.vec &lt;- gene.vec[links.keep]</span></span>
<span><span class="co">##                 coef.vec &lt;- coef.vec[links.keep]</span></span>
<span><span class="co">##                 zscore.vec &lt;- zscore.vec[links.keep]</span></span>
<span><span class="co">##                 return(list(gene = gene.vec, coef = coef.vec, </span></span>
<span><span class="co">##                   zscore = zscore.vec))</span></span>
<span><span class="co">##             }</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##     })</span></span>
<span><span class="co">##     gene.vec &lt;- do.call(what = c, args = lapply(X = res, FUN = `[[`, </span></span>
<span><span class="co">##         1))</span></span>
<span><span class="co">##     coef.vec &lt;- do.call(what = c, args = lapply(X = res, FUN = `[[`, </span></span>
<span><span class="co">##         2))</span></span>
<span><span class="co">##     zscore.vec &lt;- do.call(what = c, args = lapply(X = res, FUN = `[[`, </span></span>
<span><span class="co">##         3))</span></span>
<span><span class="co">##     if (length(x = coef.vec) == 0) {</span></span>
<span><span class="co">##         if (verbose) {</span></span>
<span><span class="co">##             message("No significant links found!")</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         return(object)</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     peak.key &lt;- seq_along(along.with = unique(x = names(x = coef.vec)))</span></span>
<span><span class="co">##     names(x = peak.key) &lt;- unique(x = names(x = coef.vec))</span></span>
<span><span class="co">##     coef.matrix &lt;- Matrix::sparseMatrix(i = gene.vec, j = peak.key[names(x = coef.vec)], </span></span>
<span><span class="co">##         x = coef.vec, dims = c(length(x = genes.use), max(peak.key)))</span></span>
<span><span class="co">##     rownames(x = coef.matrix) &lt;- genes.use</span></span>
<span><span class="co">##     colnames(x = coef.matrix) &lt;- names(x = peak.key)</span></span>
<span><span class="co">##     links &lt;- invisible(LinksToGRanges(linkmat = coef.matrix, </span></span>
<span><span class="co">##         gene.coords = gene.coords.use))</span></span>
<span><span class="co">##     z.matrix &lt;- Matrix::sparseMatrix(i = gene.vec, j = peak.key[names(x = zscore.vec)], </span></span>
<span><span class="co">##         x = zscore.vec, dims = c(length(x = genes.use), max(peak.key)))</span></span>
<span><span class="co">##     rownames(x = z.matrix) &lt;- genes.use</span></span>
<span><span class="co">##     colnames(x = z.matrix) &lt;- names(x = peak.key)</span></span>
<span><span class="co">##     z.lnk &lt;- LinksToGRanges(linkmat = z.matrix, gene.coords = gene.coords.use)</span></span>
<span><span class="co">##     links$zscore &lt;- z.lnk$score</span></span>
<span><span class="co">##     links$pvalue &lt;- pnorm(q = -abs(x = links$zscore))</span></span>
<span><span class="co">##     links &lt;- links[links$pvalue &lt; pvalue_cutoff]</span></span>
<span><span class="co">##     Signac::Links(object = object[[peak.assay]]) &lt;- links</span></span>
<span><span class="co">##     return(object)</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## [[2]]$value</span></span>
<span><span class="co">## function (lst1, lst2) </span></span>
<span><span class="co">## {</span></span>
<span><span class="co">##     overlaps &lt;- GenomicRanges::findOverlaps(query = lst1, subject = lst2, </span></span>
<span><span class="co">##         type = "any", select = "all")</span></span>
<span><span class="co">##     Matrix::sparseMatrix(i = queryHits(x = overlaps), j = subjectHits(x = overlaps), </span></span>
<span><span class="co">##         x = 1, dims = c(length(x = lst1), length(x = lst2)))</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span>
<span><span class="co">## [[3]]$value</span></span>
<span><span class="co">## function (rna_counts = NULL, atac_counts = NULL, org = "hg38", </span></span>
<span><span class="co">##     frag.file = NULL, sep = c("-", "-")) </span></span>
<span><span class="co">## {</span></span>
<span><span class="co">##     message("Loaded scRNA-seq and scATAC-seq matrices of sizes: ", </span></span>
<span><span class="co">##         nrow(rna_counts), " x ", ncol(rna_counts), " and ", nrow(atac_counts), </span></span>
<span><span class="co">##         " x ", ncol(atac_counts), ".")</span></span>
<span><span class="co">##     pbmc &lt;- Seurat::CreateSeuratObject(counts = rna_counts)</span></span>
<span><span class="co">##     pbmc[["percent.mt"]] &lt;- Seurat::PercentageFeatureSet(pbmc, </span></span>
<span><span class="co">##         pattern = "^MT-")</span></span>
<span><span class="co">##     grange.counts &lt;- Signac::StringToGRanges(rownames(atac_counts), </span></span>
<span><span class="co">##         sep = sep)</span></span>
<span><span class="co">##     grange.use &lt;- BSgenome::seqnames(grange.counts) %in% GenomeInfoDb::standardChromosomes(grange.counts)</span></span>
<span><span class="co">##     atac_counts &lt;- atac_counts[as.vector(grange.use), ]</span></span>
<span><span class="co">##     annotations &lt;- Signac::GetGRangesFromEnsDb(ensdb = org_to_DB(org = org))</span></span>
<span><span class="co">##     ensembldb::seqlevelsStyle(annotations) &lt;- "UCSC"</span></span>
<span><span class="co">##     GenomeInfoDb::genome(annotations) &lt;- org</span></span>
<span><span class="co">##     if (!is.null(frag.file)) {</span></span>
<span><span class="co">##         chrom_assay &lt;- Signac::CreateChromatinAssay(counts = atac_counts, </span></span>
<span><span class="co">##             sep = sep, genome = org, fragments = frag.file, min.cells = 10, </span></span>
<span><span class="co">##             annotation = annotations)</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     else {</span></span>
<span><span class="co">##         chrom_assay &lt;- Signac::CreateChromatinAssay(counts = atac_counts, </span></span>
<span><span class="co">##             sep = sep, genome = org, min.cells = 10, annotation = annotations)</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     pbmc[["ATAC"]] &lt;- chrom_assay</span></span>
<span><span class="co">##     pbmc</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span>
<span><span class="co">## [[4]]$value</span></span>
<span><span class="co">## function (key.TFs = NULL, db = "BioGrid", org = "human") </span></span>
<span><span class="co">## {</span></span>
<span><span class="co">##     message("Identifying proteins interacting with the ", length(key.TFs), </span></span>
<span><span class="co">##         " TFs provided by user.")</span></span>
<span><span class="co">##     if (db == "STRING") {</span></span>
<span><span class="co">##         message("Loaded the STRING database.")</span></span>
<span><span class="co">##         if (org == "mouse") {</span></span>
<span><span class="co">##             species &lt;- 10090</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         else {</span></span>
<span><span class="co">##             species &lt;- 9606</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         tmp.db &lt;- quiet(STRINGdb::STRINGdb$new(species = species))</span></span>
<span><span class="co">##         tmp.graph &lt;- quiet(tmp.db$get_graph())</span></span>
<span><span class="co">##         ensembl.lst &lt;- igraph::as_adj_list(tmp.graph, mode = "all") %&gt;% </span></span>
<span><span class="co">##             lapply(., function(x) {</span></span>
<span><span class="co">##                 strsplit(igraph::as_ids(seq = x), split = "\\.") %&gt;% </span></span>
<span><span class="co">##                   sapply(., "[[", 2)</span></span>
<span><span class="co">##             })</span></span>
<span><span class="co">##         names(ensembl.lst) &lt;- strsplit(names(ensembl.lst), split = "\\.") %&gt;% </span></span>
<span><span class="co">##             sapply(., "[[", 2)</span></span>
<span><span class="co">##         if (org == "mouse") {</span></span>
<span><span class="co">##             mart &lt;- biomaRt::useMart(host = "grcm39.ensembl.org", </span></span>
<span><span class="co">##                 biomart = "ENSEMBL_MART_ENSEMBL", dataset = "mmusculus_gene_ensembl")</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         else {</span></span>
<span><span class="co">##             mart &lt;- biomaRt::useMart(host = "grch38.ensembl.org", </span></span>
<span><span class="co">##                 biomart = "ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_symbol")</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         mart_results &lt;- biomaRt::getBM(attributes = c("ensembl_gene_id", </span></span>
<span><span class="co">##             "ensembl_peptide_id"), filters = "ensembl_peptide_id", </span></span>
<span><span class="co">##             values = protein_ids, mart = mart)</span></span>
<span><span class="co">##         ix &lt;- match(protein_ids, mart_results$ensembl_peptide_id)</span></span>
<span><span class="co">##         ix &lt;- ix[!is.na(ix)]</span></span>
<span><span class="co">##         relation.lst &lt;- pbmcapply::pbmclapply(ensembl.lst, mc.cores = parallel::detectCores(), </span></span>
<span><span class="co">##             function(protein_ids) {</span></span>
<span><span class="co">##                 newnames &lt;- protein_ids</span></span>
<span><span class="co">##                 newnames[match(mart_results[ix, "ensembl_peptide_id"], </span></span>
<span><span class="co">##                   newnames)] &lt;- mart_results[ix, "ensembl_gene_symbol"]</span></span>
<span><span class="co">##                 newnames</span></span>
<span><span class="co">##             })</span></span>
<span><span class="co">##         names(relation.lst)[match(mart_results[ix, "ensembl_peptide_id"], </span></span>
<span><span class="co">##             names(relation.lst))] &lt;- mart_results[ix, "ensembl_gene_symbol"]</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     else {</span></span>
<span><span class="co">##         message("Loaded the BioGrid database.")</span></span>
<span><span class="co">##         if (org == "mouse") {</span></span>
<span><span class="co">##             data(MouseBioGRIDInteractionOfficial)</span></span>
<span><span class="co">##             tmp.lst &lt;- MouseBioGRIDInteractionOfficial</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         else {</span></span>
<span><span class="co">##             data(HumanBioGRIDInteractionOfficial)</span></span>
<span><span class="co">##             tmp.lst &lt;- HumanBioGRIDInteractionOfficial</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         relation.lst &lt;- lapply(tmp.lst, "[[", "interactors")</span></span>
<span><span class="co">##         names(relation.lst) &lt;- sapply(tmp.lst, "[[", "name")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if (org == "mouse") {</span></span>
<span><span class="co">##         taxid &lt;- 10090</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     else {</span></span>
<span><span class="co">##         taxid &lt;- 9606</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     message("Loaded ", length(relation.lst), " protein relations from ", </span></span>
<span><span class="co">##         db, " database of ", org, ".")</span></span>
<span><span class="co">##     TFs &lt;- c(unique(unlist(relation.lst[intersect(names(relation.lst), </span></span>
<span><span class="co">##         key.TFs)])), key.TFs)</span></span>
<span><span class="co">##     message("Identified ", length(TFs), " TFs in total.")</span></span>
<span><span class="co">##     return(TFs)</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[4]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[5]]</span></span>
<span><span class="co">## [[5]]$value</span></span>
<span><span class="co">## function (pfm = NULL, peaks = NULL, org.gs = BSgenome.Hsapiens.UCSC.hg38, </span></span>
<span><span class="co">##     candidate.TFs = NULL) </span></span>
<span><span class="co">## {</span></span>
<span><span class="co">##     motif.ix &lt;- motifmatchr::motifMatches(motifmatchr::matchMotifs(pwms = pfm, </span></span>
<span><span class="co">##         subject = Signac::StringToGRanges(peaks), genome = org.gs)) %&gt;% </span></span>
<span><span class="co">##         summary</span></span>
<span><span class="co">##     TF.lst &lt;- sapply(pfm, function(x) {</span></span>
<span><span class="co">##         x@name</span></span>
<span><span class="co">##     })</span></span>
<span><span class="co">##     names(TF.lst) &lt;- NULL</span></span>
<span><span class="co">##     peak.TF.pairs &lt;- data.frame(peaks[motif.ix[, 1]], TF.lst[motif.ix[, </span></span>
<span><span class="co">##         2]])</span></span>
<span><span class="co">##     colnames(peak.TF.pairs) &lt;- c("peak", "TF")</span></span>
<span><span class="co">##     if (is.not.null(candidate.TFs)) {</span></span>
<span><span class="co">##         peak.TF.pairs &lt;- peak.TF.pairs[peak.TF.pairs$TF %in% </span></span>
<span><span class="co">##             candidate.TFs, , drop = FALSE]</span></span>
<span><span class="co">##         message("Identify putative TF binding sites via motif enrichment analysis for the TF list", </span></span>
<span><span class="co">##             " provided by the user ...")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     peak.TFs &lt;- split(peak.TF.pairs, peak.TF.pairs$peak) %&gt;% </span></span>
<span><span class="co">##         sapply(., "[[", "TF")</span></span>
<span><span class="co">##     TF.peaks &lt;- split(peak.TF.pairs, peak.TF.pairs$TF) %&gt;% sapply(., </span></span>
<span><span class="co">##         "[[", "peak")</span></span>
<span><span class="co">##     return(list(puta.bound.TFs = peak.TFs, puta.binding.peaks = TF.peaks))</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[5]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[6]]</span></span>
<span><span class="co">## [[6]]$value</span></span>
<span><span class="co">## function (regs = NULL, dbs = c("GO", "KEGG"), org = "human") </span></span>
<span><span class="co">## {</span></span>
<span><span class="co">##     genes.ll &lt;- sapply(regs, "[[", "genes")</span></span>
<span><span class="co">##     databases &lt;- c()</span></span>
<span><span class="co">##     if ("GO" %in% dbs) {</span></span>
<span><span class="co">##         message("Loading GO databases ...")</span></span>
<span><span class="co">##         databases &lt;- c("GO_Molecular_Function_2018", "GO_Cellular_Component_2018", </span></span>
<span><span class="co">##             "GO_Biological_Process_2018")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if ("KEGG" %in% dbs) {</span></span>
<span><span class="co">##         if (org == "mouse") {</span></span>
<span><span class="co">##             databases &lt;- c(databases, "KEGG_2019_Mouse")</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         else {</span></span>
<span><span class="co">##             databases &lt;- c(databases, "KEGG_2019_Human")</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         message("Loading KEGG database of ", org, " ...")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     message("Running enrichment analyses for ", length(genes.ll), </span></span>
<span><span class="co">##         " gene lists ...")</span></span>
<span><span class="co">##     quiet(require(enrichR))</span></span>
<span><span class="co">##     enriched &lt;- pbapply::pblapply(genes.ll, function(x) {</span></span>
<span><span class="co">##         enrichr(genes = x, databases = databases)</span></span>
<span><span class="co">##     })</span></span>
<span><span class="co">##     names(enriched) &lt;- seq_along(genes.ll)</span></span>
<span><span class="co">##     message("Merging enrichment analyses results ...")</span></span>
<span><span class="co">##     processed &lt;- pbmcapply::pbmclapply(databases, mc.cores = parallel::detectCores(), </span></span>
<span><span class="co">##         function(x) {</span></span>
<span><span class="co">##             message("Preparing enrichment analysis results for ", </span></span>
<span><span class="co">##                 x, " ...")</span></span>
<span><span class="co">##             x.prosessed &lt;- Reduce("rbind", lapply(seq_along(enriched), </span></span>
<span><span class="co">##                 function(i) {</span></span>
<span><span class="co">##                   cbind(Id = rep(names(enriched)[i], nrow(enriched[[i]][[x]])), </span></span>
<span><span class="co">##                     enriched[[i]][[x]])</span></span>
<span><span class="co">##                 }))</span></span>
<span><span class="co">##         })</span></span>
<span><span class="co">##     names(processed) &lt;- databases</span></span>
<span><span class="co">##     message("There are ", paste(sapply(processed, nrow), collapse = ", "), </span></span>
<span><span class="co">##         " terms/pathways for ", paste(names(processed), collapse = ", "), </span></span>
<span><span class="co">##         ".")</span></span>
<span><span class="co">##     return(processed)</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[6]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[7]]</span></span>
<span><span class="co">## [[7]]$value</span></span>
<span><span class="co">## [1] "_PACKAGE"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[7]]$visible</span></span>
<span><span class="co">## [1] TRUE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[8]]</span></span>
<span><span class="co">## [[8]]$value</span></span>
<span><span class="co">## function (obj) </span></span>
<span><span class="co">## {</span></span>
<span><span class="co">##     ltmg.df &lt;- as.data.frame(obj@assays$RNA@data)</span></span>
<span><span class="co">##     ltmg.obj &lt;- IRISFGM::ProcessData(IRISFGM::CreateIRISFGMObject(ltmg.df), </span></span>
<span><span class="co">##         normalization = "cpm", IsImputation = F)</span></span>
<span><span class="co">##     IRISFGM::RunLTMG(ltmg.obj, Gene_use = "all")</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[8]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[9]]</span></span>
<span><span class="co">## [[9]]$value</span></span>
<span><span class="co">## function (x) </span></span>
<span><span class="co">## !is.na(x)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[9]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[10]]</span></span>
<span><span class="co">## [[10]]$value</span></span>
<span><span class="co">## function (object = NULL, region = NULL, features = NULL, links = FALSE, </span></span>
<span><span class="co">##     ranges.group.by = "seurat_cluster", peaks = FALSE, bigwig = NULL, </span></span>
<span><span class="co">##     region.highlight = NULL, idents = NULL, extend.upstream = 0, </span></span>
<span><span class="co">##     extend.downstream = 0, colors = NULL, size = 10) </span></span>
<span><span class="co">## {</span></span>
<span><span class="co">##     Seurat::DefaultAssay(object) &lt;- peak.assay</span></span>
<span><span class="co">##     Seurat::Idents(object) &lt;- setNames(object@meta.data[, ranges.group.by], </span></span>
<span><span class="co">##         colnames(object))</span></span>
<span><span class="co">##     p &lt;- suppressWarnings(Signac::CoveragePlot(object = object, </span></span>
<span><span class="co">##         region = region, features = features, ranges.group.by = ranges.group.by, </span></span>
<span><span class="co">##         peaks = peaks, links = links, idents = idents, bigwig = bigwig, </span></span>
<span><span class="co">##         region.highlight = region.highlight))</span></span>
<span><span class="co">##     if (is.not.null(colors)) {</span></span>
<span><span class="co">##         p &lt;- p &amp; scale_fill_manual(values = colors)</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     return(p)</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[10]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span></code></pre>
<div class="section level2">
<h2 id="data-simulation">Data simulation<a class="anchor" aria-label="anchor" href="#data-simulation"></a>
</h2>
<p>Load a Seurat object that comprises jointly profiled scRNA-seq and
scATAC-seq assays derived from human PBMCs and produce simulated data.
Based on this object, construct a simulated Seurat object encompassing
1,000 genes, 3,000 peaks, and 1,000 cells. This object features five
eRegulons, with each eRegulon linked to roughly five target genes and
approximately two enhancers. These eRegulons are active in 100
cells.</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html" class="external-link">qread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">work.dir</span>, <span class="st">"tutorial_pbmc.qsave"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pbmc</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 6193 features across 1000 samples within 3 assays </span></span>
<span><span class="co">## Active assay: peaks (3991 features, 3991 variable features)</span></span>
<span><span class="co">##  2 other assays present: RNA, SCT</span></span>
<span><span class="co">##  3 dimensional reductions calculated: pca, lsi, umap</span></span></code></pre>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simul</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/create_rna_atac.html">create_rna_atac</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">pbmc</span>, ntfs <span class="op">=</span> <span class="fl">5</span>, ngenes <span class="op">=</span> <span class="fl">100</span>,</span>
<span>                         ncells <span class="op">=</span> <span class="fl">100</span>, all.genes <span class="op">=</span> <span class="fl">1000</span>, all.enhs <span class="op">=</span> <span class="fl">3000</span>, all.cells <span class="op">=</span> <span class="fl">1000</span>,</span>
<span>                         org <span class="op">=</span> <span class="st">"hg38"</span>, atac.assay <span class="op">=</span> <span class="st">"peaks"</span>, gene.links <span class="op">=</span> <span class="fl">2</span></span>
<span>                         <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading TF binding sites from JASPAR 2022 ...</span></span></code></pre>
<pre><code><span><span class="co">## There are 203 TFs from the JASPAR 2022 database.</span></span>
<span><span class="co">## 5 eRegulons regulated by different TFs will be generated.</span></span>
<span><span class="co">## The Seurat object contains 1101 genes, 3991 enhancers, and 1000 cells.</span></span></code></pre>
<pre><code><span><span class="co">## There are 197 candidate TFs curated by JASPAR 2022</span></span></code></pre>
<pre><code><span><span class="co">## Loaded 1207 TFs collected in DoRothEA database.</span></span></code></pre>
<pre><code><span><span class="co">## There are 500 common selected TFs between JASPAR and DoRothEA.</span></span></code></pre>
<pre><code><span><span class="co">## Generated regulons composed of 1-142 genes.</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: EnsDb.Hsapiens.v75</span></span></code></pre>
<pre><code><span><span class="co">## Finding nearby genes for each peak within 5e+05 bp ...</span></span></code></pre>
<pre><code><span><span class="co">## Identified 2849 enhancers within 5e+05 from the selected genes.</span></span></code></pre>
<pre><code><span><span class="co">## Identified the binding enhancers of 65 TFs, ranging from 47 to 322803 bound enhancers</span></span></code></pre>
<pre><code><span><span class="co">## Generated 65 eRegulons composed of 0-28 enhancer-gene relations.</span></span></code></pre>
<pre><code><span><span class="co">## Retained 5 eRegulons composed of 13-28 enhancer-gene relations.</span></span></code></pre>
<pre><code><span><span class="co">## Began building hybrid biclusters (HBCs) based on simulated eRegulons ...</span></span></code></pre>
<pre><code>## 
  |                                                        
  |                                                  |   0%
  |                                                        
  |==========                                        |  20%
  |                                                        
  |====================                              |  40%
  |                                                        
  |==============================                    |  60%
  |                                                        
  |========================================          |  80%
  |                                                        
  |==================================================| 100%</code></pre>
<pre><code><span><span class="co">## Simulated hybrid biclusters (HBCs) cover 63 genes, 83 enhancers, and 431 cells.</span></span></code></pre>
<pre><code><span><span class="co">## We have 937 genes, 2917 enhancers, and 569 cells as background.</span></span></code></pre>
<pre><code><span><span class="co">## Generated simulated scRNA-seq and scATAC-seq matrices of sizes: 1000 x 1000 and 3000 x 1000, respectively.</span></span></code></pre>
<pre><code><span><span class="co">## Loaded scRNA-seq and scATAC-seq matrices of sizes: 1000 x 1000 and 3000 x 1000.</span></span></code></pre>
<pre><code><span><span class="co">## The organism version is hg38.</span></span></code></pre>
<pre><code><span><span class="co">## Loading the database EnsDb.Hsapiens.v86 for assembly hg38</span></span></code></pre>
<pre><code><span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span>
<span></span>
<span><span class="co">## Warning in .merge_two_Seqinfo_objects(x, y): The 2 combined objects have no sequence levels in common. (Use</span></span>
<span><span class="co">##   suppressWarnings() to suppress this warning.)</span></span></code></pre>
<pre><code><span><span class="co">## The simulated scRNA-seq and scATAC-seq data contains:</span></span>
<span><span class="co">## 1. Hybrid biclusters (HBCs) in nested list,</span></span>
<span><span class="co">## 2. Seurat object containing the scRNA-seq and scATAC-seq data.</span></span></code></pre>
<div class="sourceCode" id="cb105"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">simul</span><span class="op">$</span><span class="va">Seurat</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 2826 features across 1000 samples within 2 assays </span></span>
<span><span class="co">## Active assay: RNA (1000 features, 0 variable features)</span></span>
<span><span class="co">##  1 other assay present: ATAC</span></span></code></pre>
<div class="sourceCode" id="cb107"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">simul</span><span class="op">$</span><span class="va">HBCs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "TNFRSF1B" "ANK3"     "ATP8B4"   "STX11"    "YBX3"     "PLEKHG1"</span></span></code></pre>
<div class="sourceCode" id="cb109"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">simul</span><span class="op">$</span><span class="va">HBCs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">enhancers</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "chr1-12046827-12047882"   "chr10-62050806-62051384" </span></span>
<span><span class="co">## [3] "chr10-62120709-62121465"  "chr15-50107284-50108159" </span></span>
<span><span class="co">## [5] "chr6-144149667-144151675" "chr12-10540630-10541134"</span></span></code></pre>
<div class="sourceCode" id="cb111"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">simul</span><span class="op">$</span><span class="va">HBCs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">links</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 1 metadata column:</span></span>
<span><span class="co">##       seqnames              ranges strand |        gene</span></span>
<span><span class="co">##          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1   12046827-12047882      * |    TNFRSF1B</span></span>
<span><span class="co">##   [2]    chr10   62050806-62051384      * |        ANK3</span></span>
<span><span class="co">##   [3]    chr10   62120709-62121465      * |        ANK3</span></span>
<span><span class="co">##   [4]    chr15   50107284-50108159      * |      ATP8B4</span></span>
<span><span class="co">##   [5]     chr6 144149667-144151675      * |       STX11</span></span>
<span><span class="co">##   [6]    chr12   10540630-10541134      * |        YBX3</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb113"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">simul</span><span class="op">$</span><span class="va">HBCs</span>, <span class="st">"[["</span>, <span class="st">"TF"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "BATF3"  "KLF15"  "CEBPA"  "TCF12"  "ZSCAN4"</span></span></code></pre>
<div class="sourceCode" id="cb115"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check session information</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Red Hat Enterprise Linux</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /opt/intel/2021.3/mkl/2021.3.0/lib/intel64/libmkl_gf_lp64.so.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    parallel  stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] EnsDb.Hsapiens.v75_2.99.0  stream2_2.0.0             </span></span>
<span><span class="co">##  [3] Matrix_1.6-0               igraph_1.5.0              </span></span>
<span><span class="co">##  [5] ggpubr_0.6.0               ggplot2_3.4.2             </span></span>
<span><span class="co">##  [7] dplyr_1.1.2                EnsDb.Hsapiens.v86_2.99.0 </span></span>
<span><span class="co">##  [9] EnsDb.Mmusculus.v79_2.99.0 ensembldb_2.22.0          </span></span>
<span><span class="co">## [11] AnnotationFilter_1.22.0    GenomicFeatures_1.50.4    </span></span>
<span><span class="co">## [13] AnnotationDbi_1.60.2       Biobase_2.58.0            </span></span>
<span><span class="co">## [15] GenomicRanges_1.50.2       GenomeInfoDb_1.34.9       </span></span>
<span><span class="co">## [17] IRanges_2.32.0             S4Vectors_0.36.2          </span></span>
<span><span class="co">## [19] BiocGenerics_0.44.0        pbapply_1.7-2             </span></span>
<span><span class="co">## [21] pbmcapply_1.5.1            Signac_1.10.0             </span></span>
<span><span class="co">## [23] SeuratObject_4.1.3         Seurat_4.3.0.1            </span></span>
<span><span class="co">## [25] hdf5r_1.3.8                qs_0.25.5                 </span></span>
<span><span class="co">## [27] easypackages_0.1.0         knitr_1.43                </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] rappdirs_0.3.3              rtracklayer_1.58.0         </span></span>
<span><span class="co">##   [3] scattermore_1.2             ragg_1.2.5                 </span></span>
<span><span class="co">##   [5] tidyr_1.3.0                 bit64_4.0.5                </span></span>
<span><span class="co">##   [7] irlba_2.3.5.1               DelayedArray_0.24.0        </span></span>
<span><span class="co">##   [9] rpart_4.1.16                data.table_1.14.8          </span></span>
<span><span class="co">##  [11] KEGGREST_1.38.0             RCurl_1.98-1.12            </span></span>
<span><span class="co">##  [13] generics_0.1.3              terra_1.7-39               </span></span>
<span><span class="co">##  [15] callr_3.7.3                 cowplot_1.1.1              </span></span>
<span><span class="co">##  [17] usethis_2.2.2               RSQLite_2.3.1              </span></span>
<span><span class="co">##  [19] RApiSerialize_0.1.2         RANN_2.6.1                 </span></span>
<span><span class="co">##  [21] future_1.33.0               bit_4.0.5                  </span></span>
<span><span class="co">##  [23] spatstat.data_3.0-1         xml2_1.3.5                 </span></span>
<span><span class="co">##  [25] httpuv_1.6.11               SummarizedExperiment_1.28.0</span></span>
<span><span class="co">##  [27] assertthat_0.2.1            xfun_0.40                  </span></span>
<span><span class="co">##  [29] hms_1.1.3                   jquerylib_0.1.4            </span></span>
<span><span class="co">##  [31] evaluate_0.21               promises_1.2.1             </span></span>
<span><span class="co">##  [33] fansi_1.0.4                 restfulr_0.0.15            </span></span>
<span><span class="co">##  [35] progress_1.2.2              dbplyr_2.3.3               </span></span>
<span><span class="co">##  [37] DBI_1.1.3                   htmlwidgets_1.6.2          </span></span>
<span><span class="co">##  [39] spatstat.geom_3.2-2         purrr_1.0.2                </span></span>
<span><span class="co">##  [41] ellipsis_0.3.2              backports_1.4.1            </span></span>
<span><span class="co">##  [43] RcppParallel_5.1.7          biomaRt_2.54.1             </span></span>
<span><span class="co">##  [45] deldir_1.0-9                MatrixGenerics_1.10.0      </span></span>
<span><span class="co">##  [47] vctrs_0.6.3                 SingleCellExperiment_1.20.1</span></span>
<span><span class="co">##  [49] remotes_2.4.2.1             ROCR_1.0-11                </span></span>
<span><span class="co">##  [51] abind_1.4-5                 cachem_1.0.8               </span></span>
<span><span class="co">##  [53] withr_2.5.0                 BSgenome_1.66.3            </span></span>
<span><span class="co">##  [55] progressr_0.14.0            checkmate_2.2.0            </span></span>
<span><span class="co">##  [57] sctransform_0.3.5           GenomicAlignments_1.34.1   </span></span>
<span><span class="co">##  [59] prettyunits_1.1.1           goftest_1.2-3              </span></span>
<span><span class="co">##  [61] cluster_2.1.3               lazyeval_0.2.2             </span></span>
<span><span class="co">##  [63] crayon_1.5.2                spatstat.explore_3.2-1     </span></span>
<span><span class="co">##  [65] pkgconfig_2.0.3             nlme_3.1-157               </span></span>
<span><span class="co">##  [67] pkgload_1.3.2.1             ProtGenerics_1.30.0        </span></span>
<span><span class="co">##  [69] nnet_7.3-17                 devtools_2.4.5             </span></span>
<span><span class="co">##  [71] rlang_1.1.1                 globals_0.16.2             </span></span>
<span><span class="co">##  [73] lifecycle_1.0.3             miniUI_0.1.1.1             </span></span>
<span><span class="co">##  [75] filelock_1.0.2              BiocFileCache_2.6.1        </span></span>
<span><span class="co">##  [77] dichromat_2.0-0.1           rprojroot_2.0.3            </span></span>
<span><span class="co">##  [79] polyclip_1.10-4             matrixStats_1.0.0          </span></span>
<span><span class="co">##  [81] lmtest_0.9-40               carData_3.0-5              </span></span>
<span><span class="co">##  [83] boot_1.3-28                 zoo_1.8-12                 </span></span>
<span><span class="co">##  [85] base64enc_0.1-3             ggridges_0.5.4             </span></span>
<span><span class="co">##  [87] processx_3.8.2              png_0.1-8                  </span></span>
<span><span class="co">##  [89] viridisLite_0.4.2           rjson_0.2.21               </span></span>
<span><span class="co">##  [91] stringfish_0.15.8           bitops_1.0-7               </span></span>
<span><span class="co">##  [93] KernSmooth_2.23-20          Biostrings_2.66.0          </span></span>
<span><span class="co">##  [95] blob_1.2.4                  stringr_1.5.0              </span></span>
<span><span class="co">##  [97] rTRM_1.36.0                 parallelly_1.36.0          </span></span>
<span><span class="co">##  [99] spatstat.random_3.1-5       rstatix_0.7.2              </span></span>
<span><span class="co">## [101] ggsignif_0.6.4              scales_1.2.1               </span></span>
<span><span class="co">## [103] memoise_2.0.1               magrittr_2.0.3             </span></span>
<span><span class="co">## [105] plyr_1.8.8                  ica_1.0-3                  </span></span>
<span><span class="co">## [107] zlibbioc_1.44.0             compiler_4.2.1             </span></span>
<span><span class="co">## [109] BiocIO_1.8.0                RColorBrewer_1.1-3         </span></span>
<span><span class="co">## [111] lme4_1.1-34                 fitdistrplus_1.1-11        </span></span>
<span><span class="co">## [113] Rsamtools_2.14.0            cli_3.6.1                  </span></span>
<span><span class="co">## [115] XVector_0.38.0              urlchecker_1.0.1           </span></span>
<span><span class="co">## [117] listenv_0.9.0               patchwork_1.1.2            </span></span>
<span><span class="co">## [119] ps_1.7.5                    htmlTable_2.4.1            </span></span>
<span><span class="co">## [121] Formula_1.2-5               MASS_7.3-57                </span></span>
<span><span class="co">## [123] tidyselect_1.2.0            stringi_1.7.12             </span></span>
<span><span class="co">## [125] textshaping_0.3.6           yaml_2.3.7                 </span></span>
<span><span class="co">## [127] ggrepel_0.9.3               grid_4.2.1                 </span></span>
<span><span class="co">## [129] VariantAnnotation_1.44.1    bcellViper_1.34.0          </span></span>
<span><span class="co">## [131] sass_0.4.7                  fastmatch_1.1-3            </span></span>
<span><span class="co">## [133] tools_4.2.1                 future.apply_1.11.0        </span></span>
<span><span class="co">## [135] rstudioapi_0.15.0           foreign_0.8-82             </span></span>
<span><span class="co">## [137] gridExtra_2.3               Rtsne_0.16                 </span></span>
<span><span class="co">## [139] digest_0.6.33               shiny_1.7.4.1              </span></span>
<span><span class="co">## [141] monocle3_1.3.1              Rcpp_1.0.11                </span></span>
<span><span class="co">## [143] car_3.1-2                   broom_1.0.5                </span></span>
<span><span class="co">## [145] later_1.3.1                 RcppAnnoy_0.0.21           </span></span>
<span><span class="co">## [147] httr_1.4.6                  biovizBase_1.46.0          </span></span>
<span><span class="co">## [149] colorspace_2.1-0            XML_3.99-0.14              </span></span>
<span><span class="co">## [151] fs_1.6.3                    tensor_1.5                 </span></span>
<span><span class="co">## [153] reticulate_1.31             splines_4.2.1              </span></span>
<span><span class="co">## [155] uwot_0.1.16                 RcppRoll_0.3.0             </span></span>
<span><span class="co">## [157] spatstat.utils_3.0-3        pkgdown_2.0.7              </span></span>
<span><span class="co">## [159] sp_2.0-0                    plotly_4.10.2              </span></span>
<span><span class="co">## [161] sessioninfo_1.2.2           systemfonts_1.0.4          </span></span>
<span><span class="co">## [163] dorothea_1.10.0             xtable_1.8-4               </span></span>
<span><span class="co">## [165] nloptr_2.0.3                jsonlite_1.8.7             </span></span>
<span><span class="co">## [167] R6_2.5.1                    Hmisc_5.1-0                </span></span>
<span><span class="co">## [169] profvis_0.3.8               pillar_1.9.0               </span></span>
<span><span class="co">## [171] htmltools_0.5.6             mime_0.12                  </span></span>
<span><span class="co">## [173] minqa_1.2.5                 glue_1.6.2                 </span></span>
<span><span class="co">## [175] fastmap_1.1.1               BiocParallel_1.32.6        </span></span>
<span><span class="co">## [177] codetools_0.2-18            pkgbuild_1.4.2             </span></span>
<span><span class="co">## [179] utf8_1.2.3                  lattice_0.20-45            </span></span>
<span><span class="co">## [181] bslib_0.5.1                 spatstat.sparse_3.0-2      </span></span>
<span><span class="co">## [183] tibble_3.2.1                curl_5.0.1                 </span></span>
<span><span class="co">## [185] leiden_0.4.3                survival_3.3-1             </span></span>
<span><span class="co">## [187] rmarkdown_2.23              desc_1.4.2                 </span></span>
<span><span class="co">## [189] munsell_0.5.0               GenomeInfoDbData_1.2.9     </span></span>
<span><span class="co">## [191] reshape2_1.4.4              gtable_0.3.3</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yang Li, Anjun Ma, Yizhong Wang, Cankun Wang.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
