<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>STREAM: A comprehensive guide to inferring enhancer-driven gene regulatory networks from scRNA-seq and scATAC-seq data â€¢ stream</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="STREAM: A comprehensive guide to inferring enhancer-driven gene regulatory networks from scRNA-seq and scATAC-seq data">
<meta property="og:description" content="stream">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stream</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/install.html">Install</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/data-simulation.html">Simulated data analyses</a>
    </li>
    <li>
      <a href="../articles/eRegulon-inference.html">Biological data analyses</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/YangLi-Bio/stream2" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>STREAM: A comprehensive guide to inferring
enhancer-driven gene regulatory networks from scRNA-seq and scATAC-seq
data</h1>
            
            <h4 data-toc-skip class="date">Compiled: June 16, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/YangLi-Bio/stream2/blob/HEAD/vignettes/eRegulon-inference.Rmd" class="external-link"><code>vignettes/eRegulon-inference.Rmd</code></a></small>
      <div class="hidden name"><code>eRegulon-inference.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we will guide you through the process of inferring
enhancer regulons, known as eRegulons, and enhancer-driven gene
regulatory networks or eGRNs, utilizing the STREAM tool. Our
demonstration leverages a select subset of data from the publicly
accessible 10x Genomics Multiome dataset, specifically focusing on human
Peripheral Blood Mononuclear Cells (PBMCs). For those interested in the
comprehensive dataset, the raw data captured by the 10X Genomics
Multiome ATAC+GEX can be downloaded directly from <a href="https://www.10xgenomics.com/resources/datasets/pbmc-from-a-healthy-donor-granulocytes-removed-through-cell-sorting-10-k-1-standard-1-0-0" class="external-link">10x
Genomics</a>. As for our tutorial, the subsetted data we employ, stored
in a Seurat object format, is available for download on <a href="https://figshare.com/ndownloader/files/42039747" class="external-link">Figshare</a>.</p>
<p>Initially, we ran STREAM on the PBMC data to infer eRegulons, each of
which includes its regulating TF, its included genes and enhancers (and
their relations), as well as the eRegulon-active cells. We evaluated
these eRegulons from three perspectives, i.e., pathway enrichment of
eRegulon genes, overlap of eRegulon enhancers against chromatin regions
curated in databases (e.g., histone mark ChIP-seq peaks), and
intersection of enhancer-gene relations against relations curated in
databases (e.g., EnhancerAtlas 2.0 and scEnhancer). Then, we performed
hyper-geometric test for eRegulon-active cells against each cell type.
If the eRegulon-active cells are significantly enriched in a cell type
(e.g., CD4 TCM), we denoted the corresponding eRegulon as
cell-type-active eRegulon (CD4 TCM-active eRegulon). For each cell type,
we merged all cell-type-active eRegulons, leading to an eGRN. Finally,
by splitting the eGRN into sub-networks according to the regulating TFs,
we obtained cell-type-specific eRegulons.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dynload.html" class="external-link">dyn.load</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"/users/PAS1475/liyang/libs/hdf5_1.10.6/lib/libhdf5_hl.so.100"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/warning.html" class="external-link">suppressWarnings</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/invisible.html" class="external-link">invisible</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/YangLi-Bio/stream2" class="external-link">stream2</a></span><span class="op">)</span> <span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="data-loading">Data loading<a class="anchor" aria-label="anchor" href="#data-loading"></a>
</h2>
<p>Load a <code>Seurat</code> object containing jointly profiled
scRNA-seq and scATAC-seq assays from human PBMCs.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html" class="external-link">qread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">work.dir</span>, <span class="st">"tutorial_pbmc.qsave"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pbmc</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 6193 features across 1000 samples within 3 assays </span></span>
<span><span class="co">## Active assay: peaks (3991 features, 3991 variable features)</span></span>
<span><span class="co">##  2 other assays present: RNA, SCT</span></span>
<span><span class="co">##  3 dimensional reductions calculated: pca, lsi, umap</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="eregulon-inference">eRegulon inference<a class="anchor" aria-label="anchor" href="#eregulon-inference"></a>
</h2>
<p>Run STREAM on the Seurat object with default parameters.</p>
<p>In the function <code>run_stream</code>, variable annotations are
presented as follows:</p>
<ul>
<li><p><code>obj</code>: <code>Seurat</code> object composed of both RNA
and ATAC assays</p></li>
<li><p><code>qubic.path</code>: File path of the <code>QUBIC2</code>
biclustering program</p></li>
<li><p><code>peak.assay</code>: Name of the ATAC assay</p></li>
<li><p><code>var.genes</code>: Number of highly variable genes for LTMG
modeling</p></li>
<li><p><code>top.peaks</code>: Number of top-ranked peaks to build
heterogeneous graph</p></li>
<li><p><code>min.cells</code>: Minimum number of cells for quality
control</p></li>
<li><p><code>org</code>: Organism assembly version</p></li>
<li><p><code>c.cutoff</code>: Cutoff of expression consistency among
genes in the core part of hybrid biclusters</p></li>
<li><p><code>distance</code>: Size of window to connect co-accessible
peaks</p></li>
<li><p><code>BlockOverlap</code>: Threshold of overlap between prior
identified hybrid biclusters and the current seed</p></li>
<li><p><code>Extension</code>: Cutoff of expression consistency among
genes in the dual part of hybrid biclusters</p></li>
</ul>
<p>When it comes to jointly profiled RNA+ATAC data via different
techniques, we suggest different strategies. For datasets with low
scRNA-seq depth (e.g., <a href="https://www.sciencedirect.com/science/article/pii/S0092867420312538" class="external-link">SHARE-seq</a>),
we recommend imputation using mainstream approaches like <a href="https://www.cell.com/cell/pdf/S0092-8674(18)30724-4.pdf" class="external-link">MAGIC</a>,
<a href="https://www.nature.com/articles/s41592-018-0033-z" class="external-link">SAVER</a>,
and <a href="https://www.nature.com/articles/s41467-018-03405-7" class="external-link">scImpute</a>.
For datasets with low coverage in both modalities (e.g., <a href="https://www.nature.com/articles/s41587-019-0290-0" class="external-link">SNARE-seq</a>
and <a href="https://www.nature.com/articles/s41594-019-0323-x" class="external-link">PAIRED-seq</a>),
we suggest using <a href="https://www.nature.com/articles/s41587-023-01716-9" class="external-link">SEACells</a>
to build metacells by pooling similar cells. For PAIRED-seq, which has
even lower coverage than SNARE-seq, more stringent SEACells parameters
should be used.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This process can be resource-intensive. We advise against executing it on a personal computer.</span></span>
<span><span class="va">time1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span> <span class="va">en.regs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_stream.html">run_stream</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>                      qubic.path <span class="op">=</span> <span class="st">"/users/PAS1475/liyang/software/QUBIC2/qubic"</span>,</span>
<span>                      peak.assay <span class="op">=</span> <span class="st">"peaks"</span>,</span>
<span>                      var.genes <span class="op">=</span> <span class="fl">3000</span>,</span>
<span>                      top.peaks <span class="op">=</span> <span class="fl">3000</span>,</span>
<span>                      min.cells <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                      org <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>                      c.cutoff <span class="op">=</span> <span class="fl">1.0</span>,</span>
<span>                      distance <span class="op">=</span> <span class="fl">5e+05</span>,</span>
<span>                      BlockOverlap <span class="op">=</span> <span class="fl">0.50</span>,</span>
<span>                      Extension <span class="op">=</span> <span class="fl">1.0</span></span>
<span>                      <span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<pre><code>## number of iterations= 236 
## Progress:0%
## Progress:10%
## Progress:20%
## Progress:30%
## Progress:40%
## Progress:50%
## Progress:60%
## Progress:70%
## Progress:80%
## Progress:90%
## Progress:100%
## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |                                                                      |   1%
  |                                                                            
  |=                                                                     |   1%
  |                                                                            
  |=                                                                     |   2%
  |                                                                            
  |==                                                                    |   2%
  |                                                                            
  |==                                                                    |   3%
  |                                                                            
  |===                                                                   |   4%
  |                                                                            
  |===                                                                   |   5%
  |                                                                            
  |====                                                                  |   5%
  |                                                                            
  |====                                                                  |   6%
  |                                                                            
  |=====                                                                 |   7%
  |                                                                            
  |=====                                                                 |   8%
  |                                                                            
  |======                                                                |   8%
  |                                                                            
  |======                                                                |   9%
  |                                                                            
  |=======                                                               |   9%
  |                                                                            
  |=======                                                               |  10%
  |                                                                            
  |=======                                                               |  11%
  |                                                                            
  |========                                                              |  11%
  |                                                                            
  |========                                                              |  12%
  |                                                                            
  |=========                                                             |  12%
  |                                                                            
  |=========                                                             |  13%
  |                                                                            
  |==========                                                            |  14%
  |                                                                            
  |==========                                                            |  15%
  |                                                                            
  |===========                                                           |  15%
  |                                                                            
  |===========                                                           |  16%
  |                                                                            
  |============                                                          |  17%
  |                                                                            
  |============                                                          |  18%
  |                                                                            
  |=============                                                         |  18%
  |                                                                            
  |=============                                                         |  19%
  |                                                                            
  |==============                                                        |  19%
  |                                                                            
  |==============                                                        |  20%
  |                                                                            
  |==============                                                        |  21%
  |                                                                            
  |===============                                                       |  21%
  |                                                                            
  |===============                                                       |  22%
  |                                                                            
  |================                                                      |  22%
  |                                                                            
  |================                                                      |  23%
  |                                                                            
  |=================                                                     |  24%
  |                                                                            
  |=================                                                     |  25%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |==================                                                    |  26%
  |                                                                            
  |===================                                                   |  27%
  |                                                                            
  |===================                                                   |  28%
  |                                                                            
  |====================                                                  |  28%
  |                                                                            
  |====================                                                  |  29%
  |                                                                            
  |=====================                                                 |  29%
  |                                                                            
  |=====================                                                 |  30%
  |                                                                            
  |=====================                                                 |  31%
  |                                                                            
  |======================                                                |  31%
  |                                                                            
  |======================                                                |  32%
  |                                                                            
  |=======================                                               |  32%
  |                                                                            
  |=======================                                               |  33%
  |                                                                            
  |========================                                              |  34%
  |                                                                            
  |========================                                              |  35%
  |                                                                            
  |=========================                                             |  35%
  |                                                                            
  |=========================                                             |  36%
  |                                                                            
  |==========================                                            |  37%
  |                                                                            
  |==========================                                            |  38%
  |                                                                            
  |===========================                                           |  38%
  |                                                                            
  |===========================                                           |  39%
  |                                                                            
  |============================                                          |  39%
  |                                                                            
  |============================                                          |  40%
  |                                                                            
  |============================                                          |  41%
  |                                                                            
  |=============================                                         |  41%
  |                                                                            
  |=============================                                         |  42%
  |                                                                            
  |==============================                                        |  42%
  |                                                                            
  |==============================                                        |  43%
  |                                                                            
  |===============================                                       |  44%
  |                                                                            
  |===============================                                       |  45%
  |                                                                            
  |================================                                      |  45%
  |                                                                            
  |================================                                      |  46%
  |                                                                            
  |=================================                                     |  47%
  |                                                                            
  |=================================                                     |  48%
  |                                                                            
  |==================================                                    |  48%
  |                                                                            
  |==================================                                    |  49%
  |                                                                            
  |===================================                                   |  49%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |===================================                                   |  51%
  |                                                                            
  |====================================                                  |  51%
  |                                                                            
  |====================================                                  |  52%
  |                                                                            
  |=====================================                                 |  52%
  |                                                                            
  |=====================================                                 |  53%
  |                                                                            
  |======================================                                |  54%
  |                                                                            
  |======================================                                |  55%
  |                                                                            
  |=======================================                               |  55%
  |                                                                            
  |=======================================                               |  56%
  |                                                                            
  |========================================                              |  57%
  |                                                                            
  |========================================                              |  58%
  |                                                                            
  |=========================================                             |  58%
  |                                                                            
  |=========================================                             |  59%
  |                                                                            
  |==========================================                            |  59%
  |                                                                            
  |==========================================                            |  60%
  |                                                                            
  |==========================================                            |  61%
  |                                                                            
  |===========================================                           |  61%
  |                                                                            
  |===========================================                           |  62%
  |                                                                            
  |============================================                          |  62%
  |                                                                            
  |============================================                          |  63%
  |                                                                            
  |=============================================                         |  64%
  |                                                                            
  |=============================================                         |  65%
  |                                                                            
  |==============================================                        |  65%
  |                                                                            
  |==============================================                        |  66%
  |                                                                            
  |===============================================                       |  67%
  |                                                                            
  |===============================================                       |  68%
  |                                                                            
  |================================================                      |  68%
  |                                                                            
  |================================================                      |  69%
  |                                                                            
  |=================================================                     |  69%
  |                                                                            
  |=================================================                     |  70%
  |                                                                            
  |=================================================                     |  71%
  |                                                                            
  |==================================================                    |  71%
  |                                                                            
  |==================================================                    |  72%
  |                                                                            
  |===================================================                   |  72%
  |                                                                            
  |===================================================                   |  73%
  |                                                                            
  |====================================================                  |  74%
  |                                                                            
  |====================================================                  |  75%
  |                                                                            
  |=====================================================                 |  75%
  |                                                                            
  |=====================================================                 |  76%
  |                                                                            
  |======================================================                |  77%
  |                                                                            
  |======================================================                |  78%
  |                                                                            
  |=======================================================               |  78%
  |                                                                            
  |=======================================================               |  79%
  |                                                                            
  |========================================================              |  79%
  |                                                                            
  |========================================================              |  80%
  |                                                                            
  |========================================================              |  81%
  |                                                                            
  |=========================================================             |  81%
  |                                                                            
  |=========================================================             |  82%
  |                                                                            
  |==========================================================            |  82%
  |                                                                            
  |==========================================================            |  83%
  |                                                                            
  |===========================================================           |  84%
  |                                                                            
  |===========================================================           |  85%
  |                                                                            
  |============================================================          |  85%
  |                                                                            
  |============================================================          |  86%
  |                                                                            
  |=============================================================         |  87%
  |                                                                            
  |=============================================================         |  88%
  |                                                                            
  |==============================================================        |  88%
  |                                                                            
  |==============================================================        |  89%
  |                                                                            
  |===============================================================       |  89%
  |                                                                            
  |===============================================================       |  90%
  |                                                                            
  |===============================================================       |  91%
  |                                                                            
  |================================================================      |  91%
  |                                                                            
  |================================================================      |  92%
  |                                                                            
  |=================================================================     |  92%
  |                                                                            
  |=================================================================     |  93%
  |                                                                            
  |==================================================================    |  94%
  |                                                                            
  |==================================================================    |  95%
  |                                                                            
  |===================================================================   |  95%
  |                                                                            
  |===================================================================   |  96%
  |                                                                            
  |====================================================================  |  97%
  |                                                                            
  |====================================================================  |  98%
  |                                                                            
  |===================================================================== |  98%
  |                                                                            
  |===================================================================== |  99%
  |                                                                            
  |======================================================================|  99%
  |                                                                            
  |======================================================================| 100%
## [1] "Starting Cicero"
## [1] "Calculating distance_parameter value"
## [1] "Running models"
## [1] "Assembling connections"
## [1] "Successful cicero models:  1469"
## [1] "Other models: "
## 
## Zero or one element in range 
##                        12116 
## [1] "Models with errors:  0"
## [1] "Done"
## 
  |                                                        
  |                                                  |   0%
  |                                                        
  |=                                                 |   2%
  |                                                        
  |==                                                |   3%
  |                                                        
  |==                                                |   5%
  |                                                        
  |===                                               |   7%
  |                                                        
  |====                                              |   8%
  |                                                        
  |=====                                             |  10%
  |                                                        
  |======                                            |  12%
  |                                                        
  |=======                                           |  13%
  |                                                        
  |========                                          |  15%
  |                                                        
  |========                                          |  17%
  |                                                        
  |=========                                         |  18%
  |                                                        
  |==========                                        |  20%
  |                                                        
  |===========                                       |  22%
  |                                                        
  |============                                      |  23%
  |                                                        
  |============                                      |  25%
  |                                                        
  |=============                                     |  27%
  |                                                        
  |==============                                    |  28%
  |                                                        
  |===============                                   |  30%
  |                                                        
  |================                                  |  32%
  |                                                        
  |=================                                 |  33%
  |                                                        
  |==================                                |  35%
  |                                                        
  |==================                                |  37%
  |                                                        
  |===================                               |  38%
  |                                                        
  |====================                              |  40%
  |                                                        
  |=====================                             |  42%
  |                                                        
  |======================                            |  43%
  |                                                        
  |======================                            |  45%
  |                                                        
  |=======================                           |  47%
  |                                                        
  |========================                          |  48%
  |                                                        
  |=========================                         |  50%
  |                                                        
  |==========================                        |  52%
  |                                                        
  |===========================                       |  53%
  |                                                        
  |============================                      |  55%
  |                                                        
  |============================                      |  57%
  |                                                        
  |=============================                     |  58%
  |                                                        
  |==============================                    |  60%
  |                                                        
  |===============================                   |  62%
  |                                                        
  |================================                  |  63%
  |                                                        
  |================================                  |  65%
  |                                                        
  |=================================                 |  67%
  |                                                        
  |==================================                |  68%
  |                                                        
  |===================================               |  70%
  |                                                        
  |====================================              |  72%
  |                                                        
  |=====================================             |  73%
  |                                                        
  |======================================            |  75%
  |                                                        
  |======================================            |  77%
  |                                                        
  |=======================================           |  78%
  |                                                        
  |========================================          |  80%
  |                                                        
  |=========================================         |  82%
  |                                                        
  |==========================================        |  83%
  |                                                        
  |==========================================        |  85%
  |                                                        
  |===========================================       |  87%
  |                                                        
  |============================================      |  88%
  |                                                        
  |=============================================     |  90%
  |                                                        
  |==============================================    |  92%
  |                                                        
  |===============================================   |  93%
  |                                                        
  |================================================  |  95%
  |                                                        
  |================================================  |  97%
  |                                                        
  |================================================= |  98%
  |                                                        
  |==================================================| 100%</code></pre>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qsave.html" class="external-link">qsave</a></span><span class="op">(</span><span class="va">en.regs</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">work.dir</span>, <span class="st">"eRegulons.qsave"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "terminal"   "Tier"       "TF"         "genes"      "peaks"     </span></span>
<span><span class="co">##  [6] "cells"      "atac.ratio" "score"      "weight"     "links"     </span></span>
<span><span class="co">## [11] "seed"</span></span></code></pre>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "MAN1C1" "STMN1"  "EIF2B3" "ZSWIM5" "JUN"    "EVI5"</span></span></code></pre>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">peaks</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "chr1-25904959-25907184"   "chr1-45302959-45304454"  </span></span>
<span><span class="co">## [3] "chr1-58783212-58785307"   "chr1-93179195-93181524"  </span></span>
<span><span class="co">## [5] "chr1-108660816-108661818" "chr1-150578547-150580018"</span></span></code></pre>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">links</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 1 metadata column:</span></span>
<span><span class="co">##       seqnames            ranges strand |        gene</span></span>
<span><span class="co">##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1 25904959-25907184      * |      MAN1C1</span></span>
<span><span class="co">##   [2]     chr1 25904959-25907184      * |       STMN1</span></span>
<span><span class="co">##   [3]     chr1 45302959-45304454      * |      EIF2B3</span></span>
<span><span class="co">##   [4]     chr1 45302959-45304454      * |      ZSWIM5</span></span>
<span><span class="co">##   [5]     chr1 58783212-58785307      * |         JUN</span></span>
<span><span class="co">##   [6]     chr1 93179195-93181524      * |        EVI5</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 19 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span> <span class="op">(</span><span class="st">"The running time of eRegulon prediction is: "</span>, <span class="va">time1</span><span class="op">[</span><span class="st">"elapsed"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The running time of eRegulon prediction is: 564.544</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="egrn-construction">eGRN construction<a class="anchor" aria-label="anchor" href="#egrn-construction"></a>
</h2>
<p>Construct eGRN in each cell type.</p>
<p>In the function <code>get_cts_en_GRNs</code>, variabled are defined
below:</p>
<ul>
<li><p><code>obj</code>: <code>Seurat</code> object composed of both RNA
and ATAC assays</p></li>
<li><p><code>celltype</code>: Name of the metadata column indicating
cell types</p></li>
<li><p><code>en.regs</code>: List of predicted eRegulons</p></li>
<li><p><code>peak.assay</code>: Name of the ATAC assay</p></li>
<li><p><code>rna.dims</code>: Number of RNA dimensions for multimodal
clustering using <code>Seurat</code></p></li>
<li><p><code>atac.dims</code>: Number of ATAC dimensions for multimodal
clustering using <code>Seurat</code></p></li>
<li><p><code>padj.cutoff</code>: Threshold of adjusted P-values when
mapping eRegulons onto cell types</p></li>
<li><p><code>out.dir</code>: Directory to save the final result and
intermediate results</p></li>
</ul>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">time2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span> <span class="va">en.grns</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_cts_en_GRNs.html">get_cts_en_GRNs</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">pbmc</span>, celltype <span class="op">=</span> <span class="st">"predicted.id"</span>,</span>
<span>                           en.regs <span class="op">=</span> <span class="va">en.regs</span>, peak.assay <span class="op">=</span> <span class="st">"peaks"</span>,</span>
<span>                           rna.dims <span class="op">=</span> <span class="fl">50</span>, atac.dims <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                           padj.cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                           out.dir <span class="op">=</span> <span class="st">"./"</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qsave.html" class="external-link">qsave</a></span><span class="op">(</span><span class="va">en.grns</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">work.dir</span>, <span class="st">"eGRNs.qsave"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">en.grns</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "links"     "cell.type" "cells"</span></span></code></pre>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">en.grns</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">links</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames            ranges strand |        gene          TF</span></span>
<span><span class="co">##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1 25904959-25907184      * |      MAN1C1       TCF12</span></span>
<span><span class="co">##   [2]     chr1 25904959-25907184      * |       STMN1       TCF12</span></span>
<span><span class="co">##   [3]     chr1 45302959-45304454      * |      EIF2B3       TCF12</span></span>
<span><span class="co">##   [4]     chr1 45302959-45304454      * |      ZSWIM5       TCF12</span></span>
<span><span class="co">##   [5]     chr1 58783212-58785307      * |         JUN       TCF12</span></span>
<span><span class="co">##   [6]     chr1 93179195-93181524      * |        EVI5       TCF12</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 21 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span> <span class="op">(</span><span class="st">"The running time of eGRN construction is: "</span>, <span class="va">time2</span><span class="op">[</span><span class="st">"elapsed"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The running time of eGRN construction is: 1.28699999999992</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="cell-type-specific-eregulon-prediction">Cell-type-specific eRegulon prediction<a class="anchor" aria-label="anchor" href="#cell-type-specific-eregulon-prediction"></a>
</h2>
<p>Construct cell-type-specific eRegulons.</p>
<p>In the function <code>get_cts_en_regs</code>, variables are
introduced as follows:</p>
<ul>
<li><p><code>obj</code>: <code>Seurat</code> object composed of both RNA
and ATAC assays</p></li>
<li><p><code>peak.assay</code>: Name of the ATAC assay</p></li>
<li><p><code>de.genes</code>: List of differentially expressed genes
(DEGs)</p></li>
<li><p><code>cts.en.grns</code>: Cell-type-specific eGRNs saved in
GRanges object, the name of each of which is cell type, and the GRanges
object contains metadata columns â€œgeneâ€ and â€œTFâ€.</p></li>
<li><p><code>out.dir</code>: Directory to save the final result and
intermediate results</p></li>
<li><p><code>celltype</code>: Name of the metadata column indicating
cell types</p></li>
<li><p><code>min.pct</code>: Cutoff of overlap percentage in calculating
DEGs</p></li>
<li><p><code>logfc.threshold</code>: Threshold of log-fold change in
predicting DEGs</p></li>
<li><p><code>padj.cutoff</code>: Cutoff of adjusted P-values in
identifying DEGs</p></li>
</ul>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">time3</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span> <span class="va">cts.en.regs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_cts_en_regs.html">get_cts_en_regs</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">pbmc</span>, peak.assay <span class="op">=</span> <span class="st">"peaks"</span>, de.genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                               cts.en.grns <span class="op">=</span> <span class="va">en.grns</span>, out.dir <span class="op">=</span> <span class="st">"./"</span>, celltype <span class="op">=</span> <span class="st">"predicted.id"</span>,</span>
<span>                               min.pct <span class="op">=</span> <span class="fl">0.25</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.25</span>, padj.cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qsave.html" class="external-link">qsave</a></span><span class="op">(</span><span class="va">cts.en.regs</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">work.dir</span>, <span class="st">"cell-type-specific-eRegulons.qsave"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cts.en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "TF"        "celltype"  "cells"     "genes"     "enhancers" "links"</span></span></code></pre>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cts.en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ADAMTSL4-AS1" "TSPAN5"       "ADAM19"       "ATXN1"        "EEF1A1"      </span></span>
<span><span class="co">## [6] "TNFAIP3"</span></span></code></pre>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cts.en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">enhancers</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "chr1-150320808-150322695" "chr4-98928092-98930218"  </span></span>
<span><span class="co">## [3] "chr5-157252607-157253720" "chr6-16759215-16763347"  </span></span>
<span><span class="co">## [5] "chr6-73653070-73655111"   "chr6-137707572-137708859"</span></span></code></pre>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cts.en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">links</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames              ranges strand |         gene          TF</span></span>
<span><span class="co">##          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1 150320808-150322695      * | ADAMTSL4-AS1       CEBPD</span></span>
<span><span class="co">##   [2]     chr4   98928092-98930218      * |       TSPAN5       CEBPD</span></span>
<span><span class="co">##   [3]     chr5 157252607-157253720      * |       ADAM19       CEBPD</span></span>
<span><span class="co">##   [4]     chr6   16759215-16763347      * |        ATXN1       CEBPD</span></span>
<span><span class="co">##   [5]     chr6   73653070-73655111      * |       EEF1A1       CEBPD</span></span>
<span><span class="co">##   [6]     chr6 137707572-137708859      * |      TNFAIP3       CEBPD</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 21 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span> <span class="op">(</span><span class="st">"The running time of cell-type-specific eRegulon discovery is: "</span>, <span class="va">time3</span><span class="op">[</span><span class="st">"elapsed"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The running time of cell-type-specific eRegulon discovery is: 0.673000000000002</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="evaluation-of-cell-type-specific-eregulons">Evaluation of cell-type-specific eRegulons<a class="anchor" aria-label="anchor" href="#evaluation-of-cell-type-specific-eregulons"></a>
</h2>
<p>Firstly, we performed pathway enrichment analysis for gene lists in
cell-type-specific eRegulons against Gene Ontology (GO) or KEGG.</p>
<p>In the function <code>enrich_genes</code>, variables are defined as
follows:</p>
<ul>
<li><p><code>regs</code>: The list of enhancer regulons (eRegulons) or
cell-type-specific eRegulons.</p></li>
<li><p><code>dbs</code>: The list of databases to run enrichment
analysis, c(â€œGOâ€, â€œKEGGâ€) by default.</p></li>
</ul>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/psolymos/pbapply" class="external-link">pbapply</a></span><span class="op">)</span></span>
<span><span class="va">time4</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span> <span class="va">pathways</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/enrich_genes.html">enrich_genes</a></span><span class="op">(</span>regs <span class="op">=</span> <span class="va">cts.en.regs</span>, dbs <span class="op">=</span> <span class="st">"GO"</span><span class="op">)</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Uploading data to Enrichr... Done.</span></span>
<span><span class="co">##   Querying GO_Molecular_Function_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Cellular_Component_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Biological_Process_2018... Done.</span></span>
<span><span class="co">## Parsing results... Done.</span></span>
<span><span class="co">## Uploading data to Enrichr... Done.</span></span>
<span><span class="co">##   Querying GO_Molecular_Function_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Cellular_Component_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Biological_Process_2018... Done.</span></span>
<span><span class="co">## Parsing results... Done.</span></span>
<span><span class="co">## Uploading data to Enrichr... Done.</span></span>
<span><span class="co">##   Querying GO_Molecular_Function_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Cellular_Component_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Biological_Process_2018... Done.</span></span>
<span><span class="co">## Parsing results... Done.</span></span>
<span><span class="co">## Uploading data to Enrichr... Done.</span></span>
<span><span class="co">##   Querying GO_Molecular_Function_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Cellular_Component_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Biological_Process_2018... Done.</span></span>
<span><span class="co">## Parsing results... Done.</span></span>
<span><span class="co">## Uploading data to Enrichr... Done.</span></span>
<span><span class="co">##   Querying GO_Molecular_Function_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Cellular_Component_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Biological_Process_2018... Done.</span></span>
<span><span class="co">## Parsing results... Done.</span></span></code></pre>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sig.pathways</span> <span class="op">&lt;-</span> <span class="fu">pbapply</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/pbapply/man/pbapply.html" class="external-link">pblapply</a></span><span class="op">(</span><span class="va">pathways</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">x</span><span class="op">[</span><span class="va">x</span><span class="op">$</span><span class="va">Adjusted.P.value</span> <span class="op">&lt;</span> <span class="fl">0.05</span>,, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qsave.html" class="external-link">qsave</a></span><span class="op">(</span><span class="va">sig.pathways</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">work.dir</span>, <span class="st">"enriched_pathways.qsave"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sig.pathways</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   Id                                                Term Overlap     P.value</span></span>
<span><span class="co">## 1  1                        poly(G) binding (GO:0034046)     1/7 0.003495224</span></span>
<span><span class="co">## 2  1       phospholipase inhibitor activity (GO:0004859)     1/8 0.003993646</span></span>
<span><span class="co">## 3  1                    Tat protein binding (GO:0030957)    1/10 0.004989818</span></span>
<span><span class="co">## 4  1 Lys63-specific deubiquitinase activity (GO:0061578)    1/11 0.005487568</span></span>
<span><span class="co">## 5  1          nitric-oxide synthase binding (GO:0050998)    1/11 0.005487568</span></span>
<span><span class="co">## 6  1 translation elongation factor activity (GO:0003746)    1/13 0.006482395</span></span>
<span><span class="co">##   Adjusted.P.value Old.P.value Old.Adjusted.P.value Odds.Ratio Combined.Score</span></span>
<span><span class="co">## 1       0.04006892           0                    0   370.0741      2093.2713</span></span>
<span><span class="co">## 2       0.04006892           0                    0   317.1905      1751.8591</span></span>
<span><span class="co">## 3       0.04006892           0                    0   246.6790      1307.4865</span></span>
<span><span class="co">## 4       0.04006892           0                    0   222.0000      1155.5700</span></span>
<span><span class="co">## 5       0.04006892           0                    0   222.0000      1155.5700</span></span>
<span><span class="co">## 6       0.04006892           0                    0   184.9815       932.0597</span></span>
<span><span class="co">##     Genes</span></span>
<span><span class="co">## 1   ATXN1</span></span>
<span><span class="co">## 2   ANXA1</span></span>
<span><span class="co">## 3    ACTB</span></span>
<span><span class="co">## 4 TNFAIP3</span></span>
<span><span class="co">## 5    ACTB</span></span>
<span><span class="co">## 6  EEF1A1</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span> <span class="op">(</span><span class="st">"The running time of pathway enrichment analysis for eRegulons is: "</span>, <span class="va">time4</span><span class="op">[</span><span class="st">"elapsed"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The running time of pathway enrichment analysis for eRegulons is: 2.81799999999998</span></span></code></pre>
<p>Then, we intersected the enhancer list of a cell-type-specific
eRegulon against H3K27ac ChIP-seq peaks in CD4+ T cells with ENCODE
accession code: <a href="https://www.encodeproject.org/files/ENCFF498QFU/" class="external-link">ENCFF498QFU</a>.</p>
<p>In the function <code>intersect_peaks</code>, variables are described
as follows:</p>
<ul>
<li><p><code>x</code>: The first GRanges object or data.frame.</p></li>
<li><p><code>y</code>: The first GRanges object or data.frame.</p></li>
</ul>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stuart-lab/signac" class="external-link">Signac</a></span><span class="op">)</span></span>
<span><span class="va">chipseq.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">work.dir</span>, <span class="st">"ENCFF498QFU.bed"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">V1</span>, <span class="va">V2</span>, <span class="va">V3</span><span class="op">)</span></span>
<span><span class="va">chipseq.peaks</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://stuartlab.org/signac/reference/StringToGRanges.html" class="external-link">StringToGRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">chipseq.df</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, <span class="st">"-"</span>, <span class="va">chipseq.df</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>, <span class="st">"-"</span>, <span class="va">chipseq.df</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">time5</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span> <span class="va">permTest.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/intersect_peaks.html">intersect_peaks</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu">Signac</span><span class="fu">::</span><span class="fu"><a href="https://stuartlab.org/signac/reference/StringToGRanges.html" class="external-link">StringToGRanges</a></span><span class="op">(</span><span class="va">cts.en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">enhancers</span><span class="op">)</span>, </span>
<span>                                y <span class="op">=</span> <span class="va">chipseq.peaks</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="va">permTest.res</span><span class="op">$</span><span class="va">numOverlaps</span></span></code></pre></div>
<pre><code><span><span class="co">## P-value: 0.0099009900990099</span></span>
<span><span class="co">## Z-score: 14.9248</span></span>
<span><span class="co">## Number of iterations: 100</span></span>
<span><span class="co">## Alternative: greater</span></span>
<span><span class="co">## Evaluation of the original region set: 11</span></span>
<span><span class="co">## Evaluation function: numOverlaps</span></span>
<span><span class="co">## Randomization function: randomizeRegions</span></span></code></pre>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span> <span class="op">(</span><span class="st">"The running time of evaluating TF-enhancer relations in eRegulons is: "</span>, <span class="va">time5</span><span class="op">[</span><span class="st">'elapsed'</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The running time of evaluating TF-enhancer relations in eRegulons is: 7.50099999999998</span></span></code></pre>
<p>Finally, we overlapped the enhancer-gene relations in a
cell-type-specific eRegulon against enhancer-target pairs in CD4 TCM
cells curated by <a href="http://enhanceratlas.net/scenhancer/data/download/interaction/hs/Memory_CD4_T_interaction.txt" class="external-link">scEnhancer</a>.</p>
<p>In the function <code>intersect_enhancer_gene_relations</code>,
variables are introduced as follows:</p>
<ul>
<li><p><code>x</code>: The first GRanges object saving enhancer-gene
relations with gene symbols saved in â€œgeneâ€ meta column</p></li>
<li><p><code>y</code>: The second GRanges object saving enhancer-gene
relations with gene symbols saved in â€œgeneâ€ meta column</p></li>
</ul>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Preprocess the peak-gene interaction pairs in scEnhancer database</span></span>
<span><span class="va">enhGene.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">work.dir</span>, <span class="st">"Memory_CD4_T_interaction.txt"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">V1</span>, <span class="va">V2</span><span class="op">)</span></span>
<span><span class="va">enhGene.pairs</span> <span class="op">&lt;-</span> <span class="fu">Signac</span><span class="fu">::</span><span class="fu"><a href="https://stuartlab.org/signac/reference/StringToGRanges.html" class="external-link">StringToGRanges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">enhGene.df</span><span class="op">$</span><span class="va">V1</span>, split <span class="op">=</span> <span class="st">"\\|"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">.</span>, <span class="st">"[["</span>, <span class="fl">1</span><span class="op">)</span>, sep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">":"</span>, <span class="st">"-"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">mcols</span><span class="op">(</span><span class="va">enhGene.pairs</span><span class="op">)</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">enhGene.df</span><span class="op">$</span><span class="va">V2</span>, split <span class="op">=</span> <span class="st">"[\\||\\:]"</span><span class="op">)</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">.</span>, <span class="st">"[["</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">enhGene.pairs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 1 metadata column:</span></span>
<span><span class="co">##       seqnames              ranges strand |        gene</span></span>
<span><span class="co">##          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##   [1]    chr10 100227139-100227639      * |        HPS1</span></span>
<span><span class="co">##   [2]    chr10 100227139-100227639      * |        HPS1</span></span>
<span><span class="co">##   [3]    chr10 100227139-100227639      * |        HPS1</span></span>
<span><span class="co">##   [4]    chr10 100227139-100227639      * |        HPS1</span></span>
<span><span class="co">##   [5]    chr10 101629999-101630499      * |       DNMBP</span></span>
<span><span class="co">##   [6]    chr10 101629999-101630499      * |       DNMBP</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate the overlaps between enhancer-gene relations in a cell-type-specific eRegulon</span></span>
<span><span class="va">time6</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.time.html" class="external-link">system.time</a></span><span class="op">(</span><span class="va">overlaps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/intersect_enhancer_gene_relations.html">intersect_enhancer_gene_relations</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu">resize</span><span class="op">(</span><span class="va">cts.en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">links</span>, width <span class="op">=</span> <span class="fl">50000</span>, fix <span class="op">=</span> <span class="st">"center"</span><span class="op">)</span>, </span>
<span>                                              y <span class="op">=</span> <span class="fu">resize</span><span class="op">(</span><span class="va">enhGene.pairs</span>, width <span class="op">=</span> <span class="fl">50000</span>, fix <span class="op">=</span> <span class="st">"center"</span><span class="op">)</span> <span class="op">)</span> <span class="op">)</span></span>
<span><span class="va">overlaps</span></span></code></pre></div>
<pre><code><span><span class="co">##                    x.peak                 y.peak  gene</span></span>
<span><span class="co">## 1: chr6-16736281-16786280 chr6-16686955-16736954 ATXN1</span></span>
<span><span class="co">## 2: chr6-16736281-16786280 chr6-16744524-16794523 ATXN1</span></span>
<span><span class="co">## 3: chr6-16736281-16786280 chr6-16746364-16796363 ATXN1</span></span>
<span><span class="co">## 4:   chr7-5372117-5422116   chr7-5412511-5462510  ACTB</span></span></code></pre>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/message.html" class="external-link">message</a></span> <span class="op">(</span><span class="st">"The running time of assessing enhancer-gene relations in eRegulons is: "</span>, <span class="va">time6</span><span class="op">[</span><span class="st">"elapsed"</span><span class="op">]</span>, <span class="st">"\n\n"</span>, </span>
<span>         <span class="st">"The total time for the implementation and evaluation of eRegulon/eGRN prediction is: "</span>, </span>
<span>         <span class="va">time1</span><span class="op">[</span><span class="st">"elapsed"</span><span class="op">]</span> <span class="op">+</span> <span class="va">time2</span><span class="op">[</span><span class="st">"elapsed"</span><span class="op">]</span> <span class="op">+</span> <span class="va">time3</span><span class="op">[</span><span class="st">"elapsed"</span><span class="op">]</span> <span class="op">+</span> <span class="va">time4</span><span class="op">[</span><span class="st">"elapsed"</span><span class="op">]</span> <span class="op">+</span> <span class="va">time5</span><span class="op">[</span><span class="st">"elapsed"</span><span class="op">]</span> <span class="op">+</span> <span class="va">time6</span><span class="op">[</span><span class="st">"elapsed"</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The running time of assessing enhancer-gene relations in eRegulons is: 0.0580000000001064</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## The total time for the implementation and evaluation of eRegulon/eGRN prediction is: 576.881</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yang Li, Anjun Ma, Yizhong Wang, Qi Guo, Cankun Wang.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
