<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>From Simulation to Insight: Inferring eRegulons and eGRNs using STREAM â€¢ stream2</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="From Simulation to Insight: Inferring eRegulons and eGRNs using STREAM">
<meta property="og:description" content="stream2">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stream2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/install.html">Install</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/simul-analyses.html">Simulated data analyses</a>
    </li>
    <li>
      <a href="../articles/bio-analyses.html">Biological data analyses</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/YangLi-Bio/stream2" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>From Simulation to Insight: Inferring eRegulons
and eGRNs using STREAM</h1>
            
            <h4 data-toc-skip class="date">Compiled: August 20,
2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/YangLi-Bio/stream2/blob/HEAD/vignettes/eRegulon-inference.Rmd" class="external-link"><code>vignettes/eRegulon-inference.Rmd</code></a></small>
      <div class="hidden name"><code>eRegulon-inference.Rmd</code></div>

    </div>

    
    
<p>In this tutorial, we will guide you through the process of inferring
enhancer regulons, known as eRegulons, and enhancer-driven gene
regulatory networks or eGRNs, utilizing the STREAM tool. Our
demonstration leverages a select subset of data from the publicly
accessible 10x Genomics Multiome dataset, specifically focusing on human
Peripheral Blood Mononuclear Cells (PBMCs). For those interested in the
comprehensive dataset, the raw data captured by the 10X Genomics
Multiome ATAC+GEX can be downloaded directly from [10x Genomics] (<a href="https://www.10xgenomics.com/resources/datasets/pbmc-from-a-healthy-donor-granulocytes-removed-through-cell-sorting-10-k-1-standard-1-0-0" class="external-link uri">https://www.10xgenomics.com/resources/datasets/pbmc-from-a-healthy-donor-granulocytes-removed-through-cell-sorting-10-k-1-standard-1-0-0</a>).
As for our tutorial, the subsetted data we employ, stored in a Seurat
object format, is available for download on [Figshare] (<a href="https://figshare.com/ndownloader/files/42039747" class="external-link uri">https://figshare.com/ndownloader/files/42039747</a>).</p>
<p>Initially, we ran STREAM on the PBMC data to infer eRegulons, each of
which includes its regulating TF, its included genes and enhancers (and
their relations), as well as the eRegulon-active cells. We evaluated
these eRegulons from three perspectives, i.e., pathway enrichment of
eRegulon genes, overlap of eRegulon enhancers against chromatin regions
curated in databases (e.g., histone mark ChIP-seq peaks), and
intersection of enhancer-gene relations against relations curated in
databases (e.g., EnhancerAtlas 2.0 and scEnhancer). Then, we performed
hyper-geometric test for eRegulon-active cells against each cell type.
If the eRegulon-active cells are significantly enriched in a cell type
(e.g., CD4 TCM), we denoted the corresponding eRegulon as
cell-type-active eRegulon (CD4 TCM-active eRegulon). For each cell type,
we merged all cell-type-active eRegulons, leading to an eGRN. Finally,
by splitting the eGRN into sub-networks according to the regulating TFs,
we obtained cell-type-specific eRegulons.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/dynload.html" class="external-link">dyn.load</a></span><span class="op">(</span>x <span class="op">=</span> <span class="st">"/users/PAS1475/liyang/libs/hdf5_1.10.6/lib/libhdf5_hl.so.100"</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">easypackages</span><span class="op">)</span></span>
<span><span class="va">libs</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"qs"</span>, </span>
<span>  <span class="st">"hdf5r"</span>,</span>
<span>  <span class="st">"Seurat"</span>,</span>
<span>  <span class="st">"Signac"</span>, </span>
<span>  <span class="st">"pbmcapply"</span>, </span>
<span>  <span class="st">"pbapply"</span>, </span>
<span>  <span class="st">"parallel"</span>,</span>
<span>  <span class="st">"EnsDb.Mmusculus.v79"</span>, </span>
<span>  <span class="st">"EnsDb.Hsapiens.v86"</span>,</span>
<span>  <span class="st">"dplyr"</span>, </span>
<span>  <span class="st">"ggplot2"</span>, </span>
<span>  <span class="st">"ggpubr"</span>, </span>
<span>  <span class="st">"igraph"</span>, </span>
<span>  <span class="st">"Matrix"</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/easypackages/man/libraries.html" class="external-link">libraries</a></span><span class="op">(</span><span class="va">libs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: qs</span></span></code></pre>
<pre><code><span><span class="co">## qs 0.25.5</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: hdf5r</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: Seurat</span></span></code></pre>
<pre><code><span><span class="co">## The legacy packages maptools, rgdal, and rgeos, underpinning the sp package,</span></span>
<span><span class="co">## which was just loaded, will retire in October 2023.</span></span>
<span><span class="co">## Please refer to R-spatial evolution reports for details, especially</span></span>
<span><span class="co">## https://r-spatial.org/r/2023/05/15/evolution4.html.</span></span>
<span><span class="co">## It may be desirable to make the sf package available;</span></span>
<span><span class="co">## package maintainers should consider adding sf to Suggests:.</span></span>
<span><span class="co">## The sp package is now running under evolution status 2</span></span>
<span><span class="co">##      (status 2 uses the sf package in place of rgdal)</span></span></code></pre>
<pre><code><span><span class="co">## Attaching SeuratObject</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: Signac</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: pbmcapply</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: parallel</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: pbapply</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: EnsDb.Mmusculus.v79</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: ensembldb</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: BiocGenerics</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'BiocGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     IQR, mad, sd, var, xtabs</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyDuplicated, aperm, append, as.data.frame, basename, cbind,</span></span>
<span><span class="co">##     colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,</span></span>
<span><span class="co">##     get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,</span></span>
<span><span class="co">##     match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,</span></span>
<span><span class="co">##     Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,</span></span>
<span><span class="co">##     table, tapply, union, unique, unsplit, which.max, which.min</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomicRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: stats4</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: S4Vectors</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'S4Vectors'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:hdf5r':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     values</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand.grid, I, unname</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: IRanges</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomeInfoDb</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: GenomicFeatures</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: AnnotationDbi</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: Biobase</span></span></code></pre>
<pre><code><span><span class="co">## Welcome to Bioconductor</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Vignettes contain introductory material; view with</span></span>
<span><span class="co">##     'browseVignettes()'. To cite Bioconductor, see</span></span>
<span><span class="co">##     'citation("Biobase")', and for packages 'citation("pkgname")'.</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: AnnotationFilter</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'ensembldb'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     filter</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: EnsDb.Hsapiens.v86</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: dplyr</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'dplyr'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:ensembldb':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     filter, select</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:AnnotationDbi':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     select</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     combine</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:GenomicRanges':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect, setdiff, union</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:GenomeInfoDb':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:IRanges':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     collapse, desc, intersect, setdiff, slice, union</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:S4Vectors':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     first, intersect, rename, setdiff, setequal, union</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:BiocGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     combine, intersect, setdiff, union</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     filter, lag</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     intersect, setdiff, setequal, union</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: ggplot2</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: ggpubr</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: igraph</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'igraph'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:dplyr':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     as_data_frame, groups, union</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:GenomicRanges':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     union</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:IRanges':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     union</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:S4Vectors':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     union</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:BiocGenerics':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     normalize, path, union</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     decompose, spectrum</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     union</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: Matrix</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Matrix'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:S4Vectors':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     expand</span></span></code></pre>
<pre><code><span><span class="co">## All packages loaded successfully</span></span></code></pre>
<div class="sourceCode" id="cb61"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/YangLi-Bio/stream2" class="external-link">stream2</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Warning: replacing previous import 'SummarizedExperiment::shift' by</span></span>
<span><span class="co">## 'data.table::shift' when loading 'stream2'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'data.table::last' by 'dplyr::last' when</span></span>
<span><span class="co">## loading 'stream2'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'data.table::first' by 'dplyr::first' when</span></span>
<span><span class="co">## loading 'stream2'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'data.table::between' by 'dplyr::between'</span></span>
<span><span class="co">## when loading 'stream2'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'dplyr::union' by 'igraph::union' when</span></span>
<span><span class="co">## loading 'stream2'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'dplyr::as_data_frame' by</span></span>
<span><span class="co">## 'igraph::as_data_frame' when loading 'stream2'</span></span></code></pre>
<pre><code><span><span class="co">## Warning: replacing previous import 'dplyr::groups' by 'igraph::groups' when</span></span>
<span><span class="co">## loading 'stream2'</span></span></code></pre>
<pre><code><span><span class="co">## cistrome_tools.R</span></span></code></pre>
<pre><code><span><span class="co">## epigenome_tools.R</span></span></code></pre>
<pre><code><span><span class="co">## multiome_tools.R</span></span></code></pre>
<pre><code><span><span class="co">## protein.R</span></span></code></pre>
<pre><code><span><span class="co">## reguome_tools.R</span></span></code></pre>
<pre><code><span><span class="co">## run_stream.R</span></span></code></pre>
<pre><code><span><span class="co">## stream2.R</span></span></code></pre>
<pre><code><span><span class="co">## transcriptome_tools.R</span></span></code></pre>
<pre><code><span><span class="co">## utilities.R</span></span></code></pre>
<pre><code><span><span class="co">## visual.R</span></span></code></pre>
<pre><code><span><span class="co">## [[1]]</span></span>
<span><span class="co">## [[1]]$value</span></span>
<span><span class="co">## function (object, peak.assay = "ATAC", expression.assay = "RNA", </span></span>
<span><span class="co">##     peak.slot = "counts", expression.slot = "data", method = "pearson", </span></span>
<span><span class="co">##     gene.coords = NULL, distance = 5e+05, min.distance = NULL, </span></span>
<span><span class="co">##     min.cells = 10, genes.use = NULL, n_sample = 100, pvalue_cutoff = 0.05, </span></span>
<span><span class="co">##     score_cutoff = 0.05, gene.id = FALSE, verbose = TRUE) </span></span>
<span><span class="co">## {</span></span>
<span><span class="co">##     if (!requireNamespace(package = "qlcMatrix", quietly = TRUE)) {</span></span>
<span><span class="co">##         stop("Please install qlcMatrix: install.packages('qlcMatrix')")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if (!inherits(x = object[[peak.assay]], what = "ChromatinAssay")) {</span></span>
<span><span class="co">##         stop("The requested assay is not a ChromatinAssay")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if (!is.null(x = min.distance)) {</span></span>
<span><span class="co">##         if (!is.numeric(x = min.distance)) {</span></span>
<span><span class="co">##             stop("min.distance should be a numeric value")</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         if (min.distance &lt; 0) {</span></span>
<span><span class="co">##             warning("Requested a negative min.distance value, setting min.distance to zero")</span></span>
<span><span class="co">##             min.distance &lt;- NULL</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         else if (min.distance == 0) {</span></span>
<span><span class="co">##             min.distance &lt;- NULL</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     features.match &lt;- c("GC.percent", "count", "sequence.length")</span></span>
<span><span class="co">##     if (method == "pearson") {</span></span>
<span><span class="co">##         cor_method &lt;- qlcMatrix::corSparse</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     else if (method == "spearman") {</span></span>
<span><span class="co">##         cor_method &lt;- SparseSpearmanCor</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     else {</span></span>
<span><span class="co">##         stop("method can be one of 'pearson' or 'spearman'.")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if (is.null(x = gene.coords)) {</span></span>
<span><span class="co">##         annot &lt;- Signac::Annotation(object = object[[peak.assay]])</span></span>
<span><span class="co">##         if (is.null(x = annot)) {</span></span>
<span><span class="co">##             stop("Gene annotations not found")</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         gene.coords &lt;- CollapseToLongestTranscript(ranges = annot)</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     meta.features &lt;- Seurat::GetAssayData(object = object, assay = peak.assay, </span></span>
<span><span class="co">##         slot = "meta.features")</span></span>
<span><span class="co">##     if (!(all(c("GC.percent", "sequence.length") %in% colnames(x = meta.features)))) {</span></span>
<span><span class="co">##         stop("DNA sequence information for each peak has not been computed.\n", </span></span>
<span><span class="co">##             "Run RegionsStats before calling this function.")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if (!("count" %in% colnames(x = meta.features))) {</span></span>
<span><span class="co">##         data.use &lt;- Seurat::GetAssayData(object = object[[peak.assay]], </span></span>
<span><span class="co">##             slot = "counts")</span></span>
<span><span class="co">##         hvf.info &lt;- Signac::FindTopFeatures(object = data.use, </span></span>
<span><span class="co">##             verbose = FALSE)</span></span>
<span><span class="co">##         hvf.info &lt;- hvf.info[rownames(meta.features), , drop = FALSE]</span></span>
<span><span class="co">##         meta.features &lt;- cbind(meta.features, hvf.info)</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     peak.data &lt;- Seurat::GetAssayData(object = object, assay = peak.assay, </span></span>
<span><span class="co">##         slot = peak.slot)</span></span>
<span><span class="co">##     expression.data &lt;- Seurat::GetAssayData(object = object, </span></span>
<span><span class="co">##         assay = expression.assay, slot = expression.slot)</span></span>
<span><span class="co">##     peakcounts &lt;- rowSums(x = peak.data &gt; 0)</span></span>
<span><span class="co">##     genecounts &lt;- rowSums(x = expression.data &gt; 0)</span></span>
<span><span class="co">##     peaks.keep &lt;- peakcounts &gt; min.cells</span></span>
<span><span class="co">##     genes.keep &lt;- genecounts &gt; min.cells</span></span>
<span><span class="co">##     peak.data &lt;- peak.data[peaks.keep, , drop = FALSE]</span></span>
<span><span class="co">##     if (!is.null(x = genes.use)) {</span></span>
<span><span class="co">##         genes.keep &lt;- intersect(x = names(x = genes.keep[genes.keep]), </span></span>
<span><span class="co">##             y = genes.use)</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     expression.data &lt;- expression.data[genes.keep, , drop = FALSE]</span></span>
<span><span class="co">##     if (verbose) {</span></span>
<span><span class="co">##         message("Testing ", nrow(x = expression.data), " genes and ", </span></span>
<span><span class="co">##             sum(peaks.keep), " peaks")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     genes &lt;- rownames(x = expression.data)</span></span>
<span><span class="co">##     if (gene.id) {</span></span>
<span><span class="co">##         gene.coords.use &lt;- gene.coords[gene.coords$gene_id %in% </span></span>
<span><span class="co">##             genes, ]</span></span>
<span><span class="co">##         gene.coords.use$gene_name &lt;- gene.coords.use$gene_id</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     else {</span></span>
<span><span class="co">##         gene.coords.use &lt;- gene.coords[gene.coords$gene_name %in% </span></span>
<span><span class="co">##             genes, ]</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if (length(x = gene.coords.use) == 0) {</span></span>
<span><span class="co">##         stop("Could not find gene coordinates for requested genes")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if (length(x = gene.coords.use) &lt; nrow(x = expression.data)) {</span></span>
<span><span class="co">##         message("Found gene coordinates for ", length(x = gene.coords.use), </span></span>
<span><span class="co">##             " genes")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     peaks &lt;- Signac::StringToGRanges(rownames(object[[peak.assay]]))</span></span>
<span><span class="co">##     peaks &lt;- peaks[peaks.keep]</span></span>
<span><span class="co">##     peak_distance_matrix &lt;- DistanceToTSS(peaks = peaks, genes = gene.coords.use, </span></span>
<span><span class="co">##         distance = distance)</span></span>
<span><span class="co">##     if (!is.null(x = min.distance)) {</span></span>
<span><span class="co">##         peak_distance_matrix_min &lt;- DistanceToTSS(peaks = peaks, </span></span>
<span><span class="co">##             genes = gene.coords.use, distance = min.distance)</span></span>
<span><span class="co">##         peak_distance_matrix &lt;- peak_distance_matrix - peak_distance_matrix_min</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if (sum(peak_distance_matrix) == 0) {</span></span>
<span><span class="co">##         stop("No peaks fall within distance threshold\n", "Have you set the proper genome and seqlevelsStyle for ", </span></span>
<span><span class="co">##             peak.assay, " assay?")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     genes.use &lt;- colnames(x = peak_distance_matrix)</span></span>
<span><span class="co">##     all.peaks &lt;- rownames(x = peak.data)</span></span>
<span><span class="co">##     peak.data &lt;- quiet(SeuratDisk::Transpose(peak.data))</span></span>
<span><span class="co">##     coef.vec &lt;- c()</span></span>
<span><span class="co">##     gene.vec &lt;- c()</span></span>
<span><span class="co">##     zscore.vec &lt;- c()</span></span>
<span><span class="co">##     if (TRUE) {</span></span>
<span><span class="co">##         mylapply &lt;- pbmcapply::pbmclapply</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     else {</span></span>
<span><span class="co">##         mylapply &lt;- ifelse(test = verbose, yes = pblapply, no = lapply)</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     res &lt;- mylapply(X = seq_along(along.with = genes.use), mc.cores = max(parallel::detectCores()/2, </span></span>
<span><span class="co">##         1), FUN = function(i) {</span></span>
<span><span class="co">##         peak.use &lt;- as.logical(x = peak_distance_matrix[, genes.use[[i]]])</span></span>
<span><span class="co">##         gene.expression &lt;- t(x = expression.data[genes.use[[i]], </span></span>
<span><span class="co">##             , drop = FALSE])</span></span>
<span><span class="co">##         gene.chrom &lt;- as.character(x = seqnames(x = gene.coords.use[i]))</span></span>
<span><span class="co">##         if (sum(peak.use) &lt; 2) {</span></span>
<span><span class="co">##             return(list(gene = NULL, coef = NULL, zscore = NULL))</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         else {</span></span>
<span><span class="co">##             peak.access &lt;- peak.data[, peak.use, drop = FALSE]</span></span>
<span><span class="co">##             coef.result &lt;- cor_method(X = peak.access, Y = gene.expression)</span></span>
<span><span class="co">##             rownames(x = coef.result) &lt;- colnames(x = peak.access)</span></span>
<span><span class="co">##             coef.result &lt;- coef.result[abs(x = coef.result) &gt; </span></span>
<span><span class="co">##                 score_cutoff, , drop = FALSE]</span></span>
<span><span class="co">##             if (nrow(x = coef.result) == 0) {</span></span>
<span><span class="co">##                 return(list(gene = NULL, coef = NULL, zscore = NULL))</span></span>
<span><span class="co">##             }</span></span>
<span><span class="co">##             else {</span></span>
<span><span class="co">##                 peaks.test &lt;- rownames(x = coef.result)</span></span>
<span><span class="co">##                 trans.peaks &lt;- all.peaks[!grepl(pattern = paste0("^", </span></span>
<span><span class="co">##                   gene.chrom), x = all.peaks)]</span></span>
<span><span class="co">##                 meta.use &lt;- meta.features[trans.peaks, ]</span></span>
<span><span class="co">##                 pk.use &lt;- meta.features[peaks.test, ]</span></span>
<span><span class="co">##                 bg.peaks &lt;- lapply(X = seq_len(length.out = nrow(x = pk.use)), </span></span>
<span><span class="co">##                   FUN = function(x) {</span></span>
<span><span class="co">##                     Signac::MatchRegionStats(meta.feature = meta.use, </span></span>
<span><span class="co">##                       query.feature = pk.use[x, , drop = FALSE], </span></span>
<span><span class="co">##                       features.match = features.match, n = n_sample, </span></span>
<span><span class="co">##                       verbose = FALSE)</span></span>
<span><span class="co">##                   })</span></span>
<span><span class="co">##                 bg.access &lt;- peak.data[, unlist(x = bg.peaks), </span></span>
<span><span class="co">##                   drop = FALSE]</span></span>
<span><span class="co">##                 bg.coef &lt;- cor_method(X = bg.access, Y = gene.expression)</span></span>
<span><span class="co">##                 rownames(bg.coef) &lt;- colnames(bg.access)</span></span>
<span><span class="co">##                 zscores &lt;- vector(mode = "numeric", length = length(x = peaks.test))</span></span>
<span><span class="co">##                 for (j in seq_along(along.with = peaks.test)) {</span></span>
<span><span class="co">##                   coef.use &lt;- bg.coef[(((j - 1) * n_sample) + </span></span>
<span><span class="co">##                     1):(j * n_sample), ]</span></span>
<span><span class="co">##                   z &lt;- (coef.result[j] - mean(x = coef.use))/sd(x = coef.use)</span></span>
<span><span class="co">##                   zscores[[j]] &lt;- z</span></span>
<span><span class="co">##                 }</span></span>
<span><span class="co">##                 names(x = coef.result) &lt;- peaks.test</span></span>
<span><span class="co">##                 names(x = zscores) &lt;- peaks.test</span></span>
<span><span class="co">##                 zscore.vec &lt;- c(zscore.vec, zscores)</span></span>
<span><span class="co">##                 gene.vec &lt;- c(gene.vec, rep(i, length(x = coef.result)))</span></span>
<span><span class="co">##                 coef.vec &lt;- c(coef.vec, coef.result)</span></span>
<span><span class="co">##             }</span></span>
<span><span class="co">##             gc(verbose = FALSE)</span></span>
<span><span class="co">##             pval.vec &lt;- pnorm(q = -abs(x = zscore.vec))</span></span>
<span><span class="co">##             links.keep &lt;- pval.vec &lt; pvalue_cutoff</span></span>
<span><span class="co">##             if (sum(x = links.keep) == 0) {</span></span>
<span><span class="co">##                 return(list(gene = NULL, coef = NULL, zscore = NULL))</span></span>
<span><span class="co">##             }</span></span>
<span><span class="co">##             else {</span></span>
<span><span class="co">##                 gene.vec &lt;- gene.vec[links.keep]</span></span>
<span><span class="co">##                 coef.vec &lt;- coef.vec[links.keep]</span></span>
<span><span class="co">##                 zscore.vec &lt;- zscore.vec[links.keep]</span></span>
<span><span class="co">##                 return(list(gene = gene.vec, coef = coef.vec, </span></span>
<span><span class="co">##                   zscore = zscore.vec))</span></span>
<span><span class="co">##             }</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##     })</span></span>
<span><span class="co">##     gene.vec &lt;- do.call(what = c, args = lapply(X = res, FUN = `[[`, </span></span>
<span><span class="co">##         1))</span></span>
<span><span class="co">##     coef.vec &lt;- do.call(what = c, args = lapply(X = res, FUN = `[[`, </span></span>
<span><span class="co">##         2))</span></span>
<span><span class="co">##     zscore.vec &lt;- do.call(what = c, args = lapply(X = res, FUN = `[[`, </span></span>
<span><span class="co">##         3))</span></span>
<span><span class="co">##     if (length(x = coef.vec) == 0) {</span></span>
<span><span class="co">##         if (verbose) {</span></span>
<span><span class="co">##             message("No significant links found!")</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         return(object)</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     peak.key &lt;- seq_along(along.with = unique(x = names(x = coef.vec)))</span></span>
<span><span class="co">##     names(x = peak.key) &lt;- unique(x = names(x = coef.vec))</span></span>
<span><span class="co">##     coef.matrix &lt;- Matrix::sparseMatrix(i = gene.vec, j = peak.key[names(x = coef.vec)], </span></span>
<span><span class="co">##         x = coef.vec, dims = c(length(x = genes.use), max(peak.key)))</span></span>
<span><span class="co">##     rownames(x = coef.matrix) &lt;- genes.use</span></span>
<span><span class="co">##     colnames(x = coef.matrix) &lt;- names(x = peak.key)</span></span>
<span><span class="co">##     links &lt;- invisible(LinksToGRanges(linkmat = coef.matrix, </span></span>
<span><span class="co">##         gene.coords = gene.coords.use))</span></span>
<span><span class="co">##     z.matrix &lt;- Matrix::sparseMatrix(i = gene.vec, j = peak.key[names(x = zscore.vec)], </span></span>
<span><span class="co">##         x = zscore.vec, dims = c(length(x = genes.use), max(peak.key)))</span></span>
<span><span class="co">##     rownames(x = z.matrix) &lt;- genes.use</span></span>
<span><span class="co">##     colnames(x = z.matrix) &lt;- names(x = peak.key)</span></span>
<span><span class="co">##     z.lnk &lt;- LinksToGRanges(linkmat = z.matrix, gene.coords = gene.coords.use)</span></span>
<span><span class="co">##     links$zscore &lt;- z.lnk$score</span></span>
<span><span class="co">##     links$pvalue &lt;- pnorm(q = -abs(x = links$zscore))</span></span>
<span><span class="co">##     links &lt;- links[links$pvalue &lt; pvalue_cutoff]</span></span>
<span><span class="co">##     Signac::Links(object = object[[peak.assay]]) &lt;- links</span></span>
<span><span class="co">##     return(object)</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[1]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]</span></span>
<span><span class="co">## [[2]]$value</span></span>
<span><span class="co">## function (lst1, lst2) </span></span>
<span><span class="co">## {</span></span>
<span><span class="co">##     overlaps &lt;- GenomicRanges::findOverlaps(query = lst1, subject = lst2, </span></span>
<span><span class="co">##         type = "any", select = "all")</span></span>
<span><span class="co">##     Matrix::sparseMatrix(i = queryHits(x = overlaps), j = subjectHits(x = overlaps), </span></span>
<span><span class="co">##         x = 1, dims = c(length(x = lst1), length(x = lst2)))</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[2]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]</span></span>
<span><span class="co">## [[3]]$value</span></span>
<span><span class="co">## function (rna_counts = NULL, atac_counts = NULL, org = "hg38", </span></span>
<span><span class="co">##     frag.file = NULL, sep = c("-", "-")) </span></span>
<span><span class="co">## {</span></span>
<span><span class="co">##     message("Loaded scRNA-seq and scATAC-seq matrices of sizes: ", </span></span>
<span><span class="co">##         nrow(rna_counts), " x ", ncol(rna_counts), " and ", nrow(atac_counts), </span></span>
<span><span class="co">##         " x ", ncol(atac_counts), ".")</span></span>
<span><span class="co">##     pbmc &lt;- Seurat::CreateSeuratObject(counts = rna_counts)</span></span>
<span><span class="co">##     pbmc[["percent.mt"]] &lt;- Seurat::PercentageFeatureSet(pbmc, </span></span>
<span><span class="co">##         pattern = "^MT-")</span></span>
<span><span class="co">##     grange.counts &lt;- Signac::StringToGRanges(rownames(atac_counts), </span></span>
<span><span class="co">##         sep = sep)</span></span>
<span><span class="co">##     grange.use &lt;- BSgenome::seqnames(grange.counts) %in% GenomeInfoDb::standardChromosomes(grange.counts)</span></span>
<span><span class="co">##     atac_counts &lt;- atac_counts[as.vector(grange.use), ]</span></span>
<span><span class="co">##     annotations &lt;- Signac::GetGRangesFromEnsDb(ensdb = org_to_DB(org = org))</span></span>
<span><span class="co">##     ensembldb::seqlevelsStyle(annotations) &lt;- "UCSC"</span></span>
<span><span class="co">##     GenomeInfoDb::genome(annotations) &lt;- org</span></span>
<span><span class="co">##     if (!is.null(frag.file)) {</span></span>
<span><span class="co">##         chrom_assay &lt;- Signac::CreateChromatinAssay(counts = atac_counts, </span></span>
<span><span class="co">##             sep = sep, genome = org, fragments = frag.file, min.cells = 10, </span></span>
<span><span class="co">##             annotation = annotations)</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     else {</span></span>
<span><span class="co">##         chrom_assay &lt;- Signac::CreateChromatinAssay(counts = atac_counts, </span></span>
<span><span class="co">##             sep = sep, genome = org, min.cells = 10, annotation = annotations)</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     pbmc[["ATAC"]] &lt;- chrom_assay</span></span>
<span><span class="co">##     pbmc</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[3]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[4]]</span></span>
<span><span class="co">## [[4]]$value</span></span>
<span><span class="co">## function (key.TFs = NULL, db = "BioGrid", org = "human") </span></span>
<span><span class="co">## {</span></span>
<span><span class="co">##     message("Identifying proteins interacting with the ", length(key.TFs), </span></span>
<span><span class="co">##         " TFs provided by user.")</span></span>
<span><span class="co">##     if (db == "STRING") {</span></span>
<span><span class="co">##         message("Loaded the STRING database.")</span></span>
<span><span class="co">##         if (org == "mouse") {</span></span>
<span><span class="co">##             species &lt;- 10090</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         else {</span></span>
<span><span class="co">##             species &lt;- 9606</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         tmp.db &lt;- quiet(STRINGdb::STRINGdb$new(species = species))</span></span>
<span><span class="co">##         tmp.graph &lt;- quiet(tmp.db$get_graph())</span></span>
<span><span class="co">##         ensembl.lst &lt;- igraph::as_adj_list(tmp.graph, mode = "all") %&gt;% </span></span>
<span><span class="co">##             lapply(., function(x) {</span></span>
<span><span class="co">##                 strsplit(igraph::as_ids(seq = x), split = "\\.") %&gt;% </span></span>
<span><span class="co">##                   sapply(., "[[", 2)</span></span>
<span><span class="co">##             })</span></span>
<span><span class="co">##         names(ensembl.lst) &lt;- strsplit(names(ensembl.lst), split = "\\.") %&gt;% </span></span>
<span><span class="co">##             sapply(., "[[", 2)</span></span>
<span><span class="co">##         if (org == "mouse") {</span></span>
<span><span class="co">##             mart &lt;- biomaRt::useMart(host = "grcm39.ensembl.org", </span></span>
<span><span class="co">##                 biomart = "ENSEMBL_MART_ENSEMBL", dataset = "mmusculus_gene_ensembl")</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         else {</span></span>
<span><span class="co">##             mart &lt;- biomaRt::useMart(host = "grch38.ensembl.org", </span></span>
<span><span class="co">##                 biomart = "ENSEMBL_MART_ENSEMBL", dataset = "hsapiens_gene_symbol")</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         mart_results &lt;- biomaRt::getBM(attributes = c("ensembl_gene_id", </span></span>
<span><span class="co">##             "ensembl_peptide_id"), filters = "ensembl_peptide_id", </span></span>
<span><span class="co">##             values = protein_ids, mart = mart)</span></span>
<span><span class="co">##         ix &lt;- match(protein_ids, mart_results$ensembl_peptide_id)</span></span>
<span><span class="co">##         ix &lt;- ix[!is.na(ix)]</span></span>
<span><span class="co">##         relation.lst &lt;- pbmcapply::pbmclapply(ensembl.lst, mc.cores = parallel::detectCores(), </span></span>
<span><span class="co">##             function(protein_ids) {</span></span>
<span><span class="co">##                 newnames &lt;- protein_ids</span></span>
<span><span class="co">##                 newnames[match(mart_results[ix, "ensembl_peptide_id"], </span></span>
<span><span class="co">##                   newnames)] &lt;- mart_results[ix, "ensembl_gene_symbol"]</span></span>
<span><span class="co">##                 newnames</span></span>
<span><span class="co">##             })</span></span>
<span><span class="co">##         names(relation.lst)[match(mart_results[ix, "ensembl_peptide_id"], </span></span>
<span><span class="co">##             names(relation.lst))] &lt;- mart_results[ix, "ensembl_gene_symbol"]</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     else {</span></span>
<span><span class="co">##         message("Loaded the BioGrid database.")</span></span>
<span><span class="co">##         if (org == "mouse") {</span></span>
<span><span class="co">##             data(MouseBioGRIDInteractionOfficial)</span></span>
<span><span class="co">##             tmp.lst &lt;- MouseBioGRIDInteractionOfficial</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         else {</span></span>
<span><span class="co">##             data(HumanBioGRIDInteractionOfficial)</span></span>
<span><span class="co">##             tmp.lst &lt;- HumanBioGRIDInteractionOfficial</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         relation.lst &lt;- lapply(tmp.lst, "[[", "interactors")</span></span>
<span><span class="co">##         names(relation.lst) &lt;- sapply(tmp.lst, "[[", "name")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if (org == "mouse") {</span></span>
<span><span class="co">##         taxid &lt;- 10090</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     else {</span></span>
<span><span class="co">##         taxid &lt;- 9606</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     message("Loaded ", length(relation.lst), " protein relations from ", </span></span>
<span><span class="co">##         db, " database of ", org, ".")</span></span>
<span><span class="co">##     TFs &lt;- c(unique(unlist(relation.lst[intersect(names(relation.lst), </span></span>
<span><span class="co">##         key.TFs)])), key.TFs)</span></span>
<span><span class="co">##     message("Identified ", length(TFs), " TFs in total.")</span></span>
<span><span class="co">##     return(TFs)</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[4]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[5]]</span></span>
<span><span class="co">## [[5]]$value</span></span>
<span><span class="co">## function (pfm = NULL, peaks = NULL, org.gs = BSgenome.Hsapiens.UCSC.hg38, </span></span>
<span><span class="co">##     candidate.TFs = NULL) </span></span>
<span><span class="co">## {</span></span>
<span><span class="co">##     motif.ix &lt;- motifmatchr::motifMatches(motifmatchr::matchMotifs(pwms = pfm, </span></span>
<span><span class="co">##         subject = Signac::StringToGRanges(peaks), genome = org.gs)) %&gt;% </span></span>
<span><span class="co">##         summary</span></span>
<span><span class="co">##     TF.lst &lt;- sapply(pfm, function(x) {</span></span>
<span><span class="co">##         x@name</span></span>
<span><span class="co">##     })</span></span>
<span><span class="co">##     names(TF.lst) &lt;- NULL</span></span>
<span><span class="co">##     peak.TF.pairs &lt;- data.frame(peaks[motif.ix[, 1]], TF.lst[motif.ix[, </span></span>
<span><span class="co">##         2]])</span></span>
<span><span class="co">##     colnames(peak.TF.pairs) &lt;- c("peak", "TF")</span></span>
<span><span class="co">##     if (is.not.null(candidate.TFs)) {</span></span>
<span><span class="co">##         peak.TF.pairs &lt;- peak.TF.pairs[peak.TF.pairs$TF %in% </span></span>
<span><span class="co">##             candidate.TFs, , drop = FALSE]</span></span>
<span><span class="co">##         message("Identify putative TF binding sites via motif enrichment analysis for the TF list", </span></span>
<span><span class="co">##             " provided by the user ...")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     peak.TFs &lt;- split(peak.TF.pairs, peak.TF.pairs$peak) %&gt;% </span></span>
<span><span class="co">##         sapply(., "[[", "TF")</span></span>
<span><span class="co">##     TF.peaks &lt;- split(peak.TF.pairs, peak.TF.pairs$TF) %&gt;% sapply(., </span></span>
<span><span class="co">##         "[[", "peak")</span></span>
<span><span class="co">##     return(list(puta.bound.TFs = peak.TFs, puta.binding.peaks = TF.peaks))</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[5]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[6]]</span></span>
<span><span class="co">## [[6]]$value</span></span>
<span><span class="co">## function (regs = NULL, dbs = c("GO", "KEGG"), org = "human") </span></span>
<span><span class="co">## {</span></span>
<span><span class="co">##     genes.ll &lt;- sapply(regs, "[[", "genes")</span></span>
<span><span class="co">##     databases &lt;- c()</span></span>
<span><span class="co">##     if ("GO" %in% dbs) {</span></span>
<span><span class="co">##         message("Loading GO databases ...")</span></span>
<span><span class="co">##         databases &lt;- c("GO_Molecular_Function_2018", "GO_Cellular_Component_2018", </span></span>
<span><span class="co">##             "GO_Biological_Process_2018")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     if ("KEGG" %in% dbs) {</span></span>
<span><span class="co">##         if (org == "mouse") {</span></span>
<span><span class="co">##             databases &lt;- c(databases, "KEGG_2019_Mouse")</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         else {</span></span>
<span><span class="co">##             databases &lt;- c(databases, "KEGG_2019_Human")</span></span>
<span><span class="co">##         }</span></span>
<span><span class="co">##         message("Loading KEGG database of ", org, " ...")</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     message("Running enrichment analyses for ", length(genes.ll), </span></span>
<span><span class="co">##         " gene lists ...")</span></span>
<span><span class="co">##     quiet(require(enrichR))</span></span>
<span><span class="co">##     enriched &lt;- pbapply::pblapply(genes.ll, function(x) {</span></span>
<span><span class="co">##         enrichr(genes = x, databases = databases)</span></span>
<span><span class="co">##     })</span></span>
<span><span class="co">##     names(enriched) &lt;- seq_along(genes.ll)</span></span>
<span><span class="co">##     message("Merging enrichment analyses results ...")</span></span>
<span><span class="co">##     processed &lt;- pbmcapply::pbmclapply(databases, mc.cores = parallel::detectCores(), </span></span>
<span><span class="co">##         function(x) {</span></span>
<span><span class="co">##             message("Preparing enrichment analysis results for ", </span></span>
<span><span class="co">##                 x, " ...")</span></span>
<span><span class="co">##             x.prosessed &lt;- Reduce("rbind", lapply(seq_along(enriched), </span></span>
<span><span class="co">##                 function(i) {</span></span>
<span><span class="co">##                   cbind(Id = rep(names(enriched)[i], nrow(enriched[[i]][[x]])), </span></span>
<span><span class="co">##                     enriched[[i]][[x]])</span></span>
<span><span class="co">##                 }))</span></span>
<span><span class="co">##         })</span></span>
<span><span class="co">##     names(processed) &lt;- databases</span></span>
<span><span class="co">##     message("There are ", paste(sapply(processed, nrow), collapse = ", "), </span></span>
<span><span class="co">##         " terms/pathways for ", paste(names(processed), collapse = ", "), </span></span>
<span><span class="co">##         ".")</span></span>
<span><span class="co">##     return(processed)</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[6]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[7]]</span></span>
<span><span class="co">## [[7]]$value</span></span>
<span><span class="co">## [1] "_PACKAGE"</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[7]]$visible</span></span>
<span><span class="co">## [1] TRUE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[8]]</span></span>
<span><span class="co">## [[8]]$value</span></span>
<span><span class="co">## function (obj) </span></span>
<span><span class="co">## {</span></span>
<span><span class="co">##     ltmg.df &lt;- as.data.frame(obj@assays$RNA@data)</span></span>
<span><span class="co">##     ltmg.obj &lt;- IRISFGM::ProcessData(IRISFGM::CreateIRISFGMObject(ltmg.df), </span></span>
<span><span class="co">##         normalization = "cpm", IsImputation = F)</span></span>
<span><span class="co">##     IRISFGM::RunLTMG(ltmg.obj, Gene_use = "all")</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[8]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[9]]</span></span>
<span><span class="co">## [[9]]$value</span></span>
<span><span class="co">## function (x) </span></span>
<span><span class="co">## !is.na(x)</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[9]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[10]]</span></span>
<span><span class="co">## [[10]]$value</span></span>
<span><span class="co">## function (object = NULL, region = NULL, features = NULL, links = FALSE, </span></span>
<span><span class="co">##     ranges.group.by = "seurat_cluster", peaks = FALSE, bigwig = NULL, </span></span>
<span><span class="co">##     region.highlight = NULL, idents = NULL, extend.upstream = 0, </span></span>
<span><span class="co">##     extend.downstream = 0, colors = NULL, size = 10) </span></span>
<span><span class="co">## {</span></span>
<span><span class="co">##     Seurat::DefaultAssay(object) &lt;- peak.assay</span></span>
<span><span class="co">##     Seurat::Idents(object) &lt;- setNames(object@meta.data[, ranges.group.by], </span></span>
<span><span class="co">##         colnames(object))</span></span>
<span><span class="co">##     p &lt;- suppressWarnings(Signac::CoveragePlot(object = object, </span></span>
<span><span class="co">##         region = region, features = features, ranges.group.by = ranges.group.by, </span></span>
<span><span class="co">##         peaks = peaks, links = links, idents = idents, bigwig = bigwig, </span></span>
<span><span class="co">##         region.highlight = region.highlight))</span></span>
<span><span class="co">##     if (is.not.null(colors)) {</span></span>
<span><span class="co">##         p &lt;- p &amp; scale_fill_manual(values = colors)</span></span>
<span><span class="co">##     }</span></span>
<span><span class="co">##     return(p)</span></span>
<span><span class="co">## }</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## [[10]]$visible</span></span>
<span><span class="co">## [1] FALSE</span></span></code></pre>
<div class="section level2">
<h2 id="data-loading">Data loading<a class="anchor" aria-label="anchor" href="#data-loading"></a>
</h2>
<p>Load a Seurat object containing jointly profiled scRNA-seq and
scATAC-seq assays from human PBMCs.</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pbmc</span> <span class="op">&lt;-</span> <span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qread.html" class="external-link">qread</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">work.dir</span>, <span class="st">"tutorial_pbmc.qsave"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">pbmc</span></span></code></pre></div>
<pre><code><span><span class="co">## An object of class Seurat </span></span>
<span><span class="co">## 6193 features across 1000 samples within 3 assays </span></span>
<span><span class="co">## Active assay: peaks (3991 features, 3991 variable features)</span></span>
<span><span class="co">##  2 other assays present: RNA, SCT</span></span>
<span><span class="co">##  3 dimensional reductions calculated: pca, lsi, umap</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="eregulon-inference">eRegulon inference<a class="anchor" aria-label="anchor" href="#eregulon-inference"></a>
</h2>
<p>Run STREAM on the Seurat object with default parameters.</p>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># This process can be resource-intensive. We advise against executing it on a personal computer.</span></span>
<span><span class="va">en.regs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/run_stream.html">run_stream</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">pbmc</span>,</span>
<span>                      qubic.path <span class="op">=</span> <span class="st">"/users/PAS1475/liyang/software/QUBIC2/qubic"</span>,</span>
<span>                      peak.assay <span class="op">=</span> <span class="st">"peaks"</span>,</span>
<span>                      var.genes <span class="op">=</span> <span class="fl">3000</span>,</span>
<span>                      top.peaks <span class="op">=</span> <span class="fl">3000</span>,</span>
<span>                      min.cells <span class="op">=</span> <span class="fl">10</span>,</span>
<span>                      org <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>                      c.cutoff <span class="op">=</span> <span class="fl">1.0</span>,</span>
<span>                      distance <span class="op">=</span> <span class="fl">5e+05</span>,</span>
<span>                      BlockOverlap <span class="op">=</span> <span class="fl">0.50</span>,</span>
<span>                      Extension <span class="op">=</span> <span class="fl">1.0</span></span>
<span>                      <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Began running STREAM2 at time: </span></span>
<span><span class="co">## 2023-08-20 08:43:54</span></span></code></pre>
<pre><code><span><span class="co">## Filtered out features not a gene symbol or belonging to non-standard chromosomes,</span></span>
<span><span class="co">## leading to 1098 genes and 3991 enhancers.</span></span></code></pre>
<pre><code><span><span class="co">## Saved the Seurat object after filtering to file: ./Obj_filtered.qsave.</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: BSgenome.Hsapiens.UCSC.hg38</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: BSgenome</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: Biostrings</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: XVector</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'Biostrings'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:base':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     strsplit</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: rtracklayer</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'rtracklayer'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:igraph':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     blocks</span></span></code></pre>
<pre><code><span><span class="co">## Loaded full genome sequences for hg38</span></span></code></pre>
<pre><code><span><span class="co">## Computed the GC content, region lengths, and dinucleotide base frequencies for regions in the peaks assay and add them to the feature metadata.</span></span></code></pre>
<pre><code><span><span class="co">## Wrote potential enhancer-gene pairs to file ./Potential_enhancer_gene_pairs_within_1mb.qsave</span></span></code></pre>
<pre><code><span><span class="co">## Removed the cells with zero count of RNA or ATAC, </span></span>
<span><span class="co">## leading to 1000 cells.</span></span></code></pre>
<pre><code><span><span class="co">## Filtered the enhancers that are beyond 500 kb from any genes in the Seurat object.</span></span>
<span><span class="co">## 3991 enhancers and 1098 genes were retained.</span></span></code></pre>
<pre><code><span><span class="co">## Performing TF-IDF normalization</span></span></code></pre>
<pre><code><span><span class="co">## Warning in RunTFIDF.default(object = GetAssayData(object = object, slot =</span></span>
<span><span class="co">## "counts"), : Some features contain 0 total counts</span></span></code></pre>
<pre><code><span><span class="co">## Saved the Seurat object composed of 950 genes (950 variable genes) and 3861 enhancers (3861 top-ranked enhancers) to file: ./Obj_var_genes_top_enhs.qsave.</span></span></code></pre>
<pre><code><span><span class="co">## Loaded TF binding sites from JASPAR 2022.</span></span></code></pre>
<pre><code><span><span class="co">## Finished identified 29743 TF-enhancer pairs.</span></span></code></pre>
<pre><code><span><span class="co">## Identified 201 TFs binding 3423 enhancers,</span></span>
<span><span class="co">## which were saved to file: ./TF_binding_sites_on_enhs.qsave.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Began LTMG modeling ===================</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## Registered S3 method overwritten by 'gamlss':</span></span>
<span><span class="co">##   method   from</span></span>
<span><span class="co">##   print.ri bit</span></span></code></pre>
<pre><code><span><span class="co">## </span></span></code></pre>
<pre><code><span><span class="co">## Registered S3 methods overwritten by 'treeio':</span></span>
<span><span class="co">##   method              from    </span></span>
<span><span class="co">##   MRCA.phylo          tidytree</span></span>
<span><span class="co">##   MRCA.treedata       tidytree</span></span>
<span><span class="co">##   Nnode.treedata      tidytree</span></span>
<span><span class="co">##   Ntip.treedata       tidytree</span></span>
<span><span class="co">##   ancestor.phylo      tidytree</span></span>
<span><span class="co">##   ancestor.treedata   tidytree</span></span>
<span><span class="co">##   child.phylo         tidytree</span></span>
<span><span class="co">##   child.treedata      tidytree</span></span>
<span><span class="co">##   full_join.phylo     tidytree</span></span>
<span><span class="co">##   full_join.treedata  tidytree</span></span>
<span><span class="co">##   groupClade.phylo    tidytree</span></span>
<span><span class="co">##   groupClade.treedata tidytree</span></span>
<span><span class="co">##   groupOTU.phylo      tidytree</span></span>
<span><span class="co">##   groupOTU.treedata   tidytree</span></span>
<span><span class="co">##   is.rooted.treedata  tidytree</span></span>
<span><span class="co">##   nodeid.phylo        tidytree</span></span>
<span><span class="co">##   nodeid.treedata     tidytree</span></span>
<span><span class="co">##   nodelab.phylo       tidytree</span></span>
<span><span class="co">##   nodelab.treedata    tidytree</span></span>
<span><span class="co">##   offspring.phylo     tidytree</span></span>
<span><span class="co">##   offspring.treedata  tidytree</span></span>
<span><span class="co">##   parent.phylo        tidytree</span></span>
<span><span class="co">##   parent.treedata     tidytree</span></span>
<span><span class="co">##   root.treedata       tidytree</span></span>
<span><span class="co">##   rootnode.phylo      tidytree</span></span>
<span><span class="co">##   sibling.phylo       tidytree</span></span></code></pre>
<pre><code><span><span class="co">## Creating IRISCEM object. </span></span>
<span><span class="co">## The original input file contains 1000 cells and 950 genes </span></span>
<span><span class="co">## Removed 0 genes that total expression value is equal or less than 0</span></span>
<span><span class="co">## Removed 0 cells that number of expressed gene is equal or less than 0</span></span></code></pre>
<pre><code><span><span class="co">## number of iterations= 236</span></span></code></pre>
<pre><code><span><span class="co">## using all genes.</span></span></code></pre>
<pre><code><span><span class="co">## Progress:0%</span></span>
<span><span class="co">## Progress:10%</span></span>
<span><span class="co">## Progress:20%</span></span>
<span><span class="co">## Progress:30%</span></span>
<span><span class="co">## Progress:40%</span></span>
<span><span class="co">## Progress:50%</span></span>
<span><span class="co">## Progress:60%</span></span>
<span><span class="co">## Progress:70%</span></span>
<span><span class="co">## Progress:80%</span></span>
<span><span class="co">## Progress:90%</span></span>
<span><span class="co">## Progress:100%</span></span></code></pre>
<pre><code><span><span class="co">## End LTMG modeling ===================</span></span></code></pre>
<pre><code>## 
  |                                                                            
  |                                                                      |   0%
  |                                                                            
  |                                                                      |   1%
  |                                                                            
  |=                                                                     |   1%
  |                                                                            
  |=                                                                     |   2%
  |                                                                            
  |==                                                                    |   2%
  |                                                                            
  |==                                                                    |   3%
  |                                                                            
  |===                                                                   |   4%
  |                                                                            
  |===                                                                   |   5%
  |                                                                            
  |====                                                                  |   5%
  |                                                                            
  |====                                                                  |   6%
  |                                                                            
  |=====                                                                 |   7%
  |                                                                            
  |=====                                                                 |   8%
  |                                                                            
  |======                                                                |   8%
  |                                                                            
  |======                                                                |   9%
  |                                                                            
  |=======                                                               |   9%
  |                                                                            
  |=======                                                               |  10%
  |                                                                            
  |=======                                                               |  11%
  |                                                                            
  |========                                                              |  11%
  |                                                                            
  |========                                                              |  12%
  |                                                                            
  |=========                                                             |  12%
  |                                                                            
  |=========                                                             |  13%
  |                                                                            
  |==========                                                            |  14%
  |                                                                            
  |==========                                                            |  15%
  |                                                                            
  |===========                                                           |  15%
  |                                                                            
  |===========                                                           |  16%
  |                                                                            
  |============                                                          |  17%
  |                                                                            
  |============                                                          |  18%
  |                                                                            
  |=============                                                         |  18%
  |                                                                            
  |=============                                                         |  19%
  |                                                                            
  |==============                                                        |  19%
  |                                                                            
  |==============                                                        |  20%
  |                                                                            
  |==============                                                        |  21%
  |                                                                            
  |===============                                                       |  21%
  |                                                                            
  |===============                                                       |  22%
  |                                                                            
  |================                                                      |  22%
  |                                                                            
  |================                                                      |  23%
  |                                                                            
  |=================                                                     |  24%
  |                                                                            
  |=================                                                     |  25%
  |                                                                            
  |==================                                                    |  25%
  |                                                                            
  |==================                                                    |  26%
  |                                                                            
  |===================                                                   |  27%
  |                                                                            
  |===================                                                   |  28%
  |                                                                            
  |====================                                                  |  28%
  |                                                                            
  |====================                                                  |  29%
  |                                                                            
  |=====================                                                 |  29%
  |                                                                            
  |=====================                                                 |  30%
  |                                                                            
  |=====================                                                 |  31%
  |                                                                            
  |======================                                                |  31%
  |                                                                            
  |======================                                                |  32%
  |                                                                            
  |=======================                                               |  32%
  |                                                                            
  |=======================                                               |  33%
  |                                                                            
  |========================                                              |  34%
  |                                                                            
  |========================                                              |  35%
  |                                                                            
  |=========================                                             |  35%
  |                                                                            
  |=========================                                             |  36%
  |                                                                            
  |==========================                                            |  37%
  |                                                                            
  |==========================                                            |  38%
  |                                                                            
  |===========================                                           |  38%
  |                                                                            
  |===========================                                           |  39%
  |                                                                            
  |============================                                          |  39%
  |                                                                            
  |============================                                          |  40%
  |                                                                            
  |============================                                          |  41%
  |                                                                            
  |=============================                                         |  41%
  |                                                                            
  |=============================                                         |  42%
  |                                                                            
  |==============================                                        |  42%
  |                                                                            
  |==============================                                        |  43%
  |                                                                            
  |===============================                                       |  44%
  |                                                                            
  |===============================                                       |  45%
  |                                                                            
  |================================                                      |  45%
  |                                                                            
  |================================                                      |  46%
  |                                                                            
  |=================================                                     |  47%
  |                                                                            
  |=================================                                     |  48%
  |                                                                            
  |==================================                                    |  48%
  |                                                                            
  |==================================                                    |  49%
  |                                                                            
  |===================================                                   |  49%
  |                                                                            
  |===================================                                   |  50%
  |                                                                            
  |===================================                                   |  51%
  |                                                                            
  |====================================                                  |  51%
  |                                                                            
  |====================================                                  |  52%
  |                                                                            
  |=====================================                                 |  52%
  |                                                                            
  |=====================================                                 |  53%
  |                                                                            
  |======================================                                |  54%
  |                                                                            
  |======================================                                |  55%
  |                                                                            
  |=======================================                               |  55%
  |                                                                            
  |=======================================                               |  56%
  |                                                                            
  |========================================                              |  57%
  |                                                                            
  |========================================                              |  58%
  |                                                                            
  |=========================================                             |  58%
  |                                                                            
  |=========================================                             |  59%
  |                                                                            
  |==========================================                            |  59%
  |                                                                            
  |==========================================                            |  60%
  |                                                                            
  |==========================================                            |  61%
  |                                                                            
  |===========================================                           |  61%
  |                                                                            
  |===========================================                           |  62%
  |                                                                            
  |============================================                          |  62%
  |                                                                            
  |============================================                          |  63%
  |                                                                            
  |=============================================                         |  64%
  |                                                                            
  |=============================================                         |  65%
  |                                                                            
  |==============================================                        |  65%
  |                                                                            
  |==============================================                        |  66%
  |                                                                            
  |===============================================                       |  67%
  |                                                                            
  |===============================================                       |  68%
  |                                                                            
  |================================================                      |  68%
  |                                                                            
  |================================================                      |  69%
  |                                                                            
  |=================================================                     |  69%
  |                                                                            
  |=================================================                     |  70%
  |                                                                            
  |=================================================                     |  71%
  |                                                                            
  |==================================================                    |  71%
  |                                                                            
  |==================================================                    |  72%
  |                                                                            
  |===================================================                   |  72%
  |                                                                            
  |===================================================                   |  73%
  |                                                                            
  |====================================================                  |  74%
  |                                                                            
  |====================================================                  |  75%
  |                                                                            
  |=====================================================                 |  75%
  |                                                                            
  |=====================================================                 |  76%
  |                                                                            
  |======================================================                |  77%
  |                                                                            
  |======================================================                |  78%
  |                                                                            
  |=======================================================               |  78%
  |                                                                            
  |=======================================================               |  79%
  |                                                                            
  |========================================================              |  79%
  |                                                                            
  |========================================================              |  80%
  |                                                                            
  |========================================================              |  81%
  |                                                                            
  |=========================================================             |  81%
  |                                                                            
  |=========================================================             |  82%
  |                                                                            
  |==========================================================            |  82%
  |                                                                            
  |==========================================================            |  83%
  |                                                                            
  |===========================================================           |  84%
  |                                                                            
  |===========================================================           |  85%
  |                                                                            
  |============================================================          |  85%
  |                                                                            
  |============================================================          |  86%
  |                                                                            
  |=============================================================         |  87%
  |                                                                            
  |=============================================================         |  88%
  |                                                                            
  |==============================================================        |  88%
  |                                                                            
  |==============================================================        |  89%
  |                                                                            
  |===============================================================       |  89%
  |                                                                            
  |===============================================================       |  90%
  |                                                                            
  |===============================================================       |  91%
  |                                                                            
  |================================================================      |  91%
  |                                                                            
  |================================================================      |  92%
  |                                                                            
  |=================================================================     |  92%
  |                                                                            
  |=================================================================     |  93%
  |                                                                            
  |==================================================================    |  94%
  |                                                                            
  |==================================================================    |  95%
  |                                                                            
  |===================================================================   |  95%
  |                                                                            
  |===================================================================   |  96%
  |                                                                            
  |====================================================================  |  97%
  |                                                                            
  |====================================================================  |  98%
  |                                                                            
  |===================================================================== |  98%
  |                                                                            
  |===================================================================== |  99%
  |                                                                            
  |======================================================================|  99%
  |                                                                            
  |======================================================================| 100%</code></pre>
<pre><code><span><span class="co">## Finished LTMG modeling for 950 genes across 1000 cells.</span></span>
<span><span class="co">## There are 2 transcriptional regulatory states (TRSs) identified.</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Began QUBIC biclustering ===================</span></span></code></pre>
<pre><code><span><span class="co">## End QUBIC biclustering ===================</span></span></code></pre>
<pre><code><span><span class="co">## Read co-conditioned cells (n = 976) in 1 biclusters</span></span></code></pre>
<pre><code><span><span class="co">## Read co-regulated genes (n = 29) in 1 biclusters</span></span></code></pre>
<pre><code><span><span class="co">## Testing 801 genes and 2308 peaks</span></span></code></pre>
<pre><code><span><span class="co">## Registered S3 method overwritten by 'SeuratDisk':</span></span>
<span><span class="co">##   method            from  </span></span>
<span><span class="co">##   as.sparse.H5Group Seurat</span></span></code></pre>
<pre><code><span><span class="co">## 'as(&lt;dgCMatrix&gt;, "dgTMatrix")' is deprecated.</span></span>
<span><span class="co">## Use 'as(., "TsparseMatrix")' instead.</span></span>
<span><span class="co">## See help("Deprecated") and help("Matrix-deprecated").</span></span></code></pre>
<pre><code><span><span class="co">## Built 2856 enhancer-gene relations from the total dataset.</span></span>
<span><span class="co">## Saved Signac enhancer-gene links to file: ./Signac_links.qsave</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: monocle3</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: SingleCellExperiment</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: SummarizedExperiment</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: MatrixGenerics</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: matrixStats</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'matrixStats'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:dplyr':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     count</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:Biobase':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     anyMissing, rowMedians</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'MatrixGenerics'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:matrixStats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,</span></span>
<span><span class="co">##     colCounts, colCummaxs, colCummins, colCumprods, colCumsums,</span></span>
<span><span class="co">##     colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,</span></span>
<span><span class="co">##     colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,</span></span>
<span><span class="co">##     colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,</span></span>
<span><span class="co">##     colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,</span></span>
<span><span class="co">##     colWeightedMeans, colWeightedMedians, colWeightedSds,</span></span>
<span><span class="co">##     colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,</span></span>
<span><span class="co">##     rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,</span></span>
<span><span class="co">##     rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,</span></span>
<span><span class="co">##     rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,</span></span>
<span><span class="co">##     rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,</span></span>
<span><span class="co">##     rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,</span></span>
<span><span class="co">##     rowWeightedMads, rowWeightedMeans, rowWeightedMedians,</span></span>
<span><span class="co">##     rowWeightedSds, rowWeightedVars</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:Biobase':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     rowMedians</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'SummarizedExperiment'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:SeuratObject':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Assays</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:Seurat':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Assays</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'monocle3'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:igraph':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     clusters</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:Biobase':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     exprs, fData, fData&lt;-, pData, pData&lt;-</span></span></code></pre>
<pre><code><span><span class="co">## Overlap QC metrics:</span></span>
<span><span class="co">## Cells per bin: 50</span></span>
<span><span class="co">## Maximum shared cells bin-bin: 44</span></span>
<span><span class="co">## Mean shared cells bin-bin: 2.74111218257795</span></span>
<span><span class="co">## Median shared cells bin-bin: 0</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Starting Cicero"</span></span>
<span><span class="co">## [1] "Calculating distance_parameter value"</span></span></code></pre>
<pre><code><span><span class="co">## Warning in estimate_distance_parameter(cds, window = window, maxit = 100, :</span></span>
<span><span class="co">## Could not calculate sample_num distance_parameters (26 were calculated) - see</span></span>
<span><span class="co">## documentation details</span></span></code></pre>
<pre><code><span><span class="co">## [1] "Running models"</span></span>
<span><span class="co">## [1] "Assembling connections"</span></span>
<span><span class="co">## [1] "Successful cicero models:  1469"</span></span>
<span><span class="co">## [1] "Other models: "</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Zero or one element in range </span></span>
<span><span class="co">##                        12116 </span></span>
<span><span class="co">## [1] "Models with errors:  0"</span></span>
<span><span class="co">## [1] "Done"</span></span></code></pre>
<pre><code><span><span class="co">## Built 14914 co-accessible enhancer-enhancer links.</span></span>
<span><span class="co">## Saved the Cicero enhancer-enhancer links to file: ./Cicero_links.qsave</span></span></code></pre>
<pre><code><span><span class="co">## List containing 1 Seurat objects were saved to file: ./Obj_list.qsave.</span></span></code></pre>
<pre><code><span><span class="co">## Constructed 1 heterogeneous graphs,</span></span>
<span><span class="co">## which were saved to file: ./Graph_list.qsave.</span></span></code></pre>
<pre><code><span><span class="co">## Saved the TF-target pairs to file: ./TF_target_pairs.qsave.</span></span></code></pre>
<pre><code><span><span class="co">## Identified 12 enhancer-gene relations using SFP model.</span></span></code></pre>
<pre><code><span><span class="co">## 60 seeds are identified for hybrid biclustering,</span></span>
<span><span class="co">## which were saved to file: ./Seeds.qsave.</span></span></code></pre>
<pre><code><span><span class="co">## Biclustering based on Tier-1 TF-enhancer binding annotations ...</span></span></code></pre>
<pre><code><span><span class="co">## Hybrid biclustering on the transcriptome and chromatin accessibility matrices ...</span></span></code></pre>
<pre><code>## 
  |                                                        
  |                                                  |   0%
  |                                                        
  |=                                                 |   2%
  |                                                        
  |==                                                |   3%
  |                                                        
  |==                                                |   5%
  |                                                        
  |===                                               |   7%
  |                                                        
  |====                                              |   8%
  |                                                        
  |=====                                             |  10%
  |                                                        
  |======                                            |  12%
  |                                                        
  |=======                                           |  13%
  |                                                        
  |========                                          |  15%
  |                                                        
  |========                                          |  17%
  |                                                        
  |=========                                         |  18%
  |                                                        
  |==========                                        |  20%
  |                                                        
  |===========                                       |  22%
  |                                                        
  |============                                      |  23%
  |                                                        
  |============                                      |  25%
  |                                                        
  |=============                                     |  27%
  |                                                        
  |==============                                    |  28%
  |                                                        
  |===============                                   |  30%
  |                                                        
  |================                                  |  32%
  |                                                        
  |=================                                 |  33%
  |                                                        
  |==================                                |  35%
  |                                                        
  |==================                                |  37%
  |                                                        
  |===================                               |  38%
  |                                                        
  |====================                              |  40%
  |                                                        
  |=====================                             |  42%
  |                                                        
  |======================                            |  43%
  |                                                        
  |======================                            |  45%
  |                                                        
  |=======================                           |  47%
  |                                                        
  |========================                          |  48%
  |                                                        
  |=========================                         |  50%
  |                                                        
  |==========================                        |  52%
  |                                                        
  |===========================                       |  53%
  |                                                        
  |============================                      |  55%
  |                                                        
  |============================                      |  57%
  |                                                        
  |=============================                     |  58%
  |                                                        
  |==============================                    |  60%
  |                                                        
  |===============================                   |  62%
  |                                                        
  |================================                  |  63%
  |                                                        
  |================================                  |  65%
  |                                                        
  |=================================                 |  67%
  |                                                        
  |==================================                |  68%
  |                                                        
  |===================================               |  70%
  |                                                        
  |====================================              |  72%
  |                                                        
  |=====================================             |  73%
  |                                                        
  |======================================            |  75%
  |                                                        
  |======================================            |  77%
  |                                                        
  |=======================================           |  78%
  |                                                        
  |========================================          |  80%
  |                                                        
  |=========================================         |  82%
  |                                                        
  |==========================================        |  83%
  |                                                        
  |==========================================        |  85%
  |                                                        
  |===========================================       |  87%
  |                                                        
  |============================================      |  88%
  |                                                        
  |=============================================     |  90%
  |                                                        
  |==============================================    |  92%
  |                                                        
  |===============================================   |  93%
  |                                                        
  |================================================  |  95%
  |                                                        
  |================================================  |  97%
  |                                                        
  |================================================= |  98%
  |                                                        
  |==================================================| 100%</code></pre>
<pre><code><span><span class="co">## Identified 9 hybrid biclusters (HBCs),</span></span>
<span><span class="co">## which were saved to file: ./HBCs.qsave.</span></span>
<span><span class="co">## HBCs contain 20-98 genes, 17-76 enhancers, 466-822 cells, and 9 TFs.</span></span></code></pre>
<pre><code><span><span class="co">## 9 enhancer regulons (eRegulons) were discovered</span></span></code></pre>
<div class="sourceCode" id="cb159"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qsave.html" class="external-link">qsave</a></span><span class="op">(</span><span class="va">en.regs</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">work.dir</span>, <span class="st">"eRegulons.qsave"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "terminal"   "Tier"       "TF"         "genes"      "peaks"     </span></span>
<span><span class="co">##  [6] "cells"      "atac.ratio" "score"      "weight"     "links"     </span></span>
<span><span class="co">## [11] "seed"</span></span></code></pre>
<div class="sourceCode" id="cb161"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "MAN1C1" "STMN1"  "EIF2B3" "ZSWIM5" "JUN"    "EVI5"</span></span></code></pre>
<div class="sourceCode" id="cb163"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">peaks</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "chr1-25904959-25907184"   "chr1-45302959-45304454"  </span></span>
<span><span class="co">## [3] "chr1-58783212-58785307"   "chr1-93179195-93181524"  </span></span>
<span><span class="co">## [5] "chr1-108660816-108661818" "chr1-150578547-150580018"</span></span></code></pre>
<div class="sourceCode" id="cb165"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">links</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 1 metadata column:</span></span>
<span><span class="co">##       seqnames            ranges strand |        gene</span></span>
<span><span class="co">##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1 25904959-25907184      * |      MAN1C1</span></span>
<span><span class="co">##   [2]     chr1 25904959-25907184      * |       STMN1</span></span>
<span><span class="co">##   [3]     chr1 45302959-45304454      * |      EIF2B3</span></span>
<span><span class="co">##   [4]     chr1 45302959-45304454      * |      ZSWIM5</span></span>
<span><span class="co">##   [5]     chr1 58783212-58785307      * |         JUN</span></span>
<span><span class="co">##   [6]     chr1 93179195-93181524      * |        EVI5</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 19 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="egrn-construction">eGRN construction<a class="anchor" aria-label="anchor" href="#egrn-construction"></a>
</h2>
<p>Construct eGRN in each cell type.</p>
<div class="sourceCode" id="cb167"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">en.grns</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_cts_en_GRNs.html">get_cts_en_GRNs</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">pbmc</span>, celltype <span class="op">=</span> <span class="st">"predicted.id"</span>,</span>
<span>                           en.regs <span class="op">=</span> <span class="va">en.regs</span>, peak.assay <span class="op">=</span> <span class="st">"peaks"</span>,</span>
<span>                           rna.dims <span class="op">=</span> <span class="fl">50</span>, atac.dims <span class="op">=</span> <span class="fl">50</span>,</span>
<span>                           padj.cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                           out.dir <span class="op">=</span> <span class="st">"./"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The Seurat object contains 1101 genes, 3991 enhancers, and 1000 cells.</span></span></code></pre>
<pre><code><span><span class="co">## Finished hyper-geometric test of 9 eRegulons against 3 cell types, obtaining p-values ranging from 3.48064426106208e-09 to 0.999997707597157</span></span></code></pre>
<pre><code><span><span class="co">## Finished Bonferroni correction and obtained adjusted p-values ranging from 9.39773950486761e-08 to 1</span></span></code></pre>
<pre><code><span><span class="co">## Identified enhancer gene regulatory networks (eGRNs) for cell types (n = 1) : CD4 TCM</span></span></code></pre>
<div class="sourceCode" id="cb172"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qsave.html" class="external-link">qsave</a></span><span class="op">(</span><span class="va">en.grns</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">work.dir</span>, <span class="st">"eGRNs.qsave"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">en.grns</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "links"     "cell.type" "cells"</span></span></code></pre>
<div class="sourceCode" id="cb174"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">en.grns</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">links</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames            ranges strand |        gene          TF</span></span>
<span><span class="co">##          &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1 25904959-25907184      * |      MAN1C1       TCF12</span></span>
<span><span class="co">##   [2]     chr1 25904959-25907184      * |       STMN1       TCF12</span></span>
<span><span class="co">##   [3]     chr1 45302959-45304454      * |      EIF2B3       TCF12</span></span>
<span><span class="co">##   [4]     chr1 45302959-45304454      * |      ZSWIM5       TCF12</span></span>
<span><span class="co">##   [5]     chr1 58783212-58785307      * |         JUN       TCF12</span></span>
<span><span class="co">##   [6]     chr1 93179195-93181524      * |        EVI5       TCF12</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 21 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="cell-type-specific-eregulon-prediction">Cell-type-specific eRegulon prediction<a class="anchor" aria-label="anchor" href="#cell-type-specific-eregulon-prediction"></a>
</h2>
<p>Construct cell-type-specific eRegulons.</p>
<div class="sourceCode" id="cb176"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cts.en.regs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_cts_en_regs.html">get_cts_en_regs</a></span><span class="op">(</span>obj <span class="op">=</span> <span class="va">pbmc</span>, peak.assay <span class="op">=</span> <span class="st">"peaks"</span>, de.genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                               cts.en.grns <span class="op">=</span> <span class="va">en.grns</span>, out.dir <span class="op">=</span> <span class="st">"./"</span>, celltype <span class="op">=</span> <span class="st">"predicted.id"</span>,</span>
<span>                               min.pct <span class="op">=</span> <span class="fl">0.25</span>, logfc.threshold <span class="op">=</span> <span class="fl">0.25</span>, padj.cutoff <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## The Seurat object contains 1101 genes, 3991 enhancers, and 1000 cells.</span></span></code></pre>
<pre><code><span><span class="co">## Need to filter genes based on differential expression.</span></span>
<span><span class="co">## Predicting differentially expressed genes (DEGs) ...</span></span></code></pre>
<pre><code><span><span class="co">## Warning in supportsMulticoreAndRStudio(...): [ONE-TIME WARNING] Forked</span></span>
<span><span class="co">## processing ('multicore') is not supported when running R from RStudio because</span></span>
<span><span class="co">## it is considered unstable. For more details, how to control forked processing</span></span>
<span><span class="co">## or not, and how to silence this warning in future R sessions, see</span></span>
<span><span class="co">## ?parallelly::supportsMulticore</span></span></code></pre>
<pre><code><span><span class="co">## Calculating cluster CD4 TCM</span></span></code></pre>
<pre><code><span><span class="co">## Calculating cluster CD8 Naive</span></span></code></pre>
<pre><code><span><span class="co">## Calculating cluster CD4 Naive</span></span></code></pre>
<pre><code><span><span class="co">## We have 166 differentially expressed genes (DEGs).</span></span></code></pre>
<pre><code><span><span class="co">## Identified 5 cell-type-specific eRegulons.</span></span></code></pre>
<div class="sourceCode" id="cb185"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qsave.html" class="external-link">qsave</a></span><span class="op">(</span><span class="va">cts.en.regs</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">work.dir</span>, <span class="st">"cell-type-specific-eRegulons.qsave"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cts.en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "TF"        "celltype"  "cells"     "genes"     "enhancers" "links"</span></span></code></pre>
<div class="sourceCode" id="cb187"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cts.en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">genes</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "ADAMTSL4-AS1" "TSPAN5"       "ADAM19"       "ATXN1"        "EEF1A1"      </span></span>
<span><span class="co">## [6] "TNFAIP3"</span></span></code></pre>
<div class="sourceCode" id="cb189"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cts.en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">enhancers</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "chr1-150320808-150322695" "chr4-98928092-98930218"  </span></span>
<span><span class="co">## [3] "chr5-157252607-157253720" "chr6-16759215-16763347"  </span></span>
<span><span class="co">## [5] "chr6-73653070-73655111"   "chr6-137707572-137708859"</span></span></code></pre>
<div class="sourceCode" id="cb191"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">cts.en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">links</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 2 metadata columns:</span></span>
<span><span class="co">##       seqnames              ranges strand |         gene          TF</span></span>
<span><span class="co">##          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; |  &lt;character&gt; &lt;character&gt;</span></span>
<span><span class="co">##   [1]     chr1 150320808-150322695      * | ADAMTSL4-AS1       CEBPD</span></span>
<span><span class="co">##   [2]     chr4   98928092-98930218      * |       TSPAN5       CEBPD</span></span>
<span><span class="co">##   [3]     chr5 157252607-157253720      * |       ADAM19       CEBPD</span></span>
<span><span class="co">##   [4]     chr6   16759215-16763347      * |        ATXN1       CEBPD</span></span>
<span><span class="co">##   [5]     chr6   73653070-73655111      * |       EEF1A1       CEBPD</span></span>
<span><span class="co">##   [6]     chr6 137707572-137708859      * |      TNFAIP3       CEBPD</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 21 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="evaluation-of-cell-type-specific-eregulons">Evaluation of cell-type-specific eRegulons<a class="anchor" aria-label="anchor" href="#evaluation-of-cell-type-specific-eregulons"></a>
</h2>
<p>Firstly, we performed pathway enrichment analysis for gene lists in
cell-type-specific eRegulons against Gene Ontology (GO) or KEGG.</p>
<div class="sourceCode" id="cb193"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pathways</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/enrich_genes.html">enrich_genes</a></span><span class="op">(</span>regs <span class="op">=</span> <span class="va">cts.en.regs</span>, dbs <span class="op">=</span> <span class="st">"GO"</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading GO databases ...</span></span></code></pre>
<pre><code><span><span class="co">## Running enrichment analyses for 5 gene lists ...</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: enrichR</span></span></code></pre>
<pre><code><span><span class="co">## Welcome to enrichR</span></span>
<span><span class="co">## Checking connection ...</span></span></code></pre>
<pre><code><span><span class="co">## Enrichr ... Connection is Live!</span></span>
<span><span class="co">## FlyEnrichr ... Connection is Live!</span></span>
<span><span class="co">## WormEnrichr ... Connection is Live!</span></span>
<span><span class="co">## YeastEnrichr ... Connection is Live!</span></span>
<span><span class="co">## FishEnrichr ... Connection is Live!</span></span>
<span><span class="co">## OxEnrichr ... Connection is Live!</span></span></code></pre>
<pre><code><span><span class="co">## Uploading data to Enrichr... Done.</span></span>
<span><span class="co">##   Querying GO_Molecular_Function_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Cellular_Component_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Biological_Process_2018... Done.</span></span>
<span><span class="co">## Parsing results... Done.</span></span>
<span><span class="co">## Uploading data to Enrichr... Done.</span></span>
<span><span class="co">##   Querying GO_Molecular_Function_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Cellular_Component_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Biological_Process_2018... Done.</span></span>
<span><span class="co">## Parsing results... Done.</span></span>
<span><span class="co">## Uploading data to Enrichr... Done.</span></span>
<span><span class="co">##   Querying GO_Molecular_Function_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Cellular_Component_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Biological_Process_2018... Done.</span></span>
<span><span class="co">## Parsing results... Done.</span></span>
<span><span class="co">## Uploading data to Enrichr... Done.</span></span>
<span><span class="co">##   Querying GO_Molecular_Function_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Cellular_Component_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Biological_Process_2018... Done.</span></span>
<span><span class="co">## Parsing results... Done.</span></span>
<span><span class="co">## Uploading data to Enrichr... Done.</span></span>
<span><span class="co">##   Querying GO_Molecular_Function_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Cellular_Component_2018... Done.</span></span>
<span><span class="co">##   Querying GO_Biological_Process_2018... Done.</span></span>
<span><span class="co">## Parsing results... Done.</span></span></code></pre>
<pre><code><span><span class="co">## Merging enrichment analyses results ...</span></span>
<span><span class="co">## There are 316, 167, 1263 terms/pathways for GO_Molecular_Function_2018, GO_Cellular_Component_2018, GO_Biological_Process_2018.</span></span></code></pre>
<div class="sourceCode" id="cb201"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sig.pathways</span> <span class="op">&lt;-</span> <span class="fu">pblapply</span><span class="op">(</span><span class="va">pathways</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">x</span><span class="op">[</span><span class="va">x</span><span class="op">$</span><span class="va">Adjusted.P.value</span> <span class="op">&lt;</span> <span class="fl">0.05</span>,, drop <span class="op">=</span> <span class="cn">FALSE</span><span class="op">]</span></span>
<span><span class="op">}</span><span class="op">)</span></span>
<span><span class="fu">qs</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/qs/man/qsave.html" class="external-link">qsave</a></span><span class="op">(</span><span class="va">sig.pathways</span>, <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">work.dir</span>, <span class="st">"enriched_pathways.qsave"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">sig.pathways</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   Id                                                Term Overlap     P.value</span></span>
<span><span class="co">## 1  1                        poly(G) binding (GO:0034046)     1/7 0.003495224</span></span>
<span><span class="co">## 2  1       phospholipase inhibitor activity (GO:0004859)     1/8 0.003993646</span></span>
<span><span class="co">## 3  1                    Tat protein binding (GO:0030957)    1/10 0.004989818</span></span>
<span><span class="co">## 4  1 Lys63-specific deubiquitinase activity (GO:0061578)    1/11 0.005487568</span></span>
<span><span class="co">## 5  1          nitric-oxide synthase binding (GO:0050998)    1/11 0.005487568</span></span>
<span><span class="co">## 6  1 translation elongation factor activity (GO:0003746)    1/13 0.006482395</span></span>
<span><span class="co">##   Adjusted.P.value Old.P.value Old.Adjusted.P.value Odds.Ratio Combined.Score</span></span>
<span><span class="co">## 1       0.04006892           0                    0   370.0741      2093.2713</span></span>
<span><span class="co">## 2       0.04006892           0                    0   317.1905      1751.8591</span></span>
<span><span class="co">## 3       0.04006892           0                    0   246.6790      1307.4865</span></span>
<span><span class="co">## 4       0.04006892           0                    0   222.0000      1155.5700</span></span>
<span><span class="co">## 5       0.04006892           0                    0   222.0000      1155.5700</span></span>
<span><span class="co">## 6       0.04006892           0                    0   184.9815       932.0597</span></span>
<span><span class="co">##     Genes</span></span>
<span><span class="co">## 1   ATXN1</span></span>
<span><span class="co">## 2   ANXA1</span></span>
<span><span class="co">## 3    ACTB</span></span>
<span><span class="co">## 4 TNFAIP3</span></span>
<span><span class="co">## 5    ACTB</span></span>
<span><span class="co">## 6  EEF1A1</span></span></code></pre>
<p>Then, we intersected the enhancer list of a cell-type-specific
eRegulon against H3K27ac ChIP-seq peaks in CD4+ T cells with ENCODE
accession code: [ENCFF498QFU] (<a href="https://www.encodeproject.org/files/ENCFF498QFU/" class="external-link uri">https://www.encodeproject.org/files/ENCFF498QFU/</a>)</p>
<div class="sourceCode" id="cb203"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">chipseq.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">work.dir</span>, <span class="st">"ENCFF498QFU.bed"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">V1</span>, <span class="va">V2</span>, <span class="va">V3</span><span class="op">)</span></span>
<span><span class="va">chipseq.peaks</span> <span class="op">&lt;-</span> <span class="fu">StringToGRanges</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">chipseq.df</span><span class="op">[</span>, <span class="fl">1</span><span class="op">]</span>, <span class="st">"-"</span>, <span class="va">chipseq.df</span><span class="op">[</span>, <span class="fl">2</span><span class="op">]</span>, <span class="st">"-"</span>, <span class="va">chipseq.df</span><span class="op">[</span>, <span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">permTest.res</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/intersect_peaks.html">intersect_peaks</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu">StringToGRanges</span><span class="op">(</span><span class="va">cts.en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">enhancers</span><span class="op">)</span>, </span>
<span>                                y <span class="op">=</span> <span class="va">chipseq.peaks</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Perform significance test between two GRange objects composed of 10 and 40355 peaks.</span></span></code></pre>
<pre><code><span><span class="co">## 'getOption("repos")' replaces Bioconductor standard repositories, see</span></span>
<span><span class="co">## 'help("repositories", package = "BiocManager")' for details.</span></span>
<span><span class="co">## Replacement repositories:</span></span>
<span><span class="co">##     CRAN: https://cran.rstudio.com/</span></span></code></pre>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'BSgenome.Hsapiens.UCSC.hg19'</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:BSgenome.Hsapiens.UCSC.hg38':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     Hsapiens</span></span></code></pre>
<pre><code><span><span class="co">## The masked version of 'hg19' is not installed. Using the unmasked version. This means that no automatic masking will be available.</span></span></code></pre>
<div class="sourceCode" id="cb209"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">permTest.res</span><span class="op">$</span><span class="va">numOverlaps</span></span></code></pre></div>
<pre><code><span><span class="co">## P-value: 0.0099009900990099</span></span>
<span><span class="co">## Z-score: 14.9248</span></span>
<span><span class="co">## Number of iterations: 100</span></span>
<span><span class="co">## Alternative: greater</span></span>
<span><span class="co">## Evaluation of the original region set: 11</span></span>
<span><span class="co">## Evaluation function: numOverlaps</span></span>
<span><span class="co">## Randomization function: randomizeRegions</span></span></code></pre>
<p>Finally, we overlapped the enhancer-gene relations in a
cell-type-specific eRegulon against enhancer-target pairs in CD4 TCM
cells curated by [scEnhancer] (<a href="http://enhanceratlas.net/scenhancer/data/download/interaction/hs/Memory_CD4_T_interaction.txt" class="external-link uri">http://enhanceratlas.net/scenhancer/data/download/interaction/hs/Memory_CD4_T_interaction.txt</a>).</p>
<div class="sourceCode" id="cb211"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Preprocess the peak-gene interaction pairs in scEnhancer database</span></span>
<span><span class="va">enhGene.df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">work.dir</span>, <span class="st">"Memory_CD4_T_interaction.txt"</span><span class="op">)</span>, sep <span class="op">=</span> <span class="st">"\t"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">V1</span>, <span class="va">V2</span><span class="op">)</span></span>
<span><span class="va">enhGene.pairs</span> <span class="op">&lt;-</span> <span class="fu">StringToGRanges</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">enhGene.df</span><span class="op">$</span><span class="va">V1</span>, split <span class="op">=</span> <span class="st">"\\|"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">.</span>, <span class="st">"[["</span>, <span class="fl">1</span><span class="op">)</span>, sep <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">":"</span>, <span class="st">"-"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">mcols</span><span class="op">(</span><span class="va">enhGene.pairs</span><span class="op">)</span><span class="op">$</span><span class="va">gene</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="va">enhGene.df</span><span class="op">$</span><span class="va">V2</span>, split <span class="op">=</span> <span class="st">"[\\||\\:]"</span><span class="op">)</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">.</span>, <span class="st">"[["</span>, <span class="fl">3</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">enhGene.pairs</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## GRanges object with 6 ranges and 1 metadata column:</span></span>
<span><span class="co">##       seqnames              ranges strand |        gene</span></span>
<span><span class="co">##          &lt;Rle&gt;           &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;</span></span>
<span><span class="co">##   [1]    chr10 100227139-100227639      * |        HPS1</span></span>
<span><span class="co">##   [2]    chr10 100227139-100227639      * |        HPS1</span></span>
<span><span class="co">##   [3]    chr10 100227139-100227639      * |        HPS1</span></span>
<span><span class="co">##   [4]    chr10 100227139-100227639      * |        HPS1</span></span>
<span><span class="co">##   [5]    chr10 101629999-101630499      * |       DNMBP</span></span>
<span><span class="co">##   [6]    chr10 101629999-101630499      * |       DNMBP</span></span>
<span><span class="co">##   -------</span></span>
<span><span class="co">##   seqinfo: 23 sequences from an unspecified genome; no seqlengths</span></span></code></pre>
<div class="sourceCode" id="cb213"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate the overlaps between enhancer-gene relations in a cell-type-specific eRegulon</span></span>
<span><span class="va">overlaps</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/intersect_enhancer_gene_relations.html">intersect_enhancer_gene_relations</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu">resize</span><span class="op">(</span><span class="va">cts.en.regs</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">links</span>, width <span class="op">=</span> <span class="fl">50000</span>, fix <span class="op">=</span> <span class="st">"center"</span><span class="op">)</span>, </span>
<span>                                              y <span class="op">=</span> <span class="fu">resize</span><span class="op">(</span><span class="va">enhGene.pairs</span>, width <span class="op">=</span> <span class="fl">50000</span>, fix <span class="op">=</span> <span class="st">"center"</span><span class="op">)</span> <span class="op">)</span></span>
<span><span class="va">overlaps</span></span></code></pre></div>
<pre><code><span><span class="co">##                    x.peak                 y.peak  gene</span></span>
<span><span class="co">## 1: chr6-16736281-16786280 chr6-16686955-16736954 ATXN1</span></span>
<span><span class="co">## 2: chr6-16736281-16786280 chr6-16744524-16794523 ATXN1</span></span>
<span><span class="co">## 3: chr6-16736281-16786280 chr6-16746364-16796363 ATXN1</span></span>
<span><span class="co">## 4:   chr7-5372117-5422116   chr7-5412511-5462510  ACTB</span></span></code></pre>
<div class="sourceCode" id="cb215"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Check session information</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.2.1 (2022-06-23)</span></span>
<span><span class="co">## Platform: x86_64-pc-linux-gnu (64-bit)</span></span>
<span><span class="co">## Running under: Red Hat Enterprise Linux</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS/LAPACK: /opt/intel/2021.3/mkl/2021.3.0/lib/intel64/libmkl_gf_lp64.so.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              </span></span>
<span><span class="co">##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    </span></span>
<span><span class="co">##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   </span></span>
<span><span class="co">##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 </span></span>
<span><span class="co">##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            </span></span>
<span><span class="co">## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats4    parallel  stats     graphics  grDevices utils     datasets </span></span>
<span><span class="co">## [8] methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">##  [1] BSgenome.Hsapiens.UCSC.hg19_1.4.3 enrichR_3.2                      </span></span>
<span><span class="co">##  [3] monocle3_1.3.1                    SingleCellExperiment_1.20.1      </span></span>
<span><span class="co">##  [5] SummarizedExperiment_1.28.0       MatrixGenerics_1.10.0            </span></span>
<span><span class="co">##  [7] matrixStats_1.0.0                 BSgenome.Hsapiens.UCSC.hg38_1.4.5</span></span>
<span><span class="co">##  [9] BSgenome_1.66.3                   rtracklayer_1.58.0               </span></span>
<span><span class="co">## [11] Biostrings_2.66.0                 XVector_0.38.0                   </span></span>
<span><span class="co">## [13] stream2_2.0.0                     Matrix_1.6-0                     </span></span>
<span><span class="co">## [15] igraph_1.5.0                      ggpubr_0.6.0                     </span></span>
<span><span class="co">## [17] ggplot2_3.4.2                     dplyr_1.1.2                      </span></span>
<span><span class="co">## [19] EnsDb.Hsapiens.v86_2.99.0         EnsDb.Mmusculus.v79_2.99.0       </span></span>
<span><span class="co">## [21] ensembldb_2.22.0                  AnnotationFilter_1.22.0          </span></span>
<span><span class="co">## [23] GenomicFeatures_1.50.4            AnnotationDbi_1.60.2             </span></span>
<span><span class="co">## [25] Biobase_2.58.0                    GenomicRanges_1.50.2             </span></span>
<span><span class="co">## [27] GenomeInfoDb_1.34.9               IRanges_2.32.0                   </span></span>
<span><span class="co">## [29] S4Vectors_0.36.2                  BiocGenerics_0.44.0              </span></span>
<span><span class="co">## [31] pbapply_1.7-2                     pbmcapply_1.5.1                  </span></span>
<span><span class="co">## [33] Signac_1.10.0                     SeuratObject_4.1.3               </span></span>
<span><span class="co">## [35] Seurat_4.3.0.1                    hdf5r_1.3.8                      </span></span>
<span><span class="co">## [37] qs_0.25.5                         easypackages_0.1.0               </span></span>
<span><span class="co">## [39] knitr_1.43                       </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] mixtools_2.0.0            terra_1.7-39             </span></span>
<span><span class="co">##   [3] graphlayouts_1.0.0        IRISFGM_1.6.0            </span></span>
<span><span class="co">##   [5] lattice_0.20-45           vctrs_0.6.3              </span></span>
<span><span class="co">##   [7] expm_0.999-7              usethis_2.2.2            </span></span>
<span><span class="co">##   [9] blob_1.2.4                survival_3.3-1           </span></span>
<span><span class="co">##  [11] spatstat.data_3.0-1       later_1.3.1              </span></span>
<span><span class="co">##  [13] nloptr_2.0.3              DBI_1.1.3                </span></span>
<span><span class="co">##  [15] rappdirs_0.3.3            uwot_0.1.16              </span></span>
<span><span class="co">##  [17] dqrng_0.3.0               SeuratDisk_0.0.0.9020    </span></span>
<span><span class="co">##  [19] zlibbioc_1.44.0           htmlwidgets_1.6.2        </span></span>
<span><span class="co">##  [21] mvtnorm_1.2-2             future_1.33.0            </span></span>
<span><span class="co">##  [23] miscTools_0.6-28          leiden_0.4.3             </span></span>
<span><span class="co">##  [25] scater_1.26.1             irlba_2.3.5.1            </span></span>
<span><span class="co">##  [27] tidygraph_1.2.3           Rcpp_1.0.11              </span></span>
<span><span class="co">##  [29] KernSmooth_2.23-20        promises_1.2.1           </span></span>
<span><span class="co">##  [31] DelayedArray_0.24.0       limma_3.54.2             </span></span>
<span><span class="co">##  [33] pkgload_1.3.2.1           RcppParallel_5.1.7       </span></span>
<span><span class="co">##  [35] fs_1.6.3                  textshaping_0.3.6        </span></span>
<span><span class="co">##  [37] fastmatch_1.1-3           digest_0.6.33            </span></span>
<span><span class="co">##  [39] png_0.1-8                 bluster_1.8.0            </span></span>
<span><span class="co">##  [41] qlcMatrix_0.9.7           sctransform_0.3.5        </span></span>
<span><span class="co">##  [43] scatterpie_0.2.1          cowplot_1.1.1            </span></span>
<span><span class="co">##  [45] DOSE_3.24.2               ggraph_2.1.0             </span></span>
<span><span class="co">##  [47] pkgconfig_2.0.3           GO.db_3.16.0             </span></span>
<span><span class="co">##  [49] docopt_0.7.1              spatstat.random_3.1-5    </span></span>
<span><span class="co">##  [51] DelayedMatrixStats_1.20.0 ggbeeswarm_0.7.2         </span></span>
<span><span class="co">##  [53] minqa_1.2.5               MCL_1.0                  </span></span>
<span><span class="co">##  [55] reticulate_1.31           clusterProfiler_4.6.2    </span></span>
<span><span class="co">##  [57] beeswarm_0.4.0            xfun_0.40                </span></span>
<span><span class="co">##  [59] bslib_0.5.1               zoo_1.8-12               </span></span>
<span><span class="co">##  [61] tidyselect_1.2.0          reshape2_1.4.4           </span></span>
<span><span class="co">##  [63] purrr_1.0.2               kernlab_0.9-32           </span></span>
<span><span class="co">##  [65] ica_1.0-3                 gson_0.1.0               </span></span>
<span><span class="co">##  [67] viridisLite_0.4.2         pkgbuild_1.4.2           </span></span>
<span><span class="co">##  [69] rlang_1.1.1               jquerylib_0.1.4          </span></span>
<span><span class="co">##  [71] glue_1.6.2                RColorBrewer_1.1-3       </span></span>
<span><span class="co">##  [73] stringr_1.5.0             ggsignif_0.6.4           </span></span>
<span><span class="co">##  [75] DEsingle_1.18.1           httpuv_1.6.11            </span></span>
<span><span class="co">##  [77] BiocNeighbors_1.16.0      jsonlite_1.8.7           </span></span>
<span><span class="co">##  [79] bit_4.0.5                 mime_0.12                </span></span>
<span><span class="co">##  [81] systemfonts_1.0.4         AdaptGauss_1.5.6         </span></span>
<span><span class="co">##  [83] gridExtra_2.3             Rsamtools_2.14.0         </span></span>
<span><span class="co">##  [85] stringi_1.7.12            processx_3.8.2           </span></span>
<span><span class="co">##  [87] RcppRoll_0.3.0            spatstat.sparse_3.0-2    </span></span>
<span><span class="co">##  [89] scattermore_1.2           spatstat.explore_3.2-1   </span></span>
<span><span class="co">##  [91] yulab.utils_0.0.7         bitops_1.0-7             </span></span>
<span><span class="co">##  [93] cli_3.6.1                 Polychrome_1.5.1         </span></span>
<span><span class="co">##  [95] RSQLite_2.3.1             tidyr_1.3.0              </span></span>
<span><span class="co">##  [97] pheatmap_1.0.12           data.table_1.14.8        </span></span>
<span><span class="co">##  [99] rstudioapi_0.15.0         org.Mm.eg.db_3.16.0      </span></span>
<span><span class="co">## [101] GenomicAlignments_1.34.1  nlme_3.1-157             </span></span>
<span><span class="co">## [103] qvalue_2.30.0             scran_1.26.2             </span></span>
<span><span class="co">## [105] locfit_1.5-9.8            listenv_0.9.0            </span></span>
<span><span class="co">## [107] miniUI_0.1.1.1            gridGraphics_0.5-1       </span></span>
<span><span class="co">## [109] urlchecker_1.0.1          dbplyr_2.3.3             </span></span>
<span><span class="co">## [111] segmented_1.6-4           sessioninfo_1.2.2        </span></span>
<span><span class="co">## [113] qualV_0.3-5               lifecycle_1.0.3          </span></span>
<span><span class="co">## [115] munsell_0.5.0             codetools_0.2-18         </span></span>
<span><span class="co">## [117] vipor_0.4.5               lmtest_0.9-40            </span></span>
<span><span class="co">## [119] xtable_1.8-4              ROCR_1.0-11              </span></span>
<span><span class="co">## [121] BiocManager_1.30.22       scatterplot3d_0.3-44     </span></span>
<span><span class="co">## [123] abind_1.4-5               farver_2.1.1             </span></span>
<span><span class="co">## [125] FNN_1.1.3.2               parallelly_1.36.0        </span></span>
<span><span class="co">## [127] anocva_0.1.1              RANN_2.6.1               </span></span>
<span><span class="co">## [129] aplot_0.2.0               sparsesvd_0.2-2          </span></span>
<span><span class="co">## [131] ggtree_3.6.2              BiocIO_1.8.0             </span></span>
<span><span class="co">## [133] RcppAnnoy_0.0.21          goftest_1.2-3            </span></span>
<span><span class="co">## [135] patchwork_1.1.2           tibble_3.2.1             </span></span>
<span><span class="co">## [137] profvis_0.3.8             cluster_2.1.3            </span></span>
<span><span class="co">## [139] future.apply_1.11.0       tidytree_0.4.5           </span></span>
<span><span class="co">## [141] ellipsis_0.3.2            prettyunits_1.1.1        </span></span>
<span><span class="co">## [143] ggridges_0.5.4            fgsea_1.24.0             </span></span>
<span><span class="co">## [145] remotes_2.4.2.1           slam_0.1-50              </span></span>
<span><span class="co">## [147] spatstat.utils_3.0-3      htmltools_0.5.6          </span></span>
<span><span class="co">## [149] BiocFileCache_2.6.1       yaml_2.3.7               </span></span>
<span><span class="co">## [151] utf8_1.2.3                plotly_4.10.2            </span></span>
<span><span class="co">## [153] XML_3.99-0.14             pkgdown_2.0.7            </span></span>
<span><span class="co">## [155] DrImpute_1.0              withr_2.5.0              </span></span>
<span><span class="co">## [157] scuttle_1.8.4             fitdistrplus_1.1-11      </span></span>
<span><span class="co">## [159] BiocParallel_1.32.6       bit64_4.0.5              </span></span>
<span><span class="co">## [161] ProtGenerics_1.30.0       progressr_0.14.0         </span></span>
<span><span class="co">## [163] GOSemSim_2.24.0           ragg_1.2.5               </span></span>
<span><span class="co">## [165] rsvd_1.0.5                ScaledMatrix_1.6.0       </span></span>
<span><span class="co">## [167] devtools_2.4.5            memoise_2.0.1            </span></span>
<span><span class="co">## [169] evaluate_0.21             VGAM_1.1-8               </span></span>
<span><span class="co">## [171] RApiSerialize_0.1.2       callr_3.7.3              </span></span>
<span><span class="co">## [173] gamlss_5.4-12             gamlss.data_6.0-2        </span></span>
<span><span class="co">## [175] ps_1.7.5                  curl_5.0.1               </span></span>
<span><span class="co">## [177] fansi_1.0.4               tensor_1.5               </span></span>
<span><span class="co">## [179] edgeR_3.40.2              regioneR_1.30.0          </span></span>
<span><span class="co">## [181] cachem_1.0.8              org.Hs.eg.db_3.16.0      </span></span>
<span><span class="co">## [183] desc_1.4.2                rTRM_1.36.0              </span></span>
<span><span class="co">## [185] deldir_1.0-9              HDO.db_0.99.1            </span></span>
<span><span class="co">## [187] maxLik_1.5-2              metapod_1.6.0            </span></span>
<span><span class="co">## [189] rjson_0.2.21              rstatix_0.7.2            </span></span>
<span><span class="co">## [191] ggrepel_0.9.3             rprojroot_2.0.3          </span></span>
<span><span class="co">## [193] tools_4.2.1               sass_0.4.7               </span></span>
<span><span class="co">## [195] sandwich_3.0-2            magrittr_2.0.3           </span></span>
<span><span class="co">## [197] RCurl_1.98-1.12           car_3.1-2                </span></span>
<span><span class="co">## [199] ape_5.7-1                 ggplotify_0.1.2          </span></span>
<span><span class="co">## [201] xml2_1.3.5                httr_1.4.6               </span></span>
<span><span class="co">## [203] assertthat_0.2.1          rmarkdown_2.23           </span></span>
<span><span class="co">## [205] boot_1.3-28               globals_0.16.2           </span></span>
<span><span class="co">## [207] R6_2.5.1                  progress_1.2.2           </span></span>
<span><span class="co">## [209] KEGGREST_1.38.0           treeio_1.22.0            </span></span>
<span><span class="co">## [211] statmod_1.5.0             beachmat_2.14.2          </span></span>
<span><span class="co">## [213] BiocSingular_1.14.0       splines_4.2.1            </span></span>
<span><span class="co">## [215] carData_3.0-5             ggfun_0.1.2              </span></span>
<span><span class="co">## [217] colorspace_2.1-0          generics_0.1.3           </span></span>
<span><span class="co">## [219] pracma_2.4.2              pillar_1.9.0             </span></span>
<span><span class="co">## [221] tweenr_2.0.2              sp_2.0-0                 </span></span>
<span><span class="co">## [223] GenomeInfoDbData_1.2.9    plyr_1.8.8               </span></span>
<span><span class="co">## [225] gtable_0.3.3              bdsmatrix_1.3-6          </span></span>
<span><span class="co">## [227] stringfish_0.15.8         restfulr_0.0.15          </span></span>
<span><span class="co">## [229] WriteXLS_6.4.0            gamlss.dist_6.0-5        </span></span>
<span><span class="co">## [231] shadowtext_0.1.2          biomaRt_2.54.1           </span></span>
<span><span class="co">## [233] fastmap_1.1.1             pscl_1.5.5.1             </span></span>
<span><span class="co">## [235] broom_1.0.5               scales_1.2.1             </span></span>
<span><span class="co">## [237] filelock_1.0.2            backports_1.4.1          </span></span>
<span><span class="co">## [239] lme4_1.1-34               enrichplot_1.18.4        </span></span>
<span><span class="co">## [241] hms_1.1.3                 ggforce_0.4.1            </span></span>
<span><span class="co">## [243] Rtsne_0.16                shiny_1.7.4.1            </span></span>
<span><span class="co">## [245] polyclip_1.10-4           grid_4.2.1               </span></span>
<span><span class="co">## [247] numDeriv_2016.8-1.1       glasso_1.11              </span></span>
<span><span class="co">## [249] bbmle_1.0.25              lazyeval_0.2.2           </span></span>
<span><span class="co">## [251] crayon_1.5.2              MASS_7.3-57              </span></span>
<span><span class="co">## [253] downloader_0.4            sparseMatrixStats_1.10.0 </span></span>
<span><span class="co">## [255] viridis_0.6.4             compiler_4.2.1           </span></span>
<span><span class="co">## [257] spatstat.geom_3.2-2</span></span></code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Yang Li, Anjun Ma, Yizhong Wang, Cankun Wang.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
