<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Identify cell-type-specific enhancer regulons (eRegulons) — get_cts_en_regs • stream2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Identify cell-type-specific enhancer regulons (eRegulons) — get_cts_en_regs"><meta property="og:description" content="Enhancer regulons (eRegulons) are active in various cell subpopulations. Some
of them may be enriched in one or more cell types. The successful identification of eRegulons
at the single-cell level can improve the detection of heterogeneous transcriptional regulatory
mechanisms across various cell types and allows for reliable constructions of global gene regulatory
networks encoded in complex diseases. Hence, it is critical to study cell-type-specific eRegulons.
eRegulons specific to a cell type are called cell-type-specific eRegulons. This function takes as
input a Seurat object (composed of scRNA-seq and scATAC-seq) and cell-type-specific enhancer-drive
gene regulatory networks (eGRNs) and then identify the cell-type-specific eRegulons in each cell type/cluster."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stream2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/install.html">Install</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/data-simulation.html">Simulated data analyses</a>
    </li>
    <li>
      <a href="../articles/eRegulon-inference.html">Biological data analyses</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/YangLi-Bio/stream2" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Identify cell-type-specific enhancer regulons (eRegulons)</h1>
    <small class="dont-index">Source: <a href="https://github.com/YangLi-Bio/stream2/blob/HEAD/R/run_stream.R" class="external-link"><code>R/run_stream.R</code></a></small>
    <div class="hidden name"><code>get_cts_en_regs.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Enhancer regulons (eRegulons) are active in various cell subpopulations. Some
of them may be enriched in one or more cell types. The successful identification of eRegulons
at the single-cell level can improve the detection of heterogeneous transcriptional regulatory
mechanisms across various cell types and allows for reliable constructions of global gene regulatory
networks encoded in complex diseases. Hence, it is critical to study cell-type-specific eRegulons.
eRegulons specific to a cell type are called cell-type-specific eRegulons. This function takes as
input a <code>Seurat</code> object (composed of scRNA-seq and scATAC-seq) and cell-type-specific enhancer-drive
gene regulatory networks (eGRNs) and then identify the cell-type-specific eRegulons in each cell type/cluster.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">get_cts_en_regs</span><span class="op">(</span></span>
<span>  obj <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  peak.assay <span class="op">=</span> <span class="st">"ATAC"</span>,</span>
<span>  de.genes <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  celltype <span class="op">=</span> <span class="st">"celltype"</span>,</span>
<span>  cts.en.grns <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  accessibility <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  out.dir <span class="op">=</span> <span class="st">"./"</span>,</span>
<span>  min.pct <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  logfc.threshold <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  padj.cutoff <span class="op">=</span> <span class="fl">0.05</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>obj</dt>
<dd><p>An <code>Seurat</code> object.</p></dd>


<dt>peak.assay</dt>
<dd><p>The chromatin accessibility assay, "ATAC" by default.</p></dd>


<dt>de.genes</dt>
<dd><p>A list of differentially expressed genes (DEGs).</p></dd>


<dt>celltype</dt>
<dd><p>The metadata column indicating the cell types or clusters, "seurat_clusters" by default.</p></dd>


<dt>cts.en.grns</dt>
<dd><p>Cell-type-specific eGRNs saved in <code>GRanges</code> object, the name of each of which
is cell type, and the <code>GRanges</code> object contains metadata columns "gene" and "TF".</p></dd>


<dt>accessibility</dt>
<dd><p>Whether perform differential accessibility analysis, FALSE by default.</p></dd>


<dt>out.dir</dt>
<dd><p>The directory to save the intermediate results or final results, "./" by default.</p></dd>


<dt>padj.cutoff</dt>
<dd><p>The cutoff of adjusted p-values of differential expression, 0.05 by default.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Returns a list of cell-type-specific eRegulons,  each of which contains the following attributes:</p><ul><li><p>TF: The TF of the cell-type-specific eRegulons.</p></li>
<li><p>genes: Genes of the cell-type-specific eRegulons.</p></li>
<li><p>enhancers: Enhancers of the cell-type-specific eRegulons.</p></li>
<li><p>cells: Cells where the celltype.</p></li>
<li><p>links: The enhancer-gene relations saved in <code>GRanges</code> object.</p></li>
<li><p>celltype: The celltype of the cell-type-specific eRegulons.</p></li>
</ul></div>
    <div id="references">
    <h2>References</h2>
    <p>Li, Y., Ma, A., Wang, Y., Wang, C., Chen, S., Fu, H., Liu, B. and Ma, Q., 2022.
Enhancer-driven gene regulatory networks inference from single-cell RNA-seq and ATAC-seq data.
bioRxiv, pp.2022-12.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yang Li, Anjun Ma, Yizhong Wang, Qi Guo, Cankun Wang, Bingqiang Liu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

