<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Identify enhancer regulons (eRegulons) from jointly profiled scRNA-seq and scATAC-seq data — run_stream • stream2</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Identify enhancer regulons (eRegulons) from jointly profiled scRNA-seq and scATAC-seq data — run_stream"><meta property="og:description" content="Most transcriptional regulation analyses, including prediction of transcription factor
(TF) binding sites and identification of enhancer-gene relations require the discovery of enhancer
regulons (eRegulons), which are active in a set of cells. A TF with its set of target enhancers
and target genes is called an eRegulon. This function takes as input a Seurat
object (containing both scRNA-seq and scATAC-seq) and then simultaneously identifies eRegulons as well
as the cell sets where eRegulons are active. In addition, this function determines the number of eRegulons
in a dataset by submodular optimization."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">stream2</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.0.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/install.html">Install</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/data-simulation.html">Simulated data analyses</a>
    </li>
    <li>
      <a href="../articles/eRegulon-inference.html">Biological data analyses</a>
    </li>
  </ul></li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/YangLi-Bio/stream2" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Identify enhancer regulons (eRegulons) from jointly profiled scRNA-seq and scATAC-seq data</h1>
    <small class="dont-index">Source: <a href="https://github.com/YangLi-Bio/stream2/blob/HEAD/R/run_stream.R" class="external-link"><code>R/run_stream.R</code></a></small>
    <div class="hidden name"><code>run_stream.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Most transcriptional regulation analyses, including prediction of transcription factor
(TF) binding sites and identification of enhancer-gene relations require the discovery of enhancer
regulons (eRegulons), which are active in a set of cells. A TF with its set of target enhancers
and target genes is called an eRegulon. This function takes as input a <code>Seurat</code>
object (containing both scRNA-seq and scATAC-seq) and then simultaneously identifies eRegulons as well
as the cell sets where eRegulons are active. In addition, this function determines the number of eRegulons
in a dataset by submodular optimization.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">run_stream</span><span class="op">(</span></span>
<span>  obj <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  qubic.path <span class="op">=</span> <span class="st">"/users/PAS1475/liyang/software/QUBIC2/qubic"</span>,</span>
<span>  candidate.TFs <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  peak.assay <span class="op">=</span> <span class="st">"ATAC"</span>,</span>
<span>  var.genes <span class="op">=</span> <span class="fl">3000</span>,</span>
<span>  top.peaks <span class="op">=</span> <span class="fl">3000</span>,</span>
<span>  ifPutativeTFs <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  min.cells <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  out.dir <span class="op">=</span> <span class="st">"./"</span>,</span>
<span>  org <span class="op">=</span> <span class="st">"hg38"</span>,</span>
<span>  top.ngenes <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  c.cutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  n.blocks <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  distance <span class="op">=</span> <span class="fl">5e+05</span>,</span>
<span>  filter_peaks_for_cicero <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  ifWeighted <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  cicero.covar <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span>,</span>
<span>  signac.score <span class="op">=</span> <span class="op">-</span><span class="cn">Inf</span>,</span>
<span>  signac.pval <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  intra.cutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  inter.cutoff <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  peak.cutoff <span class="op">=</span> <span class="fl">0.8</span>,</span>
<span>  score.cutoff <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  KL <span class="op">=</span> <span class="st">"min.exp"</span>,</span>
<span>  quantile.cutoff <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  BlockOverlap <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  Extension <span class="op">=</span> <span class="fl">0.9</span>,</span>
<span>  submod.step <span class="op">=</span> <span class="fl">30</span>,</span>
<span>  min.eRegs <span class="op">=</span> <span class="fl">100</span>,</span>
<span>  url.link <span class="op">=</span> <span class="st">"https://figshare.com/ndownloader/files/38794185"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>obj</dt>
<dd><p>A <code>Seurat</code> object composed of both scRNA-seq and scATAC-seq assays.</p></dd>


<dt>qubic.path</dt>
<dd><p>The path of the binary executable file of QUBIC2, e.g., "/users/PAS1475/liyang/software/QUBIC2/qubic"</p></dd>


<dt>candidate.TFs</dt>
<dd><p>The list of candidate TFs used to identify eRegulons and eGRNs, NULL by default.</p></dd>


<dt>peak.assay</dt>
<dd><p>The scATAC-seq assay, "ATAC" by default.</p></dd>


<dt>var.genes</dt>
<dd><p>The number of highly variable genes to predict used to identify the core part of HBCs,
3000 by default.</p></dd>


<dt>top.peaks</dt>
<dd><p>The number of top-ranked peaks to identify the core part of hybrid biclusters (HBCs),
3000 by default.</p></dd>


<dt>min.cells</dt>
<dd><p>The cutoff of minimum number of cells for quality control (QC), 10 by default.</p></dd>


<dt>out.dir</dt>
<dd><p>The directory to save the intermediate or final results, "./" by default.</p></dd>


<dt>org</dt>
<dd><p>The organism version, hg38 by default.</p></dd>


<dt>top.ngenes</dt>
<dd><p>The number of genes composing the core part of an HBC, 5 by default.</p></dd>


<dt>c.cutoff</dt>
<dd><p>The cutoff of consistency during hybrid biclustering process, 1.0 by default.</p></dd>


<dt>n.blocks</dt>
<dd><p>The cutoff of the maximum number of blocks output by IRIS-FGM, 100 by default.</p></dd>


<dt>distance</dt>
<dd><p>The distance cutoff to build enhancer-enhancer relations, 5e+05 by default.</p></dd>


<dt>filter_peaks_for_cicero</dt>
<dd><p>Whether filter the peaks in a neighborhood of the <code>Signac</code> results
before running <code>cicero</code>, FALSE by default.</p></dd>


<dt>intra.cutoff</dt>
<dd><p>The cutoff to calculate pairwise similarity among HBCs associated with the same TFs,
1.0 by default.</p></dd>


<dt>inter.cutoff</dt>
<dd><p>The cutoff to compute pairwise similarity among genes in HBCs associated with different TFs,
0.80 by default.</p></dd>


<dt>peak.cutoff</dt>
<dd><p>The cutoff to quantify pairwise similarity among enhancers in HBCs associated with
different TFs, 0.80 by default.</p></dd>


<dt>KL</dt>
<dd><p>Which method to use for measuring the score of HBCs, "min.exp" by default, i.e.,
the smaller one between the numbers of genes and cells in a HBC.</p></dd>


<dt>quantile.cutoff</dt>
<dd><p>The quantile cutoff of the ratio of HBC cells, where enhancers are accessible,
4 by default, indicating the top-25% among ranks.</p></dd>


<dt>BlockOverlap</dt>
<dd><p>The cutoff of maximum overlap between blocks output by <code>IRIS-FGM</code>, 0.50 by default.</p></dd>


<dt>Extension</dt>
<dd><p>The consistency level to expand a block by <code>IRIS-FGM</code>, 0.70 by default.</p></dd>


<dt>submod.step</dt>
<dd><p>The step size of, i.e., the number of HBCs to add in each step during iteration,
for submodular optimization, 30 by default.</p></dd>


<dt>min.eRegs</dt>
<dd><p>The minimum number of enhancer regulons (eRegulons) to output, 100 by default.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>When running on a <code>Seurat</code> object,
returns a list of eRegulons saved in a nested list, each of which contains the following attributes:</p><ul><li><p>terminal: The <code>IRIS-FGM</code> block used to predict the eRegulon.</p></li>
<li><p>Tier: The tier of the TF-enhancer relations: 1 represents JASPAR annotations;
2 denotes motif scanning.</p></li>
<li><p>TF: The TF of the eRegulon.</p></li>
<li><p>genes: Genes of the eRegulon.</p></li>
<li><p>peaks: Enhancers of the eRegulon.</p></li>
<li><p>cells: Cells where the eRegulon is active.</p></li>
<li><p>atac.ratio: The ratio of cells where the eRegulon enhancers are
accessible against cells in which the eRegulon genes are expressed.</p></li>
<li><p>score: The eRegulon score.</p></li>
<li><p>weight: The eRegulon weight.</p></li>
<li><p>links: The enhancer-gene relations saved in <code>GRanges</code> object.</p></li>
<li><p>seed: The seed to obtain the eRegulon.</p></li>
</ul></div>
    <div id="references">
    <h2>References</h2>
    <p>Li, Y., Ma, A., Wang, Y., Wang, C., Chen, S., Fu, H., Liu, B. and Ma, Q., 2022.
Enhancer-driven gene regulatory networks inference from single-cell RNA-seq and ATAC-seq data.
bioRxiv, pp.2022-12.</p>
<p>Chang, Y., Allen, C., Wan, C., Chung, D., Zhang, C., Li, Z. and Ma, Q., 2021.
IRIS-FGM: an integrative single-cell RNA-Seq interpretation system for functional gene module analysis.
Bioinformatics, 37(18), pp.3045-3047.</p>
<p>Xie, J., Ma, A., Zhang, Y., Liu, B., Cao, S., Wang, C., ... &amp; Ma, Q. (2020).
QUBIC2: a novel and robust biclustering algorithm for analyses and interpretation of
large-scale RNA-Seq data. Bioinformatics, 36(4), 1143-1149.</p>
<p>Stuart, T., Srivastava, A., Madad, S., Lareau, C. A., &amp; Satija, R. (2021).
Single-cell chromatin state analysis with Signac. Nature methods, 18(11), 1333-1341.</p>
<p>Pliner, H. A., Packer, J. S., McFaline-Figueroa, J. L., Cusanovich,
D. A., Daza, R. M., Aghamirzaie, D., ... &amp; Trapnell, C. (2018).
Cicero predicts cis-regulatory DNA interactions from single-cell chromatin accessibility data.
Molecular cell, 71(5), 858-871.</p>
<p>Castro-Mondragon, J. A., Riudavets-Puig, R., Rauluseviciute, I., Berhanu Lemma, R.,
Turchi, L., Blanc-Mathieu, R., ... &amp; Mathelier, A. (2022).
JASPAR 2022: the 9th release of the open-access database of transcription factor binding profiles.
Nucleic acids research, 50(D1), D165-D173.</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Yang Li, Anjun Ma, Yizhong Wang, Qi Guo, Cankun Wang, Bingqiang Liu.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

